<?xml version="1.0" encoding="ascii"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
          "DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
  <title>dose.genetic.Chromosome</title>
  <link rel="stylesheet" href="epydoc.css" type="text/css" />
  <script type="text/javascript" src="epydoc.js"></script>
</head>

<body bgcolor="white" text="black" link="blue" vlink="#204080"
      alink="#204080">
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="dose-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            ><a class="navbar" target="_top" href="https://github.com/mauriceling/dose">https://github.com/mauriceling/dose</a></th>
          </tr></table></th>
  </tr>
</table>
<table width="100%" cellpadding="0" cellspacing="0">
  <tr valign="top">
    <td width="100%">
      <span class="breadcrumbs">
        <a href="dose-module.html">Package&nbsp;dose</a> ::
        <a href="dose.genetic-module.html">Module&nbsp;genetic</a> ::
        Class&nbsp;Chromosome
      </span>
    </td>
    <td>
      <table cellpadding="0" cellspacing="0">
        <!-- hide/show private -->
        <tr><td align="right"><span class="options">[<a href="javascript:void(0);" class="privatelink"
    onclick="toggle_private();">hide&nbsp;private</a>]</span></td></tr>
        <tr><td align="right"><span class="options"
            >[<a href="frames.html" target="_top">frames</a
            >]&nbsp;|&nbsp;<a href="dose.genetic.Chromosome-class.html"
            target="_top">no&nbsp;frames</a>]</span></td></tr>
      </table>
    </td>
  </tr>
</table>
<!-- ==================== CLASS DESCRIPTION ==================== -->
<h1 class="epydoc">Class Chromosome</h1><p class="nomargin-top"><span class="codelink"><a href="dose.genetic-pysrc.html#Chromosome">source&nbsp;code</a></span></p>
<pre class="base-tree">
object --+
         |
        <strong class="uidshort">Chromosome</strong>
</pre>

<hr />
<p>Representation of a linear chromosome.</p>

<hr />
<div class="fields">      <p><strong>See Also:</strong>
        Lim, JZR, Aw, ZQ, Goh, DJW, How, JA, Low, SXZ, Loo, BZL, Ling, MHT.
        2010. A genetic algorithm framework grounded in biology. The Python
        Papers Source Codes 2: 6.
      </p>
      <p><strong>Since:</strong>
        version 0.4
      </p>
</div><!-- ==================== INSTANCE METHODS ==================== -->
<a name="section-InstanceMethods"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Instance Methods</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-InstanceMethods"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="dose.genetic.Chromosome-class.html#__init__" class="summary-sig-name">__init__</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">sequence</span>,
        <span class="summary-sig-arg">base</span>,
        <span class="summary-sig-arg">background_mutation</span>=<span class="summary-sig-default">0.0001</span>)</span><br />
      Sets up a chromosome.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="dose.genetic-pysrc.html#Chromosome.__init__">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="dose.genetic.Chromosome-class.html#rmutate" class="summary-sig-name">rmutate</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">type</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">point</code><code class="variable-quote">'</code></span>,
        <span class="summary-sig-arg">rate</span>=<span class="summary-sig-default">0.01</span>,
        <span class="summary-sig-arg">start</span>=<span class="summary-sig-default">0</span>,
        <span class="summary-sig-arg">end</span>=<span class="summary-sig-default">-1</span>)</span><br />
      Random Mutation operator - to simulate random point, insertion, 
      deletion, inversion, gene translocation and gene duplication events.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="dose.genetic-pysrc.html#Chromosome.rmutate">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="dose.genetic.Chromosome-class.html#kmutate" class="summary-sig-name">kmutate</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">type</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">point</code><code class="variable-quote">'</code></span>,
        <span class="summary-sig-arg">start</span>=<span class="summary-sig-default">0</span>,
        <span class="summary-sig-arg">end</span>=<span class="summary-sig-default">0</span>,
        <span class="summary-sig-arg">sequence</span>=<span class="summary-sig-default">None</span>,
        <span class="summary-sig-arg">tpos</span>=<span class="summary-sig-default">0</span>)</span><br />
      Known Mutation operator - to simulate a known point, insertion, 
      deletion, inversion, gene translocation or gene duplication event.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="dose.genetic-pysrc.html#Chromosome.kmutate">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="dose.genetic.Chromosome-class.html#replicate" class="summary-sig-name">replicate</a>(<span class="summary-sig-arg">self</span>)</span><br />
      Replicates (deep copy) the chromosome.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="dose.genetic-pysrc.html#Chromosome.replicate">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
  <tr>
    <td colspan="2" class="summary">
    <p class="indent-wrapped-lines"><b>Inherited from <code>object</code></b>:
      <code>__delattr__</code>,
      <code>__format__</code>,
      <code>__getattribute__</code>,
      <code>__hash__</code>,
      <code>__new__</code>,
      <code>__reduce__</code>,
      <code>__reduce_ex__</code>,
      <code>__repr__</code>,
      <code>__setattr__</code>,
      <code>__sizeof__</code>,
      <code>__str__</code>,
      <code>__subclasshook__</code>
      </p>
    </td>
  </tr>
</table>
<!-- ==================== PROPERTIES ==================== -->
<a name="section-Properties"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Properties</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-Properties"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
  <tr>
    <td colspan="2" class="summary">
    <p class="indent-wrapped-lines"><b>Inherited from <code>object</code></b>:
      <code>__class__</code>
      </p>
    </td>
  </tr>
</table>
<!-- ==================== METHOD DETAILS ==================== -->
<a name="section-MethodDetails"></a>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Method Details</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-MethodDetails"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
</table>
<a name="__init__"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">__init__</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">sequence</span>,
        <span class="sig-arg">base</span>,
        <span class="sig-arg">background_mutation</span>=<span class="sig-default">0.0001</span>)</span>
    <br /><em class="fname">(Constructor)</em>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="dose.genetic-pysrc.html#Chromosome.__init__">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Sets up a chromosome.</p>
  <dl class="fields">
    <dt>Parameters:</dt>
    <dd><ul class="nomargin-top">
        <li><strong class="pname"><code>sequence</code></strong> - a subscriptable object (list or string) representing the sequence
          of the chromosome.</li>
        <li><strong class="pname"><code>base</code></strong> - a subscriptable object (list or string) representing allowable 
          entities in the sequence.</li>
        <li><strong class="pname"><code>background_mutation</code></strong> - background mutation rate represented as the probability of number
          of mutations per base. Default = 0.0001 (0.01%).</li>
    </ul></dd>
    <dt>Overrides:
        object.__init__
    </dt>
  </dl>
<div class="fields">      <p><strong>Since:</strong>
        version 0.4
      </p>
</div></td></tr></table>
</div>
<a name="rmutate"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">rmutate</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">type</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">point</code><code class="variable-quote">'</code></span>,
        <span class="sig-arg">rate</span>=<span class="sig-default">0.01</span>,
        <span class="sig-arg">start</span>=<span class="sig-default">0</span>,
        <span class="sig-arg">end</span>=<span class="sig-default">-1</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="dose.genetic-pysrc.html#Chromosome.rmutate">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Random Mutation operator - to simulate random point, insertion, 
  deletion, inversion, gene translocation and gene duplication events.</p>
  <p>The start and end parameters are useful for simulating mutational 
  hotspots in the genome.</p>
  <dl class="fields">
    <dt>Parameters:</dt>
    <dd><ul class="nomargin-top">
        <li><strong class="pname"><code>type</code></strong> - type of mutation. Accepts 'point' (point mutation), 'insert' 
          (insert a base), 'delete' (delete a base), 'invert' (invert a 
          stretch of the chromosome), 'duplicate' (duplicate a stretch of 
          the chromosome), 'translocate' (translocate a stretch of 
          chromosome to another random position). Default = point.</li>
        <li><strong class="pname"><code>rate</code></strong> - probability of mutation per base above background mutation rate. 
          Default = 0.01 (1%). No mutation event will ever happen if (rate 
          + background_mutation) is less than zero.</li>
        <li><strong class="pname"><code>start</code></strong> - starting base on the sequence for mutation. Default = 0, start of
          the genome.</li>
        <li><strong class="pname"><code>end</code></strong> - last base on the sequence for mutation. Default = -1, end of the 
          genome.</li>
    </ul></dd>
  </dl>
<div class="fields">      <p><strong>Since:</strong>
        version 0.4
      </p>
</div></td></tr></table>
</div>
<a name="kmutate"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">kmutate</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">type</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">point</code><code class="variable-quote">'</code></span>,
        <span class="sig-arg">start</span>=<span class="sig-default">0</span>,
        <span class="sig-arg">end</span>=<span class="sig-default">0</span>,
        <span class="sig-arg">sequence</span>=<span class="sig-default">None</span>,
        <span class="sig-arg">tpos</span>=<span class="sig-default">0</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="dose.genetic-pysrc.html#Chromosome.kmutate">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Known Mutation operator - to simulate a known point, insertion, 
  deletion, inversion, gene translocation or gene duplication event.</p>
  <p>The required parameters will be determined by the type of 
  mutation:</p>
  <ul>
    <li>
      point requires
      <ul>
        <li>
          start (the position of the base to mutate)
        </li>
        <li>
          sequence (the new base)
        </li>
      </ul>
    </li>
    <li>
      delete requires
      <ul>
        <li>
          start (the position of the base to delete)
        </li>
      </ul>
    </li>
    <li>
      insert requires
      <ul>
        <li>
          start (the position of the base to begin insertion)
        </li>
        <li>
          sequence (list of sequence or a base to insert)
        </li>
      </ul>
    </li>
    <li>
      invert requires
      <ul>
        <li>
          start (the position of the base to start inversion)
        </li>
        <li>
          end (the position of the base to end inversion)
        </li>
      </ul>
    </li>
    <li>
      duplicate requires
      <ul>
        <li>
          start (the position of the starting base to duplicate)
        </li>
        <li>
          end (the position of the last base to duplicate)
        </li>
      </ul>
    </li>
    <li>
      translocate requires
      <ul>
        <li>
          start (the position of the base to start translocation)
        </li>
        <li>
          end (the position of the base to end translocation)
        </li>
        <li>
          tpos (the position of the base insert the translocated sequence)
        </li>
      </ul>
    </li>
  </ul>
  <dl class="fields">
    <dt>Parameters:</dt>
    <dd><ul class="nomargin-top">
        <li><strong class="pname"><code>type</code></strong> - type of mutation. Accepts 'point' (point mutation), 'insert' 
          (insert a base), 'delete' (delete a base), 'invert' (invert a 
          stretch of the chromosome), 'duplicate' (duplicate a stretch of 
          the chromosome), 'translocate' (translocate a stretch of 
          chromosome to another random position). Default = point.</li>
        <li><strong class="pname"><code>start</code></strong> - starting base on the chromosome for mutation. Default = 0, start 
          of the genome.</li>
        <li><strong class="pname"><code>end</code></strong> - last base on the chromosome for mutation. Default = 0.</li>
        <li><strong class="pname"><code>sequence</code></strong> - list of bases to change to (point mutation) or to insert 
          (insertion mutation)</li>
        <li><strong class="pname"><code>tpos</code></strong> - position of the chromosome to insert the translocated sequence.</li>
    </ul></dd>
  </dl>
<div class="fields">      <p><strong>Since:</strong>
        version 0.4
      </p>
</div></td></tr></table>
</div>
<a name="replicate"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">replicate</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="dose.genetic-pysrc.html#Chromosome.replicate">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Replicates (deep copy) the chromosome.</p>
  <dl class="fields">
    <dt>Returns:</dt>
        <dd>a copy of current chromosome.</dd>
  </dl>
<div class="fields">      <p><strong>Since:</strong>
        version 0.4
      </p>
</div></td></tr></table>
</div>
<br />
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="dose-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            ><a class="navbar" target="_top" href="https://github.com/mauriceling/dose">https://github.com/mauriceling/dose</a></th>
          </tr></table></th>
  </tr>
</table>
<table border="0" cellpadding="0" cellspacing="0" width="100%%">
  <tr>
    <td align="left" class="footer">
    Generated by Epydoc 3.0.1 on Thu Feb 15 15:37:56 2018
    </td>
    <td align="right" class="footer">
      <a target="mainFrame" href="http://epydoc.sourceforge.net"
        >http://epydoc.sourceforge.net</a>
    </td>
  </tr>
</table>

<script type="text/javascript">
  <!--
  // Private objects are initially displayed (because if
  // javascript is turned off then we want them to be
  // visible); but by default, we want to hide them.  So hide
  // them unless we have a cookie that says to show them.
  checkCookie();
  // -->
</script>
</body>
</html>
