<?xml version="1.0" encoding="ascii"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
          "DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
  <title>dose.genetic</title>
  <link rel="stylesheet" href="epydoc.css" type="text/css" />
  <script type="text/javascript" src="epydoc.js"></script>
</head>

<body bgcolor="white" text="black" link="blue" vlink="#204080"
      alink="#204080">
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="dose-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            ><a class="navbar" target="_top" href="https://github.com/mauriceling/dose">https://github.com/mauriceling/dose</a></th>
          </tr></table></th>
  </tr>
</table>
<table width="100%" cellpadding="0" cellspacing="0">
  <tr valign="top">
    <td width="100%">
      <span class="breadcrumbs">
        <a href="dose-module.html">Package&nbsp;dose</a> ::
        Module&nbsp;genetic
      </span>
    </td>
    <td>
      <table cellpadding="0" cellspacing="0">
        <!-- hide/show private -->
        <tr><td align="right"><span class="options">[<a href="javascript:void(0);" class="privatelink"
    onclick="toggle_private();">hide&nbsp;private</a>]</span></td></tr>
        <tr><td align="right"><span class="options"
            >[<a href="frames.html" target="_top">frames</a
            >]&nbsp;|&nbsp;<a href="dose.genetic-pysrc.html"
            target="_top">no&nbsp;frames</a>]</span></td></tr>
      </table>
    </td>
  </tr>
</table>
<h1 class="epydoc">Source Code for <a href="dose.genetic-module.html">Module dose.genetic</a></h1>
<pre class="py-src">
<a name="L1"></a><tt class="py-lineno">  1</tt>  <tt class="py-line"><tt class="py-docstring">"""</tt> </tt>
<a name="L2"></a><tt class="py-lineno">  2</tt>  <tt class="py-line"><tt class="py-docstring">Framework for Genetic Algorithm Applications.</tt> </tt>
<a name="L3"></a><tt class="py-lineno">  3</tt>  <tt class="py-line"><tt class="py-docstring">Date created: 23rd February 2010</tt> </tt>
<a name="L4"></a><tt class="py-lineno">  4</tt>  <tt class="py-line"><tt class="py-docstring"></tt> </tt>
<a name="L5"></a><tt class="py-lineno">  5</tt>  <tt class="py-line"><tt class="py-docstring">@see: Lim, JZR, Aw, ZQ, Goh, DJW, How, JA, Low, SXZ, Loo, BZL,</tt> </tt>
<a name="L6"></a><tt class="py-lineno">  6</tt>  <tt class="py-line"><tt class="py-docstring">Ling, MHT. 2010. A genetic algorithm framework grounded in biology.</tt> </tt>
<a name="L7"></a><tt class="py-lineno">  7</tt>  <tt class="py-line"><tt class="py-docstring">The Python Papers Source Codes 2: 6. </tt> </tt>
<a name="L8"></a><tt class="py-lineno">  8</tt>  <tt class="py-line"><tt class="py-docstring">"""</tt> </tt>
<a name="L9"></a><tt class="py-lineno">  9</tt>  <tt class="py-line"><tt class="py-keyword">import</tt> <tt id="link-0" class="py-name" targets="Method dose.copads.statisticsdistribution.BetaDistribution.random()=dose.copads.statisticsdistribution.BetaDistribution-class.html#random,Method dose.copads.statisticsdistribution.BradfordDistribution.random()=dose.copads.statisticsdistribution.BradfordDistribution-class.html#random,Method dose.copads.statisticsdistribution.BurrDistribution.random()=dose.copads.statisticsdistribution.BurrDistribution-class.html#random,Method dose.copads.statisticsdistribution.CauchyDistribution.random()=dose.copads.statisticsdistribution.CauchyDistribution-class.html#random,Method dose.copads.statisticsdistribution.ExponentialDistribution.random()=dose.copads.statisticsdistribution.ExponentialDistribution-class.html#random,Method dose.copads.statisticsdistribution.GumbelDistribution.random()=dose.copads.statisticsdistribution.GumbelDistribution-class.html#random,Method dose.copads.statisticsdistribution.LogNormalDistribution.random()=dose.copads.statisticsdistribution.LogNormalDistribution-class.html#random,Method dose.copads.statisticsdistribution.NormalDistribution.random()=dose.copads.statisticsdistribution.NormalDistribution-class.html#random,Method dose.copads.statisticsdistribution.ParetoDistribution.random()=dose.copads.statisticsdistribution.ParetoDistribution-class.html#random,Method dose.copads.statisticsdistribution.UniformDistribution.random()=dose.copads.statisticsdistribution.UniformDistribution-class.html#random,Method dose.copads.statisticsdistribution.WeiBullDistribution.random()=dose.copads.statisticsdistribution.WeiBullDistribution-class.html#random"><a title="dose.copads.statisticsdistribution.BetaDistribution.random
dose.copads.statisticsdistribution.BradfordDistribution.random
dose.copads.statisticsdistribution.BurrDistribution.random
dose.copads.statisticsdistribution.CauchyDistribution.random
dose.copads.statisticsdistribution.ExponentialDistribution.random
dose.copads.statisticsdistribution.GumbelDistribution.random
dose.copads.statisticsdistribution.LogNormalDistribution.random
dose.copads.statisticsdistribution.NormalDistribution.random
dose.copads.statisticsdistribution.ParetoDistribution.random
dose.copads.statisticsdistribution.UniformDistribution.random
dose.copads.statisticsdistribution.WeiBullDistribution.random" class="py-name" href="#" onclick="return doclink('link-0', 'random', 'link-0');">random</a></tt><tt class="py-op">,</tt> <tt class="py-name">os</tt><tt class="py-op">,</tt> <tt class="py-name">string</tt> </tt>
<a name="L10"></a><tt class="py-lineno"> 10</tt>  <tt class="py-line"><tt class="py-keyword">from</tt> <tt class="py-name">copy</tt> <tt class="py-keyword">import</tt> <tt class="py-name">deepcopy</tt> </tt>
<a name="L11"></a><tt class="py-lineno"> 11</tt>  <tt class="py-line"> </tt>
<a name="Chromosome"></a><div id="Chromosome-def"><a name="L12"></a><tt class="py-lineno"> 12</tt> <a class="py-toggle" href="#" id="Chromosome-toggle" onclick="return toggle('Chromosome');">-</a><tt class="py-line"><tt class="py-keyword">class</tt> <a class="py-def-name" href="dose.genetic.Chromosome-class.html">Chromosome</a><tt class="py-op">(</tt><tt class="py-base-class">object</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Chromosome-collapsed" style="display:none;" pad="+++" indent="++++"></div><div id="Chromosome-expanded"><a name="L13"></a><tt class="py-lineno"> 13</tt>  <tt class="py-line">    <tt class="py-docstring">"""</tt> </tt>
<a name="L14"></a><tt class="py-lineno"> 14</tt>  <tt class="py-line"><tt class="py-docstring">    Representation of a linear chromosome.</tt> </tt>
<a name="L15"></a><tt class="py-lineno"> 15</tt>  <tt class="py-line"><tt class="py-docstring"></tt> </tt>
<a name="L16"></a><tt class="py-lineno"> 16</tt>  <tt class="py-line"><tt class="py-docstring">    @see: Lim, JZR, Aw, ZQ, Goh, DJW, How, JA, Low, SXZ, Loo, BZL,</tt> </tt>
<a name="L17"></a><tt class="py-lineno"> 17</tt>  <tt class="py-line"><tt class="py-docstring">    Ling, MHT. 2010. A genetic algorithm framework grounded in biology.</tt> </tt>
<a name="L18"></a><tt class="py-lineno"> 18</tt>  <tt class="py-line"><tt class="py-docstring">    The Python Papers Source Codes 2: 6.</tt> </tt>
<a name="L19"></a><tt class="py-lineno"> 19</tt>  <tt class="py-line"><tt class="py-docstring">    </tt> </tt>
<a name="L20"></a><tt class="py-lineno"> 20</tt>  <tt class="py-line"><tt class="py-docstring">    @since: version 0.4</tt> </tt>
<a name="L21"></a><tt class="py-lineno"> 21</tt>  <tt class="py-line"><tt class="py-docstring">    """</tt> </tt>
<a name="Chromosome.__init__"></a><div id="Chromosome.__init__-def"><a name="L22"></a><tt class="py-lineno"> 22</tt> <a class="py-toggle" href="#" id="Chromosome.__init__-toggle" onclick="return toggle('Chromosome.__init__');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Chromosome-class.html#__init__">__init__</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">,</tt> <tt class="py-param">sequence</tt><tt class="py-op">,</tt> <tt class="py-param">base</tt><tt class="py-op">,</tt>  </tt>
<a name="L23"></a><tt class="py-lineno"> 23</tt>  <tt class="py-line">                 <tt class="py-param">background_mutation</tt><tt class="py-op">=</tt><tt class="py-number">0.0001</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Chromosome.__init__-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Chromosome.__init__-expanded"><a name="L24"></a><tt class="py-lineno"> 24</tt>  <tt class="py-line">        <tt class="py-docstring">"""</tt> </tt>
<a name="L25"></a><tt class="py-lineno"> 25</tt>  <tt class="py-line"><tt class="py-docstring">        Sets up a chromosome.</tt> </tt>
<a name="L26"></a><tt class="py-lineno"> 26</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L27"></a><tt class="py-lineno"> 27</tt>  <tt class="py-line"><tt class="py-docstring">        @param sequence: a subscriptable object (list or string) representing</tt> </tt>
<a name="L28"></a><tt class="py-lineno"> 28</tt>  <tt class="py-line"><tt class="py-docstring">            the sequence of the chromosome.</tt> </tt>
<a name="L29"></a><tt class="py-lineno"> 29</tt>  <tt class="py-line"><tt class="py-docstring">        @param base: a subscriptable object (list or string) representing</tt> </tt>
<a name="L30"></a><tt class="py-lineno"> 30</tt>  <tt class="py-line"><tt class="py-docstring">            allowable entities in the sequence.</tt> </tt>
<a name="L31"></a><tt class="py-lineno"> 31</tt>  <tt class="py-line"><tt class="py-docstring">        @param background_mutation: background mutation rate represented as the </tt> </tt>
<a name="L32"></a><tt class="py-lineno"> 32</tt>  <tt class="py-line"><tt class="py-docstring">            probability of number of mutations per base. Default = 0.0001 </tt> </tt>
<a name="L33"></a><tt class="py-lineno"> 33</tt>  <tt class="py-line"><tt class="py-docstring">            (0.01%).</tt> </tt>
<a name="L34"></a><tt class="py-lineno"> 34</tt>  <tt class="py-line"><tt class="py-docstring">            </tt> </tt>
<a name="L35"></a><tt class="py-lineno"> 35</tt>  <tt class="py-line"><tt class="py-docstring">        @since: version 0.4</tt> </tt>
<a name="L36"></a><tt class="py-lineno"> 36</tt>  <tt class="py-line"><tt class="py-docstring">        """</tt> </tt>
<a name="L37"></a><tt class="py-lineno"> 37</tt>  <tt class="py-line">        <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt> <tt class="py-op">=</tt> <tt class="py-name">sequence</tt> </tt>
<a name="L38"></a><tt class="py-lineno"> 38</tt>  <tt class="py-line">        <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">base</tt> <tt class="py-op">=</tt> <tt class="py-name">base</tt> </tt>
<a name="L39"></a><tt class="py-lineno"> 39</tt>  <tt class="py-line">        <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">background_mutation</tt> <tt class="py-op">=</tt> <tt class="py-name">background_mutation</tt> </tt>
</div><a name="L40"></a><tt class="py-lineno"> 40</tt>  <tt class="py-line">     </tt>
<a name="Chromosome.rmutate"></a><div id="Chromosome.rmutate-def"><a name="L41"></a><tt class="py-lineno"> 41</tt> <a class="py-toggle" href="#" id="Chromosome.rmutate-toggle" onclick="return toggle('Chromosome.rmutate');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Chromosome-class.html#rmutate">rmutate</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">,</tt> <tt class="py-param">type</tt><tt class="py-op">=</tt><tt class="py-string">'point'</tt><tt class="py-op">,</tt> <tt class="py-param">rate</tt><tt class="py-op">=</tt><tt class="py-number">0.01</tt><tt class="py-op">,</tt> <tt class="py-param">start</tt><tt class="py-op">=</tt><tt class="py-number">0</tt><tt class="py-op">,</tt> <tt class="py-param">end</tt><tt class="py-op">=</tt><tt class="py-op">-</tt><tt class="py-number">1</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Chromosome.rmutate-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Chromosome.rmutate-expanded"><a name="L42"></a><tt class="py-lineno"> 42</tt>  <tt class="py-line">        <tt class="py-docstring">"""</tt> </tt>
<a name="L43"></a><tt class="py-lineno"> 43</tt>  <tt class="py-line"><tt class="py-docstring">        Random Mutation operator - to simulate random point, insertion, </tt> </tt>
<a name="L44"></a><tt class="py-lineno"> 44</tt>  <tt class="py-line"><tt class="py-docstring">        deletion, inversion, gene translocation and gene duplication events.</tt> </tt>
<a name="L45"></a><tt class="py-lineno"> 45</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L46"></a><tt class="py-lineno"> 46</tt>  <tt class="py-line"><tt class="py-docstring">        The start and end parameters are useful for simulating mutational </tt> </tt>
<a name="L47"></a><tt class="py-lineno"> 47</tt>  <tt class="py-line"><tt class="py-docstring">        hotspots in the genome.</tt> </tt>
<a name="L48"></a><tt class="py-lineno"> 48</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L49"></a><tt class="py-lineno"> 49</tt>  <tt class="py-line"><tt class="py-docstring">        @param type: type of mutation. Accepts 'point' (point mutation), </tt> </tt>
<a name="L50"></a><tt class="py-lineno"> 50</tt>  <tt class="py-line"><tt class="py-docstring">            'insert' (insert a base), 'delete' (delete a base), 'invert'</tt> </tt>
<a name="L51"></a><tt class="py-lineno"> 51</tt>  <tt class="py-line"><tt class="py-docstring">            (invert a stretch of the chromosome), 'duplicate' (duplicate a</tt> </tt>
<a name="L52"></a><tt class="py-lineno"> 52</tt>  <tt class="py-line"><tt class="py-docstring">            stretch of the chromosome), 'translocate' (translocate a </tt> </tt>
<a name="L53"></a><tt class="py-lineno"> 53</tt>  <tt class="py-line"><tt class="py-docstring">            stretch of chromosome to another random position). </tt> </tt>
<a name="L54"></a><tt class="py-lineno"> 54</tt>  <tt class="py-line"><tt class="py-docstring">            Default = point.</tt> </tt>
<a name="L55"></a><tt class="py-lineno"> 55</tt>  <tt class="py-line"><tt class="py-docstring">        @param rate: probability of mutation per base above background</tt> </tt>
<a name="L56"></a><tt class="py-lineno"> 56</tt>  <tt class="py-line"><tt class="py-docstring">            mutation rate. Default = 0.01 (1%). No mutation event will ever </tt> </tt>
<a name="L57"></a><tt class="py-lineno"> 57</tt>  <tt class="py-line"><tt class="py-docstring">            happen if (rate + background_mutation) is less than zero.</tt> </tt>
<a name="L58"></a><tt class="py-lineno"> 58</tt>  <tt class="py-line"><tt class="py-docstring">        @param start: starting base on the sequence for mutation.</tt> </tt>
<a name="L59"></a><tt class="py-lineno"> 59</tt>  <tt class="py-line"><tt class="py-docstring">            Default = 0, start of the genome.</tt> </tt>
<a name="L60"></a><tt class="py-lineno"> 60</tt>  <tt class="py-line"><tt class="py-docstring">        @param end: last base on the sequence for mutation. Default = -1, </tt> </tt>
<a name="L61"></a><tt class="py-lineno"> 61</tt>  <tt class="py-line"><tt class="py-docstring">            end of the genome.</tt> </tt>
<a name="L62"></a><tt class="py-lineno"> 62</tt>  <tt class="py-line"><tt class="py-docstring">            </tt> </tt>
<a name="L63"></a><tt class="py-lineno"> 63</tt>  <tt class="py-line"><tt class="py-docstring">        @since: version 0.4</tt> </tt>
<a name="L64"></a><tt class="py-lineno"> 64</tt>  <tt class="py-line"><tt class="py-docstring">        """</tt> </tt>
<a name="L65"></a><tt class="py-lineno"> 65</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">end</tt> <tt class="py-op">&gt;</tt> <tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">)</tt> <tt class="py-op">-</tt> <tt class="py-number">1</tt><tt class="py-op">:</tt> <tt class="py-name">end</tt> <tt class="py-op">=</tt> <tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">)</tt> <tt class="py-op">-</tt> <tt class="py-number">1</tt> </tt>
<a name="L66"></a><tt class="py-lineno"> 66</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">end</tt> <tt class="py-op">==</tt> <tt class="py-op">-</tt><tt class="py-number">1</tt><tt class="py-op">:</tt> <tt class="py-name">end</tt> <tt class="py-op">=</tt> <tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">)</tt> <tt class="py-op">-</tt> <tt class="py-number">1</tt> </tt>
<a name="L67"></a><tt class="py-lineno"> 67</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">start</tt> <tt class="py-op">==</tt> <tt class="py-name">end</tt><tt class="py-op">:</tt> <tt class="py-name">start</tt> <tt class="py-op">=</tt> <tt class="py-number">0</tt> </tt>
<a name="L68"></a><tt class="py-lineno"> 68</tt>  <tt class="py-line">        <tt class="py-name">length</tt> <tt class="py-op">=</tt> <tt class="py-name">int</tt><tt class="py-op">(</tt><tt class="py-name">end</tt> <tt class="py-op">-</tt> <tt class="py-name">start</tt><tt class="py-op">)</tt> </tt>
<a name="L69"></a><tt class="py-lineno"> 69</tt>  <tt class="py-line">        <tt class="py-name">mutation</tt> <tt class="py-op">=</tt> <tt class="py-name">int</tt><tt class="py-op">(</tt><tt class="py-op">(</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">background_mutation</tt> <tt class="py-op">+</tt> <tt class="py-name">rate</tt><tt class="py-op">)</tt> <tt class="py-op">*</tt> <tt class="py-name">length</tt><tt class="py-op">)</tt> </tt>
<a name="L70"></a><tt class="py-lineno"> 70</tt>  <tt class="py-line">        <tt class="py-keyword">while</tt> <tt class="py-name">mutation</tt> <tt class="py-op">&gt;</tt> <tt class="py-number">0</tt><tt class="py-op">:</tt> </tt>
<a name="L71"></a><tt class="py-lineno"> 71</tt>  <tt class="py-line">            <tt class="py-name">position</tt> <tt class="py-op">=</tt> <tt class="py-name">int</tt><tt class="py-op">(</tt><tt class="py-name">start</tt><tt class="py-op">)</tt> <tt class="py-op">+</tt> <tt id="link-1" class="py-name"><a title="dose.copads.statisticsdistribution.BetaDistribution.random
dose.copads.statisticsdistribution.BradfordDistribution.random
dose.copads.statisticsdistribution.BurrDistribution.random
dose.copads.statisticsdistribution.CauchyDistribution.random
dose.copads.statisticsdistribution.ExponentialDistribution.random
dose.copads.statisticsdistribution.GumbelDistribution.random
dose.copads.statisticsdistribution.LogNormalDistribution.random
dose.copads.statisticsdistribution.NormalDistribution.random
dose.copads.statisticsdistribution.ParetoDistribution.random
dose.copads.statisticsdistribution.UniformDistribution.random
dose.copads.statisticsdistribution.WeiBullDistribution.random" class="py-name" href="#" onclick="return doclink('link-1', 'random', 'link-0');">random</a></tt><tt class="py-op">.</tt><tt class="py-name">randrange</tt><tt class="py-op">(</tt><tt class="py-name">length</tt> <tt class="py-op">-</tt> <tt class="py-number">1</tt><tt class="py-op">)</tt> </tt>
<a name="L72"></a><tt class="py-lineno"> 72</tt>  <tt class="py-line">            <tt class="py-name">new_base</tt> <tt class="py-op">=</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">base</tt><tt class="py-op">[</tt><tt id="link-2" class="py-name"><a title="dose.copads.statisticsdistribution.BetaDistribution.random
dose.copads.statisticsdistribution.BradfordDistribution.random
dose.copads.statisticsdistribution.BurrDistribution.random
dose.copads.statisticsdistribution.CauchyDistribution.random
dose.copads.statisticsdistribution.ExponentialDistribution.random
dose.copads.statisticsdistribution.GumbelDistribution.random
dose.copads.statisticsdistribution.LogNormalDistribution.random
dose.copads.statisticsdistribution.NormalDistribution.random
dose.copads.statisticsdistribution.ParetoDistribution.random
dose.copads.statisticsdistribution.UniformDistribution.random
dose.copads.statisticsdistribution.WeiBullDistribution.random" class="py-name" href="#" onclick="return doclink('link-2', 'random', 'link-0');">random</a></tt><tt class="py-op">.</tt><tt class="py-name">randrange</tt><tt class="py-op">(</tt><tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">base</tt><tt class="py-op">)</tt><tt class="py-op">)</tt><tt class="py-op">]</tt> </tt>
<a name="L73"></a><tt class="py-lineno"> 73</tt>  <tt class="py-line">            <tt class="py-keyword">if</tt> <tt class="py-name">type</tt> <tt class="py-op">==</tt> <tt class="py-string">'point'</tt><tt class="py-op">:</tt>  </tt>
<a name="L74"></a><tt class="py-lineno"> 74</tt>  <tt class="py-line">                <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">[</tt><tt class="py-name">position</tt><tt class="py-op">]</tt> <tt class="py-op">=</tt> <tt class="py-name">new_base</tt> </tt>
<a name="L75"></a><tt class="py-lineno"> 75</tt>  <tt class="py-line">            <tt class="py-keyword">if</tt> <tt class="py-name">type</tt> <tt class="py-op">==</tt> <tt class="py-string">'delete'</tt><tt class="py-op">:</tt>  </tt>
<a name="L76"></a><tt class="py-lineno"> 76</tt>  <tt class="py-line">                <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">.</tt><tt class="py-name">pop</tt><tt class="py-op">(</tt><tt class="py-name">position</tt><tt class="py-op">)</tt> </tt>
<a name="L77"></a><tt class="py-lineno"> 77</tt>  <tt class="py-line">            <tt class="py-keyword">if</tt> <tt class="py-name">type</tt> <tt class="py-op">==</tt> <tt class="py-string">'insert'</tt><tt class="py-op">:</tt>  </tt>
<a name="L78"></a><tt class="py-lineno"> 78</tt>  <tt class="py-line">                <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">.</tt><tt class="py-name">insert</tt><tt class="py-op">(</tt><tt class="py-name">position</tt><tt class="py-op">,</tt> <tt class="py-name">new_base</tt><tt class="py-op">)</tt> </tt>
<a name="L79"></a><tt class="py-lineno"> 79</tt>  <tt class="py-line">            <tt class="py-keyword">if</tt> <tt class="py-name">type</tt> <tt class="py-op">==</tt> <tt class="py-string">'duplicate'</tt><tt class="py-op">:</tt> </tt>
<a name="L80"></a><tt class="py-lineno"> 80</tt>  <tt class="py-line">                <tt class="py-name">end_pos</tt> <tt class="py-op">=</tt> <tt id="link-3" class="py-name"><a title="dose.copads.statisticsdistribution.BetaDistribution.random
dose.copads.statisticsdistribution.BradfordDistribution.random
dose.copads.statisticsdistribution.BurrDistribution.random
dose.copads.statisticsdistribution.CauchyDistribution.random
dose.copads.statisticsdistribution.ExponentialDistribution.random
dose.copads.statisticsdistribution.GumbelDistribution.random
dose.copads.statisticsdistribution.LogNormalDistribution.random
dose.copads.statisticsdistribution.NormalDistribution.random
dose.copads.statisticsdistribution.ParetoDistribution.random
dose.copads.statisticsdistribution.UniformDistribution.random
dose.copads.statisticsdistribution.WeiBullDistribution.random" class="py-name" href="#" onclick="return doclink('link-3', 'random', 'link-0');">random</a></tt><tt class="py-op">.</tt><tt class="py-name">randrange</tt><tt class="py-op">(</tt><tt class="py-name">position</tt> <tt class="py-op">+</tt> <tt class="py-number">1</tt><tt class="py-op">,</tt> <tt class="py-name">end</tt><tt class="py-op">)</tt> </tt>
<a name="L81"></a><tt class="py-lineno"> 81</tt>  <tt class="py-line">                <tt class="py-name">fragment</tt> <tt class="py-op">=</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">[</tt><tt class="py-name">position</tt><tt class="py-op">:</tt><tt class="py-name">end_pos</tt><tt class="py-op">]</tt> </tt>
<a name="L82"></a><tt class="py-lineno"> 82</tt>  <tt class="py-line">                <tt class="py-keyword">for</tt> <tt class="py-name">i</tt> <tt class="py-keyword">in</tt> <tt id="link-4" class="py-name" targets="Method dose.copads.samplestatistics.SingleSample.range()=dose.copads.samplestatistics.SingleSample-class.html#range"><a title="dose.copads.samplestatistics.SingleSample.range" class="py-name" href="#" onclick="return doclink('link-4', 'range', 'link-4');">range</a></tt><tt class="py-op">(</tt><tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">fragment</tt><tt class="py-op">)</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
<a name="L83"></a><tt class="py-lineno"> 83</tt>  <tt class="py-line">                    <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">.</tt><tt class="py-name">insert</tt><tt class="py-op">(</tt><tt class="py-name">end_pos</tt> <tt class="py-op">+</tt> <tt class="py-name">i</tt><tt class="py-op">,</tt> <tt class="py-name">fragment</tt><tt class="py-op">[</tt><tt class="py-name">i</tt><tt class="py-op">]</tt><tt class="py-op">)</tt> </tt>
<a name="L84"></a><tt class="py-lineno"> 84</tt>  <tt class="py-line">            <tt class="py-keyword">if</tt> <tt class="py-name">type</tt> <tt class="py-op">==</tt> <tt class="py-string">'invert'</tt><tt class="py-op">:</tt> </tt>
<a name="L85"></a><tt class="py-lineno"> 85</tt>  <tt class="py-line">                <tt class="py-name">end_pos</tt> <tt class="py-op">=</tt> <tt id="link-5" class="py-name"><a title="dose.copads.statisticsdistribution.BetaDistribution.random
dose.copads.statisticsdistribution.BradfordDistribution.random
dose.copads.statisticsdistribution.BurrDistribution.random
dose.copads.statisticsdistribution.CauchyDistribution.random
dose.copads.statisticsdistribution.ExponentialDistribution.random
dose.copads.statisticsdistribution.GumbelDistribution.random
dose.copads.statisticsdistribution.LogNormalDistribution.random
dose.copads.statisticsdistribution.NormalDistribution.random
dose.copads.statisticsdistribution.ParetoDistribution.random
dose.copads.statisticsdistribution.UniformDistribution.random
dose.copads.statisticsdistribution.WeiBullDistribution.random" class="py-name" href="#" onclick="return doclink('link-5', 'random', 'link-0');">random</a></tt><tt class="py-op">.</tt><tt class="py-name">randrange</tt><tt class="py-op">(</tt><tt class="py-name">position</tt> <tt class="py-op">+</tt> <tt class="py-number">1</tt><tt class="py-op">,</tt> <tt class="py-name">end</tt><tt class="py-op">)</tt> </tt>
<a name="L86"></a><tt class="py-lineno"> 86</tt>  <tt class="py-line">                <tt class="py-name">fragment</tt> <tt class="py-op">=</tt> <tt class="py-op">[</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">.</tt><tt class="py-name">pop</tt><tt class="py-op">(</tt><tt class="py-name">position</tt><tt class="py-op">)</tt>  </tt>
<a name="L87"></a><tt class="py-lineno"> 87</tt>  <tt class="py-line">                            <tt class="py-keyword">for</tt> <tt class="py-name">i</tt> <tt class="py-keyword">in</tt> <tt id="link-6" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.range" class="py-name" href="#" onclick="return doclink('link-6', 'range', 'link-4');">range</a></tt><tt class="py-op">(</tt><tt class="py-name">end_pos</tt> <tt class="py-op">-</tt> <tt class="py-name">position</tt><tt class="py-op">)</tt><tt class="py-op">]</tt> </tt>
<a name="L88"></a><tt class="py-lineno"> 88</tt>  <tt class="py-line">                <tt class="py-name">fragment</tt><tt class="py-op">.</tt><tt class="py-name">reverse</tt><tt class="py-op">(</tt><tt class="py-op">)</tt> </tt>
<a name="L89"></a><tt class="py-lineno"> 89</tt>  <tt class="py-line">                <tt class="py-keyword">for</tt> <tt class="py-name">base</tt> <tt class="py-keyword">in</tt> <tt class="py-name">fragment</tt><tt class="py-op">:</tt> </tt>
<a name="L90"></a><tt class="py-lineno"> 90</tt>  <tt class="py-line">                    <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">.</tt><tt class="py-name">insert</tt><tt class="py-op">(</tt><tt class="py-name">position</tt><tt class="py-op">,</tt> <tt class="py-name">base</tt><tt class="py-op">)</tt> </tt>
<a name="L91"></a><tt class="py-lineno"> 91</tt>  <tt class="py-line">            <tt class="py-keyword">if</tt> <tt class="py-name">type</tt> <tt class="py-op">==</tt> <tt class="py-string">'translocate'</tt><tt class="py-op">:</tt> </tt>
<a name="L92"></a><tt class="py-lineno"> 92</tt>  <tt class="py-line">                <tt class="py-name">end_pos</tt> <tt class="py-op">=</tt> <tt id="link-7" class="py-name"><a title="dose.copads.statisticsdistribution.BetaDistribution.random
dose.copads.statisticsdistribution.BradfordDistribution.random
dose.copads.statisticsdistribution.BurrDistribution.random
dose.copads.statisticsdistribution.CauchyDistribution.random
dose.copads.statisticsdistribution.ExponentialDistribution.random
dose.copads.statisticsdistribution.GumbelDistribution.random
dose.copads.statisticsdistribution.LogNormalDistribution.random
dose.copads.statisticsdistribution.NormalDistribution.random
dose.copads.statisticsdistribution.ParetoDistribution.random
dose.copads.statisticsdistribution.UniformDistribution.random
dose.copads.statisticsdistribution.WeiBullDistribution.random" class="py-name" href="#" onclick="return doclink('link-7', 'random', 'link-0');">random</a></tt><tt class="py-op">.</tt><tt class="py-name">randrange</tt><tt class="py-op">(</tt><tt class="py-name">position</tt> <tt class="py-op">+</tt> <tt class="py-number">1</tt><tt class="py-op">,</tt> <tt class="py-name">end</tt><tt class="py-op">)</tt> </tt>
<a name="L93"></a><tt class="py-lineno"> 93</tt>  <tt class="py-line">                <tt class="py-name">fragment</tt> <tt class="py-op">=</tt> <tt class="py-op">[</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">.</tt><tt class="py-name">pop</tt><tt class="py-op">(</tt><tt class="py-name">position</tt><tt class="py-op">)</tt>  </tt>
<a name="L94"></a><tt class="py-lineno"> 94</tt>  <tt class="py-line">                            <tt class="py-keyword">for</tt> <tt class="py-name">i</tt> <tt class="py-keyword">in</tt> <tt id="link-8" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.range" class="py-name" href="#" onclick="return doclink('link-8', 'range', 'link-4');">range</a></tt><tt class="py-op">(</tt><tt class="py-name">end_pos</tt> <tt class="py-op">-</tt> <tt class="py-name">position</tt><tt class="py-op">)</tt><tt class="py-op">]</tt> </tt>
<a name="L95"></a><tt class="py-lineno"> 95</tt>  <tt class="py-line">                <tt class="py-name">insertion_point</tt> <tt class="py-op">=</tt> <tt id="link-9" class="py-name"><a title="dose.copads.statisticsdistribution.BetaDistribution.random
dose.copads.statisticsdistribution.BradfordDistribution.random
dose.copads.statisticsdistribution.BurrDistribution.random
dose.copads.statisticsdistribution.CauchyDistribution.random
dose.copads.statisticsdistribution.ExponentialDistribution.random
dose.copads.statisticsdistribution.GumbelDistribution.random
dose.copads.statisticsdistribution.LogNormalDistribution.random
dose.copads.statisticsdistribution.NormalDistribution.random
dose.copads.statisticsdistribution.ParetoDistribution.random
dose.copads.statisticsdistribution.UniformDistribution.random
dose.copads.statisticsdistribution.WeiBullDistribution.random" class="py-name" href="#" onclick="return doclink('link-9', 'random', 'link-0');">random</a></tt><tt class="py-op">.</tt><tt class="py-name">randint</tt><tt class="py-op">(</tt><tt class="py-number">0</tt><tt class="py-op">,</tt> <tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">)</tt><tt class="py-op">)</tt> </tt>
<a name="L96"></a><tt class="py-lineno"> 96</tt>  <tt class="py-line">                <tt class="py-keyword">for</tt> <tt class="py-name">i</tt> <tt class="py-keyword">in</tt> <tt id="link-10" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.range" class="py-name" href="#" onclick="return doclink('link-10', 'range', 'link-4');">range</a></tt><tt class="py-op">(</tt><tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">fragment</tt><tt class="py-op">)</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
<a name="L97"></a><tt class="py-lineno"> 97</tt>  <tt class="py-line">                    <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">.</tt><tt class="py-name">insert</tt><tt class="py-op">(</tt><tt class="py-name">insertion_point</tt> <tt class="py-op">+</tt> <tt class="py-name">i</tt><tt class="py-op">,</tt> <tt class="py-name">fragment</tt><tt class="py-op">[</tt><tt class="py-name">i</tt><tt class="py-op">]</tt><tt class="py-op">)</tt> </tt>
<a name="L98"></a><tt class="py-lineno"> 98</tt>  <tt class="py-line">            <tt class="py-name">mutation</tt> <tt class="py-op">=</tt> <tt class="py-name">mutation</tt> <tt class="py-op">-</tt> <tt class="py-number">1</tt> </tt>
</div><a name="L99"></a><tt class="py-lineno"> 99</tt>  <tt class="py-line">     </tt>
<a name="Chromosome.kmutate"></a><div id="Chromosome.kmutate-def"><a name="L100"></a><tt class="py-lineno">100</tt> <a class="py-toggle" href="#" id="Chromosome.kmutate-toggle" onclick="return toggle('Chromosome.kmutate');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Chromosome-class.html#kmutate">kmutate</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">,</tt> <tt class="py-param">type</tt><tt class="py-op">=</tt><tt class="py-string">'point'</tt><tt class="py-op">,</tt> <tt class="py-param">start</tt><tt class="py-op">=</tt><tt class="py-number">0</tt><tt class="py-op">,</tt> <tt class="py-param">end</tt><tt class="py-op">=</tt><tt class="py-number">0</tt><tt class="py-op">,</tt> <tt class="py-param">sequence</tt><tt class="py-op">=</tt><tt class="py-name">None</tt><tt class="py-op">,</tt> <tt class="py-param">tpos</tt><tt class="py-op">=</tt><tt class="py-number">0</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Chromosome.kmutate-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Chromosome.kmutate-expanded"><a name="L101"></a><tt class="py-lineno">101</tt>  <tt class="py-line">        <tt class="py-docstring">"""</tt> </tt>
<a name="L102"></a><tt class="py-lineno">102</tt>  <tt class="py-line"><tt class="py-docstring">        Known Mutation operator - to simulate a known point, insertion, </tt> </tt>
<a name="L103"></a><tt class="py-lineno">103</tt>  <tt class="py-line"><tt class="py-docstring">        deletion, inversion, gene translocation or gene duplication event.</tt> </tt>
<a name="L104"></a><tt class="py-lineno">104</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L105"></a><tt class="py-lineno">105</tt>  <tt class="py-line"><tt class="py-docstring">        The required parameters will be determined by the type of mutation:</tt> </tt>
<a name="L106"></a><tt class="py-lineno">106</tt>  <tt class="py-line"><tt class="py-docstring">            - point requires </tt> </tt>
<a name="L107"></a><tt class="py-lineno">107</tt>  <tt class="py-line"><tt class="py-docstring">                - start (the position of the base to mutate)</tt> </tt>
<a name="L108"></a><tt class="py-lineno">108</tt>  <tt class="py-line"><tt class="py-docstring">                - sequence (the new base)</tt> </tt>
<a name="L109"></a><tt class="py-lineno">109</tt>  <tt class="py-line"><tt class="py-docstring">            - delete requires</tt> </tt>
<a name="L110"></a><tt class="py-lineno">110</tt>  <tt class="py-line"><tt class="py-docstring">                - start (the position of the base to delete)</tt> </tt>
<a name="L111"></a><tt class="py-lineno">111</tt>  <tt class="py-line"><tt class="py-docstring">            - insert requires</tt> </tt>
<a name="L112"></a><tt class="py-lineno">112</tt>  <tt class="py-line"><tt class="py-docstring">                - start (the position of the base to begin insertion)</tt> </tt>
<a name="L113"></a><tt class="py-lineno">113</tt>  <tt class="py-line"><tt class="py-docstring">                - sequence (list of sequence or a base to insert)</tt> </tt>
<a name="L114"></a><tt class="py-lineno">114</tt>  <tt class="py-line"><tt class="py-docstring">            - invert requires</tt> </tt>
<a name="L115"></a><tt class="py-lineno">115</tt>  <tt class="py-line"><tt class="py-docstring">                - start (the position of the base to start inversion)</tt> </tt>
<a name="L116"></a><tt class="py-lineno">116</tt>  <tt class="py-line"><tt class="py-docstring">                - end (the position of the base to end inversion)</tt> </tt>
<a name="L117"></a><tt class="py-lineno">117</tt>  <tt class="py-line"><tt class="py-docstring">            - duplicate requires</tt> </tt>
<a name="L118"></a><tt class="py-lineno">118</tt>  <tt class="py-line"><tt class="py-docstring">                - start (the position of the starting base to duplicate)</tt> </tt>
<a name="L119"></a><tt class="py-lineno">119</tt>  <tt class="py-line"><tt class="py-docstring">                - end (the position of the last base to duplicate)</tt> </tt>
<a name="L120"></a><tt class="py-lineno">120</tt>  <tt class="py-line"><tt class="py-docstring">            - translocate requires</tt> </tt>
<a name="L121"></a><tt class="py-lineno">121</tt>  <tt class="py-line"><tt class="py-docstring">                - start (the position of the base to start translocation)</tt> </tt>
<a name="L122"></a><tt class="py-lineno">122</tt>  <tt class="py-line"><tt class="py-docstring">                - end (the position of the base to end translocation)</tt> </tt>
<a name="L123"></a><tt class="py-lineno">123</tt>  <tt class="py-line"><tt class="py-docstring">                - tpos (the position of the base insert the translocated </tt> </tt>
<a name="L124"></a><tt class="py-lineno">124</tt>  <tt class="py-line"><tt class="py-docstring">                    sequence)</tt> </tt>
<a name="L125"></a><tt class="py-lineno">125</tt>  <tt class="py-line"><tt class="py-docstring">                </tt> </tt>
<a name="L126"></a><tt class="py-lineno">126</tt>  <tt class="py-line"><tt class="py-docstring">        @param type: type of mutation. Accepts 'point' (point mutation), </tt> </tt>
<a name="L127"></a><tt class="py-lineno">127</tt>  <tt class="py-line"><tt class="py-docstring">            'insert' (insert a base), 'delete' (delete a base), 'invert'</tt> </tt>
<a name="L128"></a><tt class="py-lineno">128</tt>  <tt class="py-line"><tt class="py-docstring">            (invert a stretch of the chromosome), 'duplicate' (duplicate a</tt> </tt>
<a name="L129"></a><tt class="py-lineno">129</tt>  <tt class="py-line"><tt class="py-docstring">            stretch of the chromosome), 'translocate' (translocate a stretch </tt> </tt>
<a name="L130"></a><tt class="py-lineno">130</tt>  <tt class="py-line"><tt class="py-docstring">            of chromosome to another random position). Default = point.</tt> </tt>
<a name="L131"></a><tt class="py-lineno">131</tt>  <tt class="py-line"><tt class="py-docstring">        @param start: starting base on the chromosome for mutation.</tt> </tt>
<a name="L132"></a><tt class="py-lineno">132</tt>  <tt class="py-line"><tt class="py-docstring">            Default = 0, start of the genome.</tt> </tt>
<a name="L133"></a><tt class="py-lineno">133</tt>  <tt class="py-line"><tt class="py-docstring">        @param end: last base on the chromosome for mutation. Default = 0.</tt> </tt>
<a name="L134"></a><tt class="py-lineno">134</tt>  <tt class="py-line"><tt class="py-docstring">        @param sequence: list of bases to change to (point mutation) or to </tt> </tt>
<a name="L135"></a><tt class="py-lineno">135</tt>  <tt class="py-line"><tt class="py-docstring">            insert (insertion mutation)</tt> </tt>
<a name="L136"></a><tt class="py-lineno">136</tt>  <tt class="py-line"><tt class="py-docstring">        @param tpos: position of the chromosome to insert the translocated</tt> </tt>
<a name="L137"></a><tt class="py-lineno">137</tt>  <tt class="py-line"><tt class="py-docstring">            sequence.</tt> </tt>
<a name="L138"></a><tt class="py-lineno">138</tt>  <tt class="py-line"><tt class="py-docstring">            </tt> </tt>
<a name="L139"></a><tt class="py-lineno">139</tt>  <tt class="py-line"><tt class="py-docstring">        @since: version 0.4</tt> </tt>
<a name="L140"></a><tt class="py-lineno">140</tt>  <tt class="py-line"><tt class="py-docstring">        """</tt> </tt>
<a name="L141"></a><tt class="py-lineno">141</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">type</tt> <tt class="py-op">==</tt> <tt class="py-string">'point'</tt><tt class="py-op">:</tt> </tt>
<a name="L142"></a><tt class="py-lineno">142</tt>  <tt class="py-line">            <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">[</tt><tt class="py-name">start</tt><tt class="py-op">]</tt> <tt class="py-op">=</tt> <tt class="py-name">sequence</tt> </tt>
<a name="L143"></a><tt class="py-lineno">143</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">type</tt> <tt class="py-op">==</tt> <tt class="py-string">'delete'</tt><tt class="py-op">:</tt>  </tt>
<a name="L144"></a><tt class="py-lineno">144</tt>  <tt class="py-line">            <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">.</tt><tt class="py-name">pop</tt><tt class="py-op">(</tt><tt class="py-name">start</tt><tt class="py-op">)</tt> </tt>
<a name="L145"></a><tt class="py-lineno">145</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">type</tt> <tt class="py-op">==</tt> <tt class="py-string">'insert'</tt><tt class="py-op">:</tt> </tt>
<a name="L146"></a><tt class="py-lineno">146</tt>  <tt class="py-line">            <tt class="py-keyword">for</tt> <tt class="py-name">base</tt> <tt class="py-keyword">in</tt> <tt class="py-name">list</tt><tt class="py-op">(</tt><tt class="py-name">sequence</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
<a name="L147"></a><tt class="py-lineno">147</tt>  <tt class="py-line">                <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">.</tt><tt class="py-name">insert</tt><tt class="py-op">(</tt><tt class="py-name">start</tt><tt class="py-op">,</tt> <tt class="py-name">base</tt><tt class="py-op">)</tt> </tt>
<a name="L148"></a><tt class="py-lineno">148</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">type</tt> <tt class="py-op">==</tt> <tt class="py-string">'duplicate'</tt><tt class="py-op">:</tt> </tt>
<a name="L149"></a><tt class="py-lineno">149</tt>  <tt class="py-line">            <tt class="py-name">fragment</tt> <tt class="py-op">=</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">[</tt><tt class="py-name">start</tt><tt class="py-op">:</tt><tt class="py-name">end</tt> <tt class="py-op">+</tt> <tt class="py-number">1</tt><tt class="py-op">]</tt> </tt>
<a name="L150"></a><tt class="py-lineno">150</tt>  <tt class="py-line">            <tt class="py-keyword">for</tt> <tt class="py-name">i</tt> <tt class="py-keyword">in</tt> <tt id="link-11" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.range" class="py-name" href="#" onclick="return doclink('link-11', 'range', 'link-4');">range</a></tt><tt class="py-op">(</tt><tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">fragment</tt><tt class="py-op">)</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
<a name="L151"></a><tt class="py-lineno">151</tt>  <tt class="py-line">                <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">.</tt><tt class="py-name">insert</tt><tt class="py-op">(</tt><tt class="py-name">end</tt> <tt class="py-op">+</tt> <tt class="py-name">i</tt><tt class="py-op">,</tt> <tt class="py-name">fragment</tt><tt class="py-op">[</tt><tt class="py-name">i</tt><tt class="py-op">]</tt><tt class="py-op">)</tt> </tt>
<a name="L152"></a><tt class="py-lineno">152</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">type</tt> <tt class="py-op">==</tt> <tt class="py-string">'invert'</tt><tt class="py-op">:</tt> </tt>
<a name="L153"></a><tt class="py-lineno">153</tt>  <tt class="py-line">            <tt class="py-keyword">for</tt> <tt class="py-name">base</tt> <tt class="py-keyword">in</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">[</tt><tt class="py-name">start</tt><tt class="py-op">:</tt><tt class="py-name">end</tt><tt class="py-op">]</tt><tt class="py-op">:</tt> </tt>
<a name="L154"></a><tt class="py-lineno">154</tt>  <tt class="py-line">                <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">.</tt><tt class="py-name">insert</tt><tt class="py-op">(</tt><tt class="py-name">start</tt><tt class="py-op">,</tt> <tt class="py-name">base</tt><tt class="py-op">)</tt> </tt>
<a name="L155"></a><tt class="py-lineno">155</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">type</tt> <tt class="py-op">==</tt> <tt class="py-string">'translocate'</tt><tt class="py-op">:</tt> </tt>
<a name="L156"></a><tt class="py-lineno">156</tt>  <tt class="py-line">            <tt class="py-name">fragment</tt> <tt class="py-op">=</tt> <tt class="py-op">[</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">.</tt><tt class="py-name">pop</tt><tt class="py-op">(</tt><tt class="py-name">start</tt><tt class="py-op">)</tt>  </tt>
<a name="L157"></a><tt class="py-lineno">157</tt>  <tt class="py-line">                        <tt class="py-keyword">for</tt> <tt class="py-name">i</tt> <tt class="py-keyword">in</tt> <tt id="link-12" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.range" class="py-name" href="#" onclick="return doclink('link-12', 'range', 'link-4');">range</a></tt><tt class="py-op">(</tt><tt class="py-name">end</tt> <tt class="py-op">-</tt> <tt class="py-name">start</tt><tt class="py-op">)</tt><tt class="py-op">]</tt> </tt>
<a name="L158"></a><tt class="py-lineno">158</tt>  <tt class="py-line">            <tt class="py-keyword">for</tt> <tt class="py-name">i</tt> <tt class="py-keyword">in</tt> <tt id="link-13" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.range" class="py-name" href="#" onclick="return doclink('link-13', 'range', 'link-4');">range</a></tt><tt class="py-op">(</tt><tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">fragment</tt><tt class="py-op">)</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
<a name="L159"></a><tt class="py-lineno">159</tt>  <tt class="py-line">                <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">.</tt><tt class="py-name">insert</tt><tt class="py-op">(</tt><tt class="py-name">tpos</tt> <tt class="py-op">+</tt> <tt class="py-name">i</tt><tt class="py-op">,</tt> <tt class="py-name">fragment</tt><tt class="py-op">[</tt><tt class="py-name">i</tt><tt class="py-op">]</tt><tt class="py-op">)</tt> </tt>
</div><a name="L160"></a><tt class="py-lineno">160</tt>  <tt class="py-line"> </tt>
<a name="Chromosome.replicate"></a><div id="Chromosome.replicate-def"><a name="L161"></a><tt class="py-lineno">161</tt> <a class="py-toggle" href="#" id="Chromosome.replicate-toggle" onclick="return toggle('Chromosome.replicate');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Chromosome-class.html#replicate">replicate</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Chromosome.replicate-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Chromosome.replicate-expanded"><a name="L162"></a><tt class="py-lineno">162</tt>  <tt class="py-line">        <tt class="py-docstring">"""</tt> </tt>
<a name="L163"></a><tt class="py-lineno">163</tt>  <tt class="py-line"><tt class="py-docstring">        Replicates (deep copy) the chromosome.</tt> </tt>
<a name="L164"></a><tt class="py-lineno">164</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L165"></a><tt class="py-lineno">165</tt>  <tt class="py-line"><tt class="py-docstring">        @return: a copy of current chromosome.</tt> </tt>
<a name="L166"></a><tt class="py-lineno">166</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L167"></a><tt class="py-lineno">167</tt>  <tt class="py-line"><tt class="py-docstring">        @since: version 0.4</tt> </tt>
<a name="L168"></a><tt class="py-lineno">168</tt>  <tt class="py-line"><tt class="py-docstring">        """</tt> </tt>
<a name="L169"></a><tt class="py-lineno">169</tt>  <tt class="py-line">        <tt class="py-keyword">return</tt> <tt class="py-name">deepcopy</tt><tt class="py-op">(</tt><tt class="py-name">self</tt><tt class="py-op">)</tt> </tt>
</div></div><a name="L170"></a><tt class="py-lineno">170</tt>  <tt class="py-line">  </tt>
<a name="L171"></a><tt class="py-lineno">171</tt>  <tt class="py-line">         </tt>
<a name="Organism"></a><div id="Organism-def"><a name="L172"></a><tt class="py-lineno">172</tt> <a class="py-toggle" href="#" id="Organism-toggle" onclick="return toggle('Organism');">-</a><tt class="py-line"><tt class="py-keyword">class</tt> <a class="py-def-name" href="dose.genetic.Organism-class.html">Organism</a><tt class="py-op">(</tt><tt class="py-base-class">object</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Organism-collapsed" style="display:none;" pad="+++" indent="++++"></div><div id="Organism-expanded"><a name="L173"></a><tt class="py-lineno">173</tt>  <tt class="py-line">    <tt class="py-docstring">"""</tt> </tt>
<a name="L174"></a><tt class="py-lineno">174</tt>  <tt class="py-line"><tt class="py-docstring">    An organism represented by a list of chromosomes and a status table. </tt> </tt>
<a name="L175"></a><tt class="py-lineno">175</tt>  <tt class="py-line"><tt class="py-docstring">    This class should almost never be instantiated on its own but acts as </tt> </tt>
<a name="L176"></a><tt class="py-lineno">176</tt>  <tt class="py-line"><tt class="py-docstring">    an ancestor class as some functions need to be over-ridden in the </tt> </tt>
<a name="L177"></a><tt class="py-lineno">177</tt>  <tt class="py-line"><tt class="py-docstring">    inherited class for the desired use. </tt> </tt>
<a name="L178"></a><tt class="py-lineno">178</tt>  <tt class="py-line"><tt class="py-docstring">    </tt> </tt>
<a name="L179"></a><tt class="py-lineno">179</tt>  <tt class="py-line"><tt class="py-docstring">    Each organism is identifiable by a randomly generated 32-character </tt> </tt>
<a name="L180"></a><tt class="py-lineno">180</tt>  <tt class="py-line"><tt class="py-docstring">    'name' as Organism.identity.</tt> </tt>
<a name="L181"></a><tt class="py-lineno">181</tt>  <tt class="py-line"><tt class="py-docstring">    </tt> </tt>
<a name="L182"></a><tt class="py-lineno">182</tt>  <tt class="py-line"><tt class="py-docstring">    Methods to be over-ridden in the inherited class or substituted are</tt> </tt>
<a name="L183"></a><tt class="py-lineno">183</tt>  <tt class="py-line"><tt class="py-docstring">        - fitness</tt> </tt>
<a name="L184"></a><tt class="py-lineno">184</tt>  <tt class="py-line"><tt class="py-docstring">        - mutation_scheme</tt> </tt>
<a name="L185"></a><tt class="py-lineno">185</tt>  <tt class="py-line"><tt class="py-docstring">    </tt> </tt>
<a name="L186"></a><tt class="py-lineno">186</tt>  <tt class="py-line"><tt class="py-docstring">    Pre-defined status are</tt> </tt>
<a name="L187"></a><tt class="py-lineno">187</tt>  <tt class="py-line"><tt class="py-docstring">        1. alive - is the organism alive? True or False.</tt> </tt>
<a name="L188"></a><tt class="py-lineno">188</tt>  <tt class="py-line"><tt class="py-docstring">        2. vitality - percentage of maximum vitality.</tt> </tt>
<a name="L189"></a><tt class="py-lineno">189</tt>  <tt class="py-line"><tt class="py-docstring">        3. age - the current age of the organism.</tt> </tt>
<a name="L190"></a><tt class="py-lineno">190</tt>  <tt class="py-line"><tt class="py-docstring">        4. lifespan - pre-defined maximum lifespan to be set based on scenario. </tt> </tt>
<a name="L191"></a><tt class="py-lineno">191</tt>  <tt class="py-line"><tt class="py-docstring">        5. fitness - how fit the organism is? Set to maximum fitness.</tt> </tt>
<a name="L192"></a><tt class="py-lineno">192</tt>  <tt class="py-line"><tt class="py-docstring">        6. death - reason of death (as death code).</tt> </tt>
<a name="L193"></a><tt class="py-lineno">193</tt>  <tt class="py-line"><tt class="py-docstring">        7. parents - identity of parents.</tt> </tt>
<a name="L194"></a><tt class="py-lineno">194</tt>  <tt class="py-line"><tt class="py-docstring">        8. gender - gender of organism. </tt> </tt>
<a name="L195"></a><tt class="py-lineno">195</tt>  <tt class="py-line"><tt class="py-docstring">        9. blood - result of genomic interpretation or expression by </tt> </tt>
<a name="L196"></a><tt class="py-lineno">196</tt>  <tt class="py-line"><tt class="py-docstring">        Rajaraga interpreter.</tt> </tt>
<a name="L197"></a><tt class="py-lineno">197</tt>  <tt class="py-line"><tt class="py-docstring">        10. identity - 32-character randomly generated name.</tt> </tt>
<a name="L198"></a><tt class="py-lineno">198</tt>  <tt class="py-line"><tt class="py-docstring">        11. deme - defined as a sub-population or local population.</tt> </tt>
<a name="L199"></a><tt class="py-lineno">199</tt>  <tt class="py-line"><tt class="py-docstring">        12. location - location of the organism within the ecosystem.</tt> </tt>
<a name="L200"></a><tt class="py-lineno">200</tt>  <tt class="py-line"><tt class="py-docstring"></tt> </tt>
<a name="L201"></a><tt class="py-lineno">201</tt>  <tt class="py-line"><tt class="py-docstring">    </tt> </tt>
<a name="L202"></a><tt class="py-lineno">202</tt>  <tt class="py-line"><tt class="py-docstring">    List of defined death codes</tt> </tt>
<a name="L203"></a><tt class="py-lineno">203</tt>  <tt class="py-line"><tt class="py-docstring">        1. death01 - zero vitality</tt> </tt>
<a name="L204"></a><tt class="py-lineno">204</tt>  <tt class="py-line"><tt class="py-docstring">        2. death02 - maximum age reached</tt> </tt>
<a name="L205"></a><tt class="py-lineno">205</tt>  <tt class="py-line"><tt class="py-docstring">        3. death03 - unknown death cause</tt> </tt>
<a name="L206"></a><tt class="py-lineno">206</tt>  <tt class="py-line"><tt class="py-docstring"></tt> </tt>
<a name="L207"></a><tt class="py-lineno">207</tt>  <tt class="py-line"><tt class="py-docstring">    @see: Lim, JZR, Aw, ZQ, Goh, DJW, How, JA, Low, SXZ, Loo, BZL,</tt> </tt>
<a name="L208"></a><tt class="py-lineno">208</tt>  <tt class="py-line"><tt class="py-docstring">    Ling, MHT. 2010. A genetic algorithm framework grounded in biology.</tt> </tt>
<a name="L209"></a><tt class="py-lineno">209</tt>  <tt class="py-line"><tt class="py-docstring">    The Python Papers Source Codes 2: 6.</tt> </tt>
<a name="L210"></a><tt class="py-lineno">210</tt>  <tt class="py-line"><tt class="py-docstring">    </tt> </tt>
<a name="L211"></a><tt class="py-lineno">211</tt>  <tt class="py-line"><tt class="py-docstring">    @since: version 0.4</tt> </tt>
<a name="L212"></a><tt class="py-lineno">212</tt>  <tt class="py-line"><tt class="py-docstring">    """</tt> </tt>
<a name="L213"></a><tt class="py-lineno">213</tt>  <tt class="py-line">     </tt>
<a name="Organism.__init__"></a><div id="Organism.__init__-def"><a name="L214"></a><tt class="py-lineno">214</tt> <a class="py-toggle" href="#" id="Organism.__init__-toggle" onclick="return toggle('Organism.__init__');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Organism-class.html#__init__">__init__</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">,</tt> <tt class="py-param">genome</tt><tt class="py-op">=</tt><tt class="py-string">'dummy'</tt><tt class="py-op">,</tt> <tt class="py-param">mutation_type</tt><tt class="py-op">=</tt><tt class="py-string">'point'</tt><tt class="py-op">,</tt> </tt>
<a name="L215"></a><tt class="py-lineno">215</tt>  <tt class="py-line">                 <tt class="py-param">additional_mutation_rate</tt><tt class="py-op">=</tt><tt class="py-number">0.01</tt><tt class="py-op">,</tt> <tt class="py-param">gender</tt><tt class="py-op">=</tt><tt class="py-name">None</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Organism.__init__-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Organism.__init__-expanded"><a name="L216"></a><tt class="py-lineno">216</tt>  <tt class="py-line">        <tt class="py-docstring">"""</tt> </tt>
<a name="L217"></a><tt class="py-lineno">217</tt>  <tt class="py-line"><tt class="py-docstring">        Sets up a new organism with default status (age = 0, vitality = 100,</tt> </tt>
<a name="L218"></a><tt class="py-lineno">218</tt>  <tt class="py-line"><tt class="py-docstring">        lifespan = 100, fitness = 100, alive = True)</tt> </tt>
<a name="L219"></a><tt class="py-lineno">219</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L220"></a><tt class="py-lineno">220</tt>  <tt class="py-line"><tt class="py-docstring">        @param genome: list of chromosomes to inherit. Default = 'dummy'. </tt> </tt>
<a name="L221"></a><tt class="py-lineno">221</tt>  <tt class="py-line"><tt class="py-docstring">            This creates a 'dummy' chromosome which is basically a one-base </tt> </tt>
<a name="L222"></a><tt class="py-lineno">222</tt>  <tt class="py-line"><tt class="py-docstring">            chromosome - this is for applications which does not utilize </tt> </tt>
<a name="L223"></a><tt class="py-lineno">223</tt>  <tt class="py-line"><tt class="py-docstring">            the chromosome.</tt> </tt>
<a name="L224"></a><tt class="py-lineno">224</tt>  <tt class="py-line"><tt class="py-docstring">        @param mutation_type: type of mutation. Accepts 'point' (point </tt> </tt>
<a name="L225"></a><tt class="py-lineno">225</tt>  <tt class="py-line"><tt class="py-docstring">            mutation), 'insert' (insert a base), 'delete' (delete a base), </tt> </tt>
<a name="L226"></a><tt class="py-lineno">226</tt>  <tt class="py-line"><tt class="py-docstring">            'invert' (invert a stretch of the chromosome), 'duplicate' </tt> </tt>
<a name="L227"></a><tt class="py-lineno">227</tt>  <tt class="py-line"><tt class="py-docstring">            (duplicate a stretch of the chromosome), 'translocate' </tt> </tt>
<a name="L228"></a><tt class="py-lineno">228</tt>  <tt class="py-line"><tt class="py-docstring">            (translocate a stretch of chromosome to another random position). </tt> </tt>
<a name="L229"></a><tt class="py-lineno">229</tt>  <tt class="py-line"><tt class="py-docstring">            Default = point.</tt> </tt>
<a name="L230"></a><tt class="py-lineno">230</tt>  <tt class="py-line"><tt class="py-docstring">        @param additional_mutation_rate: probability of mutation per base </tt> </tt>
<a name="L231"></a><tt class="py-lineno">231</tt>  <tt class="py-line"><tt class="py-docstring">            above background mutation rate. Default = 0.01 (1%). No mutation </tt> </tt>
<a name="L232"></a><tt class="py-lineno">232</tt>  <tt class="py-line"><tt class="py-docstring">            event will ever happen if (rate + background_mutation) is less </tt> </tt>
<a name="L233"></a><tt class="py-lineno">233</tt>  <tt class="py-line"><tt class="py-docstring">            than zero.</tt> </tt>
<a name="L234"></a><tt class="py-lineno">234</tt>  <tt class="py-line"><tt class="py-docstring">        @param gender: establishes the gender of the organism which may be </tt> </tt>
<a name="L235"></a><tt class="py-lineno">235</tt>  <tt class="py-line"><tt class="py-docstring">            used for mating routines.</tt> </tt>
<a name="L236"></a><tt class="py-lineno">236</tt>  <tt class="py-line"><tt class="py-docstring">            </tt> </tt>
<a name="L237"></a><tt class="py-lineno">237</tt>  <tt class="py-line"><tt class="py-docstring">        @since: version 0.4</tt> </tt>
<a name="L238"></a><tt class="py-lineno">238</tt>  <tt class="py-line"><tt class="py-docstring">        """</tt> </tt>
<a name="L239"></a><tt class="py-lineno">239</tt>  <tt class="py-line">        <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">status</tt> <tt class="py-op">=</tt> <tt class="py-op">{</tt><tt class="py-string">'alive'</tt><tt class="py-op">:</tt> <tt class="py-name">True</tt><tt class="py-op">,</tt>            <tt class="py-comment"># is the organism alive?</tt> </tt>
<a name="L240"></a><tt class="py-lineno">240</tt>  <tt class="py-line">                       <tt class="py-string">'vitality'</tt><tt class="py-op">:</tt> <tt class="py-number">100.0</tt><tt class="py-op">,</tt>        <tt class="py-comment"># % of vitality</tt> </tt>
<a name="L241"></a><tt class="py-lineno">241</tt>  <tt class="py-line">                       <tt class="py-string">'parents'</tt><tt class="py-op">:</tt> <tt class="py-name">None</tt><tt class="py-op">,</tt>          <tt class="py-comment"># identity of parent(s)</tt> </tt>
<a name="L242"></a><tt class="py-lineno">242</tt>  <tt class="py-line">                       <tt class="py-string">'age'</tt><tt class="py-op">:</tt> <tt class="py-number">0.0</tt><tt class="py-op">,</tt>               <tt class="py-comment"># age of the organism</tt> </tt>
<a name="L243"></a><tt class="py-lineno">243</tt>  <tt class="py-line">                       <tt class="py-string">'gender'</tt><tt class="py-op">:</tt> <tt class="py-name">None</tt><tt class="py-op">,</tt>           <tt class="py-comment"># gender of organism</tt> </tt>
<a name="L244"></a><tt class="py-lineno">244</tt>  <tt class="py-line">                       <tt class="py-string">'lifespan'</tt><tt class="py-op">:</tt> <tt class="py-number">100.0</tt><tt class="py-op">,</tt>        <tt class="py-comment"># maximum lifespan</tt> </tt>
<a name="L245"></a><tt class="py-lineno">245</tt>  <tt class="py-line">                       <tt class="py-string">'fitness'</tt><tt class="py-op">:</tt> <tt class="py-number">100.0</tt><tt class="py-op">,</tt>         <tt class="py-comment"># % of fitness</tt> </tt>
<a name="L246"></a><tt class="py-lineno">246</tt>  <tt class="py-line">                       <tt class="py-string">'blood'</tt><tt class="py-op">:</tt> <tt class="py-name">None</tt><tt class="py-op">,</tt>            <tt class="py-comment"># interpreted chromosome</tt> </tt>
<a name="L247"></a><tt class="py-lineno">247</tt>  <tt class="py-line">                       <tt class="py-string">'identity'</tt><tt class="py-op">:</tt> <tt class="py-name">None</tt><tt class="py-op">,</tt>         <tt class="py-comment"># name of the organism</tt> </tt>
<a name="L248"></a><tt class="py-lineno">248</tt>  <tt class="py-line">                       <tt class="py-string">'deme'</tt><tt class="py-op">:</tt> <tt class="py-name">None</tt><tt class="py-op">,</tt>             <tt class="py-comment"># sub-population or race</tt> </tt>
<a name="L249"></a><tt class="py-lineno">249</tt>  <tt class="py-line">                       <tt class="py-string">'location'</tt><tt class="py-op">:</tt> <tt class="py-name">None</tt><tt class="py-op">,</tt>         <tt class="py-comment"># location of the organism</tt> </tt>
<a name="L250"></a><tt class="py-lineno">250</tt>  <tt class="py-line">                       <tt class="py-string">'death'</tt><tt class="py-op">:</tt> <tt class="py-name">None</tt><tt class="py-op">}</tt> </tt>
<a name="L251"></a><tt class="py-lineno">251</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">genome</tt> <tt class="py-op">==</tt> <tt class="py-string">'dummy'</tt><tt class="py-op">:</tt> </tt>
<a name="L252"></a><tt class="py-lineno">252</tt>  <tt class="py-line">            <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">genome</tt> <tt class="py-op">=</tt> <tt class="py-op">[</tt><tt id="link-14" class="py-name" targets="Class dose.genetic.Chromosome=dose.genetic.Chromosome-class.html"><a title="dose.genetic.Chromosome" class="py-name" href="#" onclick="return doclink('link-14', 'Chromosome', 'link-14');">Chromosome</a></tt><tt class="py-op">(</tt><tt class="py-op">[</tt><tt class="py-number">0</tt><tt class="py-op">]</tt><tt class="py-op">,</tt> <tt class="py-op">[</tt><tt class="py-number">0</tt><tt class="py-op">]</tt><tt class="py-op">)</tt><tt class="py-op">]</tt> </tt>
<a name="L253"></a><tt class="py-lineno">253</tt>  <tt class="py-line">        <tt class="py-keyword">else</tt><tt class="py-op">:</tt>  </tt>
<a name="L254"></a><tt class="py-lineno">254</tt>  <tt class="py-line">            <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">genome</tt> <tt class="py-op">=</tt> <tt class="py-name">genome</tt> </tt>
<a name="L255"></a><tt class="py-lineno">255</tt>  <tt class="py-line">        <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">mutation_type</tt> <tt class="py-op">=</tt> <tt class="py-name">mutation_type</tt> </tt>
<a name="L256"></a><tt class="py-lineno">256</tt>  <tt class="py-line">        <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">additional_mutation_rate</tt> <tt class="py-op">=</tt> <tt class="py-name">additional_mutation_rate</tt> </tt>
<a name="L257"></a><tt class="py-lineno">257</tt>  <tt class="py-line">        <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">status</tt><tt class="py-op">[</tt><tt class="py-string">'gender'</tt><tt class="py-op">]</tt> <tt class="py-op">=</tt> <tt class="py-name">gender</tt> </tt>
</div><a name="L258"></a><tt class="py-lineno">258</tt>  <tt class="py-line">         </tt>
<a name="Organism.generate_name"></a><div id="Organism.generate_name-def"><a name="L259"></a><tt class="py-lineno">259</tt> <a class="py-toggle" href="#" id="Organism.generate_name-toggle" onclick="return toggle('Organism.generate_name');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Organism-class.html#generate_name">generate_name</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Organism.generate_name-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Organism.generate_name-expanded"><a name="L260"></a><tt class="py-lineno">260</tt>  <tt class="py-line">        <tt id="link-15" class="py-name" targets="Variable dose.copads.samplestatistics.SingleSample.name=dose.copads.samplestatistics.SingleSample-class.html#name"><a title="dose.copads.samplestatistics.SingleSample.name" class="py-name" href="#" onclick="return doclink('link-15', 'name', 'link-15');">name</a></tt> <tt class="py-op">=</tt> <tt class="py-string">''</tt><tt class="py-op">.</tt><tt class="py-name">join</tt><tt class="py-op">(</tt><tt class="py-op">[</tt><tt id="link-16" class="py-name"><a title="dose.copads.statisticsdistribution.BetaDistribution.random
dose.copads.statisticsdistribution.BradfordDistribution.random
dose.copads.statisticsdistribution.BurrDistribution.random
dose.copads.statisticsdistribution.CauchyDistribution.random
dose.copads.statisticsdistribution.ExponentialDistribution.random
dose.copads.statisticsdistribution.GumbelDistribution.random
dose.copads.statisticsdistribution.LogNormalDistribution.random
dose.copads.statisticsdistribution.NormalDistribution.random
dose.copads.statisticsdistribution.ParetoDistribution.random
dose.copads.statisticsdistribution.UniformDistribution.random
dose.copads.statisticsdistribution.WeiBullDistribution.random" class="py-name" href="#" onclick="return doclink('link-16', 'random', 'link-0');">random</a></tt><tt class="py-op">.</tt><tt class="py-name">choice</tt><tt class="py-op">(</tt><tt class="py-op">(</tt><tt class="py-string">'1'</tt><tt class="py-op">,</tt> <tt class="py-string">'2'</tt><tt class="py-op">,</tt> <tt class="py-string">'3'</tt><tt class="py-op">,</tt> <tt class="py-string">'4'</tt><tt class="py-op">,</tt> <tt class="py-string">'5'</tt><tt class="py-op">,</tt> <tt class="py-string">'6'</tt><tt class="py-op">,</tt> <tt class="py-string">'7'</tt><tt class="py-op">,</tt> </tt>
<a name="L261"></a><tt class="py-lineno">261</tt>  <tt class="py-line">                                       <tt class="py-string">'8'</tt><tt class="py-op">,</tt> <tt class="py-string">'9'</tt><tt class="py-op">,</tt> <tt class="py-string">'A'</tt><tt class="py-op">,</tt> <tt class="py-string">'B'</tt><tt class="py-op">,</tt> <tt class="py-string">'C'</tt><tt class="py-op">,</tt> <tt class="py-string">'D'</tt><tt class="py-op">,</tt> <tt class="py-string">'E'</tt><tt class="py-op">,</tt> </tt>
<a name="L262"></a><tt class="py-lineno">262</tt>  <tt class="py-line">                                       <tt class="py-string">'F'</tt><tt class="py-op">,</tt> <tt class="py-string">'G'</tt><tt class="py-op">,</tt> <tt class="py-string">'H'</tt><tt class="py-op">,</tt> <tt class="py-string">'I'</tt><tt class="py-op">,</tt> <tt class="py-string">'J'</tt><tt class="py-op">,</tt> <tt class="py-string">'K'</tt><tt class="py-op">,</tt> <tt class="py-string">'L'</tt><tt class="py-op">,</tt> </tt>
<a name="L263"></a><tt class="py-lineno">263</tt>  <tt class="py-line">                                       <tt class="py-string">'M'</tt><tt class="py-op">,</tt> <tt class="py-string">'N'</tt><tt class="py-op">,</tt> <tt class="py-string">'O'</tt><tt class="py-op">,</tt> <tt class="py-string">'P'</tt><tt class="py-op">,</tt> <tt class="py-string">'Q'</tt><tt class="py-op">,</tt> <tt class="py-string">'R'</tt><tt class="py-op">,</tt> <tt class="py-string">'S'</tt><tt class="py-op">,</tt> </tt>
<a name="L264"></a><tt class="py-lineno">264</tt>  <tt class="py-line">                                       <tt class="py-string">'T'</tt><tt class="py-op">,</tt> <tt class="py-string">'U'</tt><tt class="py-op">,</tt> <tt class="py-string">'V'</tt><tt class="py-op">,</tt> <tt class="py-string">'W'</tt><tt class="py-op">,</tt> <tt class="py-string">'X'</tt><tt class="py-op">,</tt> <tt class="py-string">'Y'</tt><tt class="py-op">,</tt> <tt class="py-string">'Z'</tt><tt class="py-op">,</tt> </tt>
<a name="L265"></a><tt class="py-lineno">265</tt>  <tt class="py-line">                                       <tt class="py-string">'a'</tt><tt class="py-op">,</tt> <tt class="py-string">'b'</tt><tt class="py-op">,</tt> <tt class="py-string">'c'</tt><tt class="py-op">,</tt> <tt class="py-string">'d'</tt><tt class="py-op">,</tt> <tt class="py-string">'e'</tt><tt class="py-op">,</tt> <tt class="py-string">'f'</tt><tt class="py-op">,</tt> <tt class="py-string">'g'</tt><tt class="py-op">,</tt> </tt>
<a name="L266"></a><tt class="py-lineno">266</tt>  <tt class="py-line">                                       <tt class="py-string">'h'</tt><tt class="py-op">,</tt> <tt class="py-string">'i'</tt><tt class="py-op">,</tt> <tt class="py-string">'j'</tt><tt class="py-op">,</tt> <tt class="py-string">'k'</tt><tt class="py-op">,</tt> <tt class="py-string">'l'</tt><tt class="py-op">,</tt> <tt class="py-string">'m'</tt><tt class="py-op">,</tt> <tt class="py-string">'n'</tt><tt class="py-op">,</tt> </tt>
<a name="L267"></a><tt class="py-lineno">267</tt>  <tt class="py-line">                                       <tt class="py-string">'o'</tt><tt class="py-op">,</tt> <tt class="py-string">'p'</tt><tt class="py-op">,</tt> <tt class="py-string">'q'</tt><tt class="py-op">,</tt> <tt class="py-string">'r'</tt><tt class="py-op">,</tt> <tt class="py-string">'s'</tt><tt class="py-op">,</tt> <tt class="py-string">'t'</tt><tt class="py-op">,</tt> <tt class="py-string">'u'</tt><tt class="py-op">,</tt> </tt>
<a name="L268"></a><tt class="py-lineno">268</tt>  <tt class="py-line">                                       <tt class="py-string">'v'</tt><tt class="py-op">,</tt> <tt class="py-string">'w'</tt><tt class="py-op">,</tt> <tt class="py-string">'x'</tt><tt class="py-op">,</tt> <tt class="py-string">'y'</tt><tt class="py-op">,</tt> <tt class="py-string">'z'</tt><tt class="py-op">,</tt> <tt class="py-string">'='</tt><tt class="py-op">,</tt> <tt class="py-string">'#'</tt><tt class="py-op">,</tt> </tt>
<a name="L269"></a><tt class="py-lineno">269</tt>  <tt class="py-line">                                       <tt class="py-string">'$'</tt><tt class="py-op">,</tt> <tt class="py-string">'%'</tt><tt class="py-op">,</tt> <tt class="py-string">'&amp;'</tt><tt class="py-op">,</tt> <tt class="py-string">'@'</tt><tt class="py-op">,</tt> <tt class="py-string">'&lt;'</tt><tt class="py-op">,</tt> <tt class="py-string">'&gt;'</tt><tt class="py-op">,</tt> <tt class="py-string">'?'</tt> </tt>
<a name="L270"></a><tt class="py-lineno">270</tt>  <tt class="py-line">                                       <tt class="py-op">)</tt><tt class="py-op">)</tt> <tt class="py-keyword">for</tt> <tt class="py-name">x</tt> <tt class="py-keyword">in</tt> <tt id="link-17" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.range" class="py-name" href="#" onclick="return doclink('link-17', 'range', 'link-4');">range</a></tt><tt class="py-op">(</tt><tt class="py-number">32</tt><tt class="py-op">)</tt><tt class="py-op">]</tt><tt class="py-op">)</tt> </tt>
<a name="L271"></a><tt class="py-lineno">271</tt>  <tt class="py-line"> </tt>
<a name="L272"></a><tt class="py-lineno">272</tt>  <tt class="py-line">        <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">status</tt><tt class="py-op">[</tt><tt class="py-string">'identity'</tt><tt class="py-op">]</tt> <tt class="py-op">=</tt> <tt id="link-18" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.name" class="py-name" href="#" onclick="return doclink('link-18', 'name', 'link-15');">name</a></tt> </tt>
</div><a name="L273"></a><tt class="py-lineno">273</tt>  <tt class="py-line">         </tt>
<a name="Organism.fitness"></a><div id="Organism.fitness-def"><a name="L274"></a><tt class="py-lineno">274</tt> <a class="py-toggle" href="#" id="Organism.fitness-toggle" onclick="return toggle('Organism.fitness');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Organism-class.html#fitness">fitness</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Organism.fitness-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Organism.fitness-expanded"><a name="L275"></a><tt class="py-lineno">275</tt>  <tt class="py-line">        <tt class="py-docstring">"""</tt> </tt>
<a name="L276"></a><tt class="py-lineno">276</tt>  <tt class="py-line"><tt class="py-docstring">        Function to calculate the fitness of the current organism. </tt> </tt>
<a name="L277"></a><tt class="py-lineno">277</tt>  <tt class="py-line"><tt class="py-docstring">        B{This function MUST be over-ridden by the inherited class or </tt> </tt>
<a name="L278"></a><tt class="py-lineno">278</tt>  <tt class="py-line"><tt class="py-docstring">        substituted as fitness function is highly dependent on utility.} </tt> </tt>
<a name="L279"></a><tt class="py-lineno">279</tt>  <tt class="py-line"><tt class="py-docstring">        The only requirement is that a fitness score must be returned.</tt> </tt>
<a name="L280"></a><tt class="py-lineno">280</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L281"></a><tt class="py-lineno">281</tt>  <tt class="py-line"><tt class="py-docstring">        Here, the sample implementation calculates fitness as proportion of </tt> </tt>
<a name="L282"></a><tt class="py-lineno">282</tt>  <tt class="py-line"><tt class="py-docstring">        the genome with '1's.</tt> </tt>
<a name="L283"></a><tt class="py-lineno">283</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L284"></a><tt class="py-lineno">284</tt>  <tt class="py-line"><tt class="py-docstring">        @return: fitness score or fitness list</tt> </tt>
<a name="L285"></a><tt class="py-lineno">285</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L286"></a><tt class="py-lineno">286</tt>  <tt class="py-line"><tt class="py-docstring">        @since: version 0.4</tt> </tt>
<a name="L287"></a><tt class="py-lineno">287</tt>  <tt class="py-line"><tt class="py-docstring">        """</tt> </tt>
<a name="L288"></a><tt class="py-lineno">288</tt>  <tt class="py-line">        <tt class="py-name">one_count</tt> <tt class="py-op">=</tt> <tt class="py-name">sum</tt><tt class="py-op">(</tt><tt class="py-op">[</tt><tt class="py-name">sum</tt><tt class="py-op">(</tt><tt class="py-name">chromosome</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">)</tt>  </tt>
<a name="L289"></a><tt class="py-lineno">289</tt>  <tt class="py-line">                         <tt class="py-keyword">for</tt> <tt class="py-name">chromosome</tt> <tt class="py-keyword">in</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">genome</tt><tt class="py-op">]</tt><tt class="py-op">)</tt> </tt>
<a name="L290"></a><tt class="py-lineno">290</tt>  <tt class="py-line">        <tt class="py-name">length_of_genome</tt> <tt class="py-op">=</tt> <tt class="py-name">sum</tt><tt class="py-op">(</tt><tt class="py-op">[</tt><tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">chromosome</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">)</tt>  </tt>
<a name="L291"></a><tt class="py-lineno">291</tt>  <tt class="py-line">                                <tt class="py-keyword">for</tt> <tt class="py-name">chromosome</tt> <tt class="py-keyword">in</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">genome</tt><tt class="py-op">]</tt><tt class="py-op">)</tt> </tt>
<a name="L292"></a><tt class="py-lineno">292</tt>  <tt class="py-line">        <tt class="py-keyword">return</tt> <tt class="py-name">float</tt><tt class="py-op">(</tt><tt class="py-name">one_count</tt><tt class="py-op">)</tt> <tt class="py-op">/</tt> <tt class="py-name">float</tt><tt class="py-op">(</tt><tt class="py-name">length_of_genome</tt><tt class="py-op">)</tt> </tt>
</div><a name="L293"></a><tt class="py-lineno">293</tt>  <tt class="py-line">     </tt>
<a name="Organism.mutation_scheme"></a><div id="Organism.mutation_scheme-def"><a name="L294"></a><tt class="py-lineno">294</tt> <a class="py-toggle" href="#" id="Organism.mutation_scheme-toggle" onclick="return toggle('Organism.mutation_scheme');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Organism-class.html#mutation_scheme">mutation_scheme</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">,</tt> <tt class="py-param">type</tt><tt class="py-op">=</tt><tt class="py-name">None</tt><tt class="py-op">,</tt> <tt class="py-param">rate</tt><tt class="py-op">=</tt><tt class="py-name">None</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Organism.mutation_scheme-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Organism.mutation_scheme-expanded"><a name="L295"></a><tt class="py-lineno">295</tt>  <tt class="py-line">        <tt class="py-docstring">"""</tt> </tt>
<a name="L296"></a><tt class="py-lineno">296</tt>  <tt class="py-line"><tt class="py-docstring">        Function to trigger mutation events in each chromosome. B{This </tt> </tt>
<a name="L297"></a><tt class="py-lineno">297</tt>  <tt class="py-line"><tt class="py-docstring">        function may be over-ridden by the inherited class or substituted </tt> </tt>
<a name="L298"></a><tt class="py-lineno">298</tt>  <tt class="py-line"><tt class="py-docstring">        to cater for specific mutation schemes but not an absolute </tt> </tt>
<a name="L299"></a><tt class="py-lineno">299</tt>  <tt class="py-line"><tt class="py-docstring">        requirement to do so.}</tt> </tt>
<a name="L300"></a><tt class="py-lineno">300</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L301"></a><tt class="py-lineno">301</tt>  <tt class="py-line"><tt class="py-docstring">        Both type and rate must be defined at the same time, otherwise the </tt> </tt>
<a name="L302"></a><tt class="py-lineno">302</tt>  <tt class="py-line"><tt class="py-docstring">        initiated mutation_type and additional_mutation_rate will be used.</tt> </tt>
<a name="L303"></a><tt class="py-lineno">303</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L304"></a><tt class="py-lineno">304</tt>  <tt class="py-line"><tt class="py-docstring">        @param type: type of mutation. Accepts 'point' (point mutation), </tt> </tt>
<a name="L305"></a><tt class="py-lineno">305</tt>  <tt class="py-line"><tt class="py-docstring">            'insert' (insert a base), 'delete' (delete a base), 'invert' </tt> </tt>
<a name="L306"></a><tt class="py-lineno">306</tt>  <tt class="py-line"><tt class="py-docstring">            (invert a stretch of the chromosome), 'duplicate' (duplicate a </tt> </tt>
<a name="L307"></a><tt class="py-lineno">307</tt>  <tt class="py-line"><tt class="py-docstring">            stretch of the chromosome), 'translocate' (translocate a stretch </tt> </tt>
<a name="L308"></a><tt class="py-lineno">308</tt>  <tt class="py-line"><tt class="py-docstring">            of chromosome to another random position). Default = None.</tt> </tt>
<a name="L309"></a><tt class="py-lineno">309</tt>  <tt class="py-line"><tt class="py-docstring">        @param rate: probability of mutation per base above background mutation </tt> </tt>
<a name="L310"></a><tt class="py-lineno">310</tt>  <tt class="py-line"><tt class="py-docstring">            rate. Default = None. No mutation event will ever happen if </tt> </tt>
<a name="L311"></a><tt class="py-lineno">311</tt>  <tt class="py-line"><tt class="py-docstring">            (rate + background_mutation) is less than zero.</tt> </tt>
<a name="L312"></a><tt class="py-lineno">312</tt>  <tt class="py-line"><tt class="py-docstring">            </tt> </tt>
<a name="L313"></a><tt class="py-lineno">313</tt>  <tt class="py-line"><tt class="py-docstring">        @since: version 0.4</tt> </tt>
<a name="L314"></a><tt class="py-lineno">314</tt>  <tt class="py-line"><tt class="py-docstring">        """</tt> </tt>
<a name="L315"></a><tt class="py-lineno">315</tt>  <tt class="py-line">        <tt class="py-keyword">for</tt> <tt class="py-name">chromosome</tt> <tt class="py-keyword">in</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">genome</tt><tt class="py-op">:</tt>  </tt>
<a name="L316"></a><tt class="py-lineno">316</tt>  <tt class="py-line">            <tt class="py-keyword">if</tt> <tt class="py-keyword">not</tt> <tt class="py-name">type</tt> <tt class="py-keyword">and</tt> <tt class="py-keyword">not</tt> <tt class="py-name">rate</tt><tt class="py-op">:</tt> </tt>
<a name="L317"></a><tt class="py-lineno">317</tt>  <tt class="py-line">                <tt class="py-name">chromosome</tt><tt class="py-op">.</tt><tt id="link-19" class="py-name" targets="Method dose.genetic.Chromosome.rmutate()=dose.genetic.Chromosome-class.html#rmutate"><a title="dose.genetic.Chromosome.rmutate" class="py-name" href="#" onclick="return doclink('link-19', 'rmutate', 'link-19');">rmutate</a></tt><tt class="py-op">(</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">mutation_type</tt><tt class="py-op">,</tt>  </tt>
<a name="L318"></a><tt class="py-lineno">318</tt>  <tt class="py-line">                                   <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">additional_mutation_rate</tt><tt class="py-op">)</tt> </tt>
<a name="L319"></a><tt class="py-lineno">319</tt>  <tt class="py-line">            <tt class="py-keyword">if</tt> <tt class="py-name">type</tt> <tt class="py-keyword">and</tt> <tt class="py-name">rate</tt><tt class="py-op">:</tt> </tt>
<a name="L320"></a><tt class="py-lineno">320</tt>  <tt class="py-line">                <tt class="py-name">chromosome</tt><tt class="py-op">.</tt><tt id="link-20" class="py-name"><a title="dose.genetic.Chromosome.rmutate" class="py-name" href="#" onclick="return doclink('link-20', 'rmutate', 'link-19');">rmutate</a></tt><tt class="py-op">(</tt><tt class="py-name">type</tt><tt class="py-op">,</tt> <tt class="py-name">rate</tt><tt class="py-op">)</tt> </tt>
</div><a name="L321"></a><tt class="py-lineno">321</tt>  <tt class="py-line">         </tt>
<a name="Organism.setStatus"></a><div id="Organism.setStatus-def"><a name="L322"></a><tt class="py-lineno">322</tt> <a class="py-toggle" href="#" id="Organism.setStatus-toggle" onclick="return toggle('Organism.setStatus');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Organism-class.html#setStatus">setStatus</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">,</tt> <tt class="py-param">variable</tt><tt class="py-op">,</tt> <tt class="py-param">value</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Organism.setStatus-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Organism.setStatus-expanded"><a name="L323"></a><tt class="py-lineno">323</tt>  <tt class="py-line">        <tt class="py-docstring">"""</tt> </tt>
<a name="L324"></a><tt class="py-lineno">324</tt>  <tt class="py-line"><tt class="py-docstring">        Sets new status or change status of the organism. However, the </tt> </tt>
<a name="L325"></a><tt class="py-lineno">325</tt>  <tt class="py-line"><tt class="py-docstring">        following status change will result in death of the organism</tt> </tt>
<a name="L326"></a><tt class="py-lineno">326</tt>  <tt class="py-line"><tt class="py-docstring">            1. 'alive' to False</tt> </tt>
<a name="L327"></a><tt class="py-lineno">327</tt>  <tt class="py-line"><tt class="py-docstring">            2. 'vitality' to or below zero</tt> </tt>
<a name="L328"></a><tt class="py-lineno">328</tt>  <tt class="py-line"><tt class="py-docstring">            3. 'age' to or above lifespan</tt> </tt>
<a name="L329"></a><tt class="py-lineno">329</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L330"></a><tt class="py-lineno">330</tt>  <tt class="py-line"><tt class="py-docstring">        @param variable: name of status to change</tt> </tt>
<a name="L331"></a><tt class="py-lineno">331</tt>  <tt class="py-line"><tt class="py-docstring">        @param value: new value of the status</tt> </tt>
<a name="L332"></a><tt class="py-lineno">332</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L333"></a><tt class="py-lineno">333</tt>  <tt class="py-line"><tt class="py-docstring">        @since: version 0.4</tt> </tt>
<a name="L334"></a><tt class="py-lineno">334</tt>  <tt class="py-line"><tt class="py-docstring">        """</tt> </tt>
<a name="L335"></a><tt class="py-lineno">335</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">variable</tt> <tt class="py-op">==</tt> <tt class="py-string">'alive'</tt> <tt class="py-keyword">and</tt> <tt class="py-name">value</tt> <tt class="py-op">==</tt> <tt class="py-name">True</tt><tt class="py-op">:</tt>  </tt>
<a name="L336"></a><tt class="py-lineno">336</tt>  <tt class="py-line">            <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">status</tt><tt class="py-op">[</tt><tt class="py-string">'alive'</tt><tt class="py-op">]</tt> <tt class="py-op">=</tt> <tt class="py-name">value</tt> </tt>
<a name="L337"></a><tt class="py-lineno">337</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">variable</tt> <tt class="py-op">==</tt> <tt class="py-string">'alive'</tt> <tt class="py-keyword">and</tt> <tt class="py-name">value</tt> <tt class="py-op">==</tt> <tt class="py-name">False</tt><tt class="py-op">:</tt>  </tt>
<a name="L338"></a><tt class="py-lineno">338</tt>  <tt class="py-line">            <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">status</tt><tt class="py-op">[</tt><tt class="py-string">'alive'</tt><tt class="py-op">]</tt> <tt class="py-op">=</tt> <tt class="py-name">value</tt> </tt>
<a name="L339"></a><tt class="py-lineno">339</tt>  <tt class="py-line">            <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">status</tt><tt class="py-op">[</tt><tt class="py-string">'death'</tt><tt class="py-op">]</tt> <tt class="py-op">=</tt> <tt class="py-string">'death03'</tt> </tt>
<a name="L340"></a><tt class="py-lineno">340</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">variable</tt> <tt class="py-op">==</tt> <tt class="py-string">'vitality'</tt><tt class="py-op">:</tt> </tt>
<a name="L341"></a><tt class="py-lineno">341</tt>  <tt class="py-line">            <tt class="py-keyword">if</tt> <tt class="py-name">float</tt><tt class="py-op">(</tt><tt class="py-name">value</tt><tt class="py-op">)</tt> <tt class="py-op">&gt;</tt> <tt class="py-number">100.1</tt><tt class="py-op">:</tt> </tt>
<a name="L342"></a><tt class="py-lineno">342</tt>  <tt class="py-line">                <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">status</tt><tt class="py-op">[</tt><tt class="py-string">'vitality'</tt><tt class="py-op">]</tt> <tt class="py-op">=</tt> <tt class="py-number">100.0</tt> </tt>
<a name="L343"></a><tt class="py-lineno">343</tt>  <tt class="py-line">            <tt class="py-keyword">if</tt> <tt class="py-name">float</tt><tt class="py-op">(</tt><tt class="py-name">value</tt><tt class="py-op">)</tt> <tt class="py-op">&gt;</tt> <tt class="py-number">0.0</tt> <tt class="py-keyword">and</tt> <tt class="py-name">float</tt><tt class="py-op">(</tt><tt class="py-name">value</tt><tt class="py-op">)</tt> <tt class="py-op">&lt;</tt> <tt class="py-number">100.1</tt><tt class="py-op">:</tt> </tt>
<a name="L344"></a><tt class="py-lineno">344</tt>  <tt class="py-line">                <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">status</tt><tt class="py-op">[</tt><tt class="py-string">'vitality'</tt><tt class="py-op">]</tt> <tt class="py-op">=</tt> <tt class="py-name">float</tt><tt class="py-op">(</tt><tt class="py-name">value</tt><tt class="py-op">)</tt> </tt>
<a name="L345"></a><tt class="py-lineno">345</tt>  <tt class="py-line">            <tt class="py-keyword">if</tt> <tt class="py-name">float</tt><tt class="py-op">(</tt><tt class="py-name">value</tt><tt class="py-op">)</tt> <tt class="py-op">==</tt> <tt class="py-number">0</tt> <tt class="py-keyword">or</tt> <tt class="py-name">float</tt><tt class="py-op">(</tt><tt class="py-name">value</tt><tt class="py-op">)</tt> <tt class="py-op">&lt;</tt> <tt class="py-number">0</tt><tt class="py-op">:</tt> </tt>
<a name="L346"></a><tt class="py-lineno">346</tt>  <tt class="py-line">                <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">status</tt><tt class="py-op">[</tt><tt class="py-string">'vitality'</tt><tt class="py-op">]</tt> <tt class="py-op">=</tt> <tt class="py-number">0</tt> </tt>
<a name="L347"></a><tt class="py-lineno">347</tt>  <tt class="py-line">                <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">status</tt><tt class="py-op">[</tt><tt class="py-string">'alive'</tt><tt class="py-op">]</tt> <tt class="py-op">=</tt> <tt class="py-name">False</tt> </tt>
<a name="L348"></a><tt class="py-lineno">348</tt>  <tt class="py-line">                <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">status</tt><tt class="py-op">[</tt><tt class="py-string">'death'</tt><tt class="py-op">]</tt> <tt class="py-op">=</tt> <tt class="py-string">'death01'</tt> </tt>
<a name="L349"></a><tt class="py-lineno">349</tt>  <tt class="py-line">        <tt class="py-keyword">elif</tt> <tt class="py-name">variable</tt> <tt class="py-op">==</tt> <tt class="py-string">'age'</tt><tt class="py-op">:</tt> </tt>
<a name="L350"></a><tt class="py-lineno">350</tt>  <tt class="py-line">            <tt class="py-keyword">if</tt> <tt class="py-name">float</tt><tt class="py-op">(</tt><tt class="py-name">value</tt><tt class="py-op">)</tt> <tt class="py-op">&lt;</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">status</tt><tt class="py-op">[</tt><tt class="py-string">'lifespan'</tt><tt class="py-op">]</tt><tt class="py-op">:</tt> </tt>
<a name="L351"></a><tt class="py-lineno">351</tt>  <tt class="py-line">                <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">status</tt><tt class="py-op">[</tt><tt class="py-string">'age'</tt><tt class="py-op">]</tt> <tt class="py-op">=</tt> <tt class="py-name">float</tt><tt class="py-op">(</tt><tt class="py-name">value</tt><tt class="py-op">)</tt> </tt>
<a name="L352"></a><tt class="py-lineno">352</tt>  <tt class="py-line">            <tt class="py-keyword">else</tt><tt class="py-op">:</tt> </tt>
<a name="L353"></a><tt class="py-lineno">353</tt>  <tt class="py-line">                <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">status</tt><tt class="py-op">[</tt><tt class="py-string">'age'</tt><tt class="py-op">]</tt> <tt class="py-op">=</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">status</tt><tt class="py-op">[</tt><tt class="py-string">'lifespan'</tt><tt class="py-op">]</tt> </tt>
<a name="L354"></a><tt class="py-lineno">354</tt>  <tt class="py-line">                <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">status</tt><tt class="py-op">[</tt><tt class="py-string">'alive'</tt><tt class="py-op">]</tt> <tt class="py-op">=</tt> <tt class="py-name">False</tt> </tt>
<a name="L355"></a><tt class="py-lineno">355</tt>  <tt class="py-line">                <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">status</tt><tt class="py-op">[</tt><tt class="py-string">'death'</tt><tt class="py-op">]</tt> <tt class="py-op">=</tt> <tt class="py-string">'death02'</tt> </tt>
<a name="L356"></a><tt class="py-lineno">356</tt>  <tt class="py-line">        <tt class="py-keyword">else</tt><tt class="py-op">:</tt> </tt>
<a name="L357"></a><tt class="py-lineno">357</tt>  <tt class="py-line">            <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">status</tt><tt class="py-op">[</tt><tt class="py-name">variable</tt><tt class="py-op">]</tt> <tt class="py-op">=</tt> <tt class="py-name">value</tt> </tt>
</div><a name="L358"></a><tt class="py-lineno">358</tt>  <tt class="py-line">         </tt>
<a name="Organism.getStatus"></a><div id="Organism.getStatus-def"><a name="L359"></a><tt class="py-lineno">359</tt> <a class="py-toggle" href="#" id="Organism.getStatus-toggle" onclick="return toggle('Organism.getStatus');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Organism-class.html#getStatus">getStatus</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">,</tt> <tt class="py-param">variable</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Organism.getStatus-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Organism.getStatus-expanded"><a name="L360"></a><tt class="py-lineno">360</tt>  <tt class="py-line">        <tt class="py-docstring">"""</tt> </tt>
<a name="L361"></a><tt class="py-lineno">361</tt>  <tt class="py-line"><tt class="py-docstring">        Returns a status variable of the current organism</tt> </tt>
<a name="L362"></a><tt class="py-lineno">362</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L363"></a><tt class="py-lineno">363</tt>  <tt class="py-line"><tt class="py-docstring">        @param variable: name of the status variable</tt> </tt>
<a name="L364"></a><tt class="py-lineno">364</tt>  <tt class="py-line"><tt class="py-docstring">        @return: status or a KeyError if status is not found</tt> </tt>
<a name="L365"></a><tt class="py-lineno">365</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L366"></a><tt class="py-lineno">366</tt>  <tt class="py-line"><tt class="py-docstring">        @since: version 0.4</tt> </tt>
<a name="L367"></a><tt class="py-lineno">367</tt>  <tt class="py-line"><tt class="py-docstring">        """</tt> </tt>
<a name="L368"></a><tt class="py-lineno">368</tt>  <tt class="py-line">        <tt class="py-keyword">try</tt><tt class="py-op">:</tt> </tt>
<a name="L369"></a><tt class="py-lineno">369</tt>  <tt class="py-line">            <tt class="py-keyword">return</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">status</tt><tt class="py-op">[</tt><tt class="py-name">variable</tt><tt class="py-op">]</tt> </tt>
<a name="L370"></a><tt class="py-lineno">370</tt>  <tt class="py-line">        <tt class="py-keyword">except</tt><tt class="py-op">:</tt> </tt>
<a name="L371"></a><tt class="py-lineno">371</tt>  <tt class="py-line">            <tt class="py-keyword">raise</tt> <tt class="py-name">KeyError</tt><tt class="py-op">(</tt><tt class="py-string">'%s not found in organism status'</tt> <tt class="py-op">%</tt> <tt class="py-name">str</tt><tt class="py-op">(</tt><tt class="py-name">variable</tt><tt class="py-op">)</tt><tt class="py-op">)</tt> </tt>
</div><a name="L372"></a><tt class="py-lineno">372</tt>  <tt class="py-line">     </tt>
<a name="Organism.__str__"></a><div id="Organism.__str__-def"><a name="L373"></a><tt class="py-lineno">373</tt> <a class="py-toggle" href="#" id="Organism.__str__-toggle" onclick="return toggle('Organism.__str__');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Organism-class.html#__str__">__str__</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Organism.__str__-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Organism.__str__-expanded"><a name="L374"></a><tt class="py-lineno">374</tt>  <tt class="py-line">        <tt class="py-docstring">"""Returns the genome of the organism"""</tt> </tt>
<a name="L375"></a><tt class="py-lineno">375</tt>  <tt class="py-line">        <tt class="py-keyword">return</tt> <tt class="py-name">str</tt><tt class="py-op">(</tt><tt class="py-op">[</tt><tt class="py-name">chromosome</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt> <tt class="py-keyword">for</tt> <tt class="py-name">chromosome</tt> <tt class="py-keyword">in</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">genome</tt><tt class="py-op">]</tt><tt class="py-op">)</tt> </tt>
</div><a name="L376"></a><tt class="py-lineno">376</tt>  <tt class="py-line">         </tt>
<a name="Organism.clone"></a><div id="Organism.clone-def"><a name="L377"></a><tt class="py-lineno">377</tt> <a class="py-toggle" href="#" id="Organism.clone-toggle" onclick="return toggle('Organism.clone');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Organism-class.html#clone">clone</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Organism.clone-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Organism.clone-expanded"><a name="L378"></a><tt class="py-lineno">378</tt>  <tt class="py-line">        <tt class="py-docstring">"""</tt> </tt>
<a name="L379"></a><tt class="py-lineno">379</tt>  <tt class="py-line"><tt class="py-docstring">        Cloness (deep copy) the organism.</tt> </tt>
<a name="L380"></a><tt class="py-lineno">380</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L381"></a><tt class="py-lineno">381</tt>  <tt class="py-line"><tt class="py-docstring">        @return: a copy of current organism.</tt> </tt>
<a name="L382"></a><tt class="py-lineno">382</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L383"></a><tt class="py-lineno">383</tt>  <tt class="py-line"><tt class="py-docstring">        @since: version 0.4</tt> </tt>
<a name="L384"></a><tt class="py-lineno">384</tt>  <tt class="py-line"><tt class="py-docstring">        """</tt> </tt>
<a name="L385"></a><tt class="py-lineno">385</tt>  <tt class="py-line">        <tt class="py-name">org</tt> <tt class="py-op">=</tt> <tt class="py-name">deepcopy</tt><tt class="py-op">(</tt><tt class="py-name">self</tt><tt class="py-op">)</tt> </tt>
<a name="L386"></a><tt class="py-lineno">386</tt>  <tt class="py-line">        <tt class="py-keyword">return</tt> <tt class="py-name">org</tt> </tt>
</div></div><a name="L387"></a><tt class="py-lineno">387</tt>  <tt class="py-line">         </tt>
<a name="Population"></a><div id="Population-def"><a name="L388"></a><tt class="py-lineno">388</tt> <a class="py-toggle" href="#" id="Population-toggle" onclick="return toggle('Population');">-</a><tt class="py-line"><tt class="py-keyword">class</tt> <a class="py-def-name" href="dose.genetic.Population-class.html">Population</a><tt class="py-op">(</tt><tt class="py-base-class">object</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Population-collapsed" style="display:none;" pad="+++" indent="++++"></div><div id="Population-expanded"><a name="L389"></a><tt class="py-lineno">389</tt>  <tt class="py-line">    <tt class="py-docstring">"""</tt> </tt>
<a name="L390"></a><tt class="py-lineno">390</tt>  <tt class="py-line"><tt class="py-docstring">    Representation of a population as a list of organisms. The entire </tt> </tt>
<a name="L391"></a><tt class="py-lineno">391</tt>  <tt class="py-line"><tt class="py-docstring">    population is in memory.</tt> </tt>
<a name="L392"></a><tt class="py-lineno">392</tt>  <tt class="py-line"><tt class="py-docstring">    </tt> </tt>
<a name="L393"></a><tt class="py-lineno">393</tt>  <tt class="py-line"><tt class="py-docstring">    Methods to be over-ridden in the inherited class or substituted are</tt> </tt>
<a name="L394"></a><tt class="py-lineno">394</tt>  <tt class="py-line"><tt class="py-docstring">        - prepopulation_control</tt> </tt>
<a name="L395"></a><tt class="py-lineno">395</tt>  <tt class="py-line"><tt class="py-docstring">        - mating</tt> </tt>
<a name="L396"></a><tt class="py-lineno">396</tt>  <tt class="py-line"><tt class="py-docstring">        - postpopulation_control</tt> </tt>
<a name="L397"></a><tt class="py-lineno">397</tt>  <tt class="py-line"><tt class="py-docstring">        - generation_events</tt> </tt>
<a name="L398"></a><tt class="py-lineno">398</tt>  <tt class="py-line"><tt class="py-docstring">        - report</tt> </tt>
<a name="L399"></a><tt class="py-lineno">399</tt>  <tt class="py-line"><tt class="py-docstring"></tt> </tt>
<a name="L400"></a><tt class="py-lineno">400</tt>  <tt class="py-line"><tt class="py-docstring">    @see: Lim, JZR, Aw, ZQ, Goh, DJW, How, JA, Low, SXZ, Loo, BZL,</tt> </tt>
<a name="L401"></a><tt class="py-lineno">401</tt>  <tt class="py-line"><tt class="py-docstring">    Ling, MHT. 2010. A genetic algorithm framework grounded in biology.</tt> </tt>
<a name="L402"></a><tt class="py-lineno">402</tt>  <tt class="py-line"><tt class="py-docstring">    The Python Papers Source Codes 2: 6.</tt> </tt>
<a name="L403"></a><tt class="py-lineno">403</tt>  <tt class="py-line"><tt class="py-docstring">    </tt> </tt>
<a name="L404"></a><tt class="py-lineno">404</tt>  <tt class="py-line"><tt class="py-docstring">    @since: version 0.4</tt> </tt>
<a name="L405"></a><tt class="py-lineno">405</tt>  <tt class="py-line"><tt class="py-docstring">    """</tt> </tt>
<a name="L406"></a><tt class="py-lineno">406</tt>  <tt class="py-line">     </tt>
<a name="Population.__init__"></a><div id="Population.__init__-def"><a name="L407"></a><tt class="py-lineno">407</tt> <a class="py-toggle" href="#" id="Population.__init__-toggle" onclick="return toggle('Population.__init__');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Population-class.html#__init__">__init__</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">,</tt> <tt class="py-param">goal</tt><tt class="py-op">,</tt> <tt class="py-param">maxgenerations</tt><tt class="py-op">=</tt><tt class="py-string">'infinite'</tt><tt class="py-op">,</tt> <tt class="py-param">agents</tt><tt class="py-op">=</tt><tt class="py-op">[</tt><tt class="py-op">]</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Population.__init__-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Population.__init__-expanded"><a name="L408"></a><tt class="py-lineno">408</tt>  <tt class="py-line">        <tt class="py-docstring">"""</tt> </tt>
<a name="L409"></a><tt class="py-lineno">409</tt>  <tt class="py-line"><tt class="py-docstring">        Establishes a population of organisms.</tt> </tt>
<a name="L410"></a><tt class="py-lineno">410</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L411"></a><tt class="py-lineno">411</tt>  <tt class="py-line"><tt class="py-docstring">        @param goal: the goal to be reached by the population.</tt> </tt>
<a name="L412"></a><tt class="py-lineno">412</tt>  <tt class="py-line"><tt class="py-docstring">        @type goal: the return type of Organism.fitness()</tt> </tt>
<a name="L413"></a><tt class="py-lineno">413</tt>  <tt class="py-line"><tt class="py-docstring">        @param maxgenerations: maximum number of generations to evolve.</tt> </tt>
<a name="L414"></a><tt class="py-lineno">414</tt>  <tt class="py-line"><tt class="py-docstring">            Default = 'infinite'.</tt> </tt>
<a name="L415"></a><tt class="py-lineno">415</tt>  <tt class="py-line"><tt class="py-docstring">        @param agents: organisms making up the initial population.</tt> </tt>
<a name="L416"></a><tt class="py-lineno">416</tt>  <tt class="py-line"><tt class="py-docstring">        @type agents: list of Organism objects</tt> </tt>
<a name="L417"></a><tt class="py-lineno">417</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L418"></a><tt class="py-lineno">418</tt>  <tt class="py-line"><tt class="py-docstring">        @since: version 0.4</tt> </tt>
<a name="L419"></a><tt class="py-lineno">419</tt>  <tt class="py-line"><tt class="py-docstring">        """</tt> </tt>
<a name="L420"></a><tt class="py-lineno">420</tt>  <tt class="py-line">        <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt> <tt class="py-op">=</tt> <tt class="py-name">agents</tt> </tt>
<a name="L421"></a><tt class="py-lineno">421</tt>  <tt class="py-line">        <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">goal</tt> <tt class="py-op">=</tt> <tt class="py-name">goal</tt> </tt>
<a name="L422"></a><tt class="py-lineno">422</tt>  <tt class="py-line">        <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">maxgenerations</tt> <tt class="py-op">=</tt> <tt class="py-name">maxgenerations</tt> </tt>
<a name="L423"></a><tt class="py-lineno">423</tt>  <tt class="py-line">        <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">generation</tt> <tt class="py-op">=</tt> <tt class="py-number">0</tt> </tt>
</div><a name="L424"></a><tt class="py-lineno">424</tt>  <tt class="py-line">     </tt>
<a name="Population.prepopulation_control"></a><div id="Population.prepopulation_control-def"><a name="L425"></a><tt class="py-lineno">425</tt> <a class="py-toggle" href="#" id="Population.prepopulation_control-toggle" onclick="return toggle('Population.prepopulation_control');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Population-class.html#prepopulation_control">prepopulation_control</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Population.prepopulation_control-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Population.prepopulation_control-expanded"><a name="L426"></a><tt class="py-lineno">426</tt>  <tt class="py-line">        <tt class="py-docstring">"""</tt> </tt>
<a name="L427"></a><tt class="py-lineno">427</tt>  <tt class="py-line"><tt class="py-docstring">        Function to trigger population control events before mating event in </tt> </tt>
<a name="L428"></a><tt class="py-lineno">428</tt>  <tt class="py-line"><tt class="py-docstring">        each generation (For example, to simulate pre-puberty death). B{This </tt> </tt>
<a name="L429"></a><tt class="py-lineno">429</tt>  <tt class="py-line"><tt class="py-docstring">        function may be over-ridden by the inherited class or substituted to </tt> </tt>
<a name="L430"></a><tt class="py-lineno">430</tt>  <tt class="py-line"><tt class="py-docstring">        cater for specific events.} Although this is not an absolute </tt> </tt>
<a name="L431"></a><tt class="py-lineno">431</tt>  <tt class="py-line"><tt class="py-docstring">        requirement, it is extremely encouraged to prevent exhaustion of </tt> </tt>
<a name="L432"></a><tt class="py-lineno">432</tt>  <tt class="py-line"><tt class="py-docstring">        memory space. Without population control, it will seems like a </tt> </tt>
<a name="L433"></a><tt class="py-lineno">433</tt>  <tt class="py-line"><tt class="py-docstring">        reproducing immortal population.</tt> </tt>
<a name="L434"></a><tt class="py-lineno">434</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L435"></a><tt class="py-lineno">435</tt>  <tt class="py-line"><tt class="py-docstring">        Here, the sample implementation eliminates the bottom half of the </tt> </tt>
<a name="L436"></a><tt class="py-lineno">436</tt>  <tt class="py-line"><tt class="py-docstring">        population based on fitness score unless the number of organisms is</tt> </tt>
<a name="L437"></a><tt class="py-lineno">437</tt>  <tt class="py-line"><tt class="py-docstring">        less than 20. This is to prevent extinction. However, if the number </tt> </tt>
<a name="L438"></a><tt class="py-lineno">438</tt>  <tt class="py-line"><tt class="py-docstring">        of organisms is more than 2000 (more than 100x initial population </tt> </tt>
<a name="L439"></a><tt class="py-lineno">439</tt>  <tt class="py-line"><tt class="py-docstring">        size, a random selection of 2000 will be used for the next generation.</tt> </tt>
<a name="L440"></a><tt class="py-lineno">440</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L441"></a><tt class="py-lineno">441</tt>  <tt class="py-line"><tt class="py-docstring">        @since: version 0.4</tt> </tt>
<a name="L442"></a><tt class="py-lineno">442</tt>  <tt class="py-line"><tt class="py-docstring">        """</tt> </tt>
<a name="L443"></a><tt class="py-lineno">443</tt>  <tt class="py-line">        <tt class="py-name">size</tt> <tt class="py-op">=</tt> <tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt><tt class="py-op">)</tt> </tt>
<a name="L444"></a><tt class="py-lineno">444</tt>  <tt class="py-line">        <tt class="py-name">sfitness</tt> <tt class="py-op">=</tt> <tt class="py-op">[</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt><tt class="py-op">[</tt><tt class="py-name">x</tt><tt class="py-op">]</tt><tt class="py-op">.</tt><tt id="link-21" class="py-name" targets="Method dose.dose.dose_functions.fitness()=dose.dose.dose_functions-class.html#fitness,Method dose.genetic.Organism.fitness()=dose.genetic.Organism-class.html#fitness"><a title="dose.dose.dose_functions.fitness
dose.genetic.Organism.fitness" class="py-name" href="#" onclick="return doclink('link-21', 'fitness', 'link-21');">fitness</a></tt><tt class="py-op">(</tt><tt class="py-op">)</tt> <tt class="py-keyword">for</tt> <tt class="py-name">x</tt> <tt class="py-keyword">in</tt> <tt id="link-22" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.range" class="py-name" href="#" onclick="return doclink('link-22', 'range', 'link-4');">range</a></tt><tt class="py-op">(</tt><tt class="py-name">size</tt><tt class="py-op">)</tt><tt class="py-op">]</tt> </tt>
<a name="L445"></a><tt class="py-lineno">445</tt>  <tt class="py-line">        <tt class="py-name">threshold</tt> <tt class="py-op">=</tt> <tt class="py-name">sum</tt><tt class="py-op">(</tt><tt class="py-name">sfitness</tt><tt class="py-op">)</tt> <tt class="py-op">/</tt> <tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">sfitness</tt><tt class="py-op">)</tt> </tt>
<a name="L446"></a><tt class="py-lineno">446</tt>  <tt class="py-line">        <tt class="py-name">temp</tt> <tt class="py-op">=</tt> <tt class="py-op">[</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt><tt class="py-op">[</tt><tt class="py-name">x</tt><tt class="py-op">]</tt> </tt>
<a name="L447"></a><tt class="py-lineno">447</tt>  <tt class="py-line">                <tt class="py-keyword">for</tt> <tt class="py-name">x</tt> <tt class="py-keyword">in</tt> <tt id="link-23" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.range" class="py-name" href="#" onclick="return doclink('link-23', 'range', 'link-4');">range</a></tt><tt class="py-op">(</tt><tt class="py-name">size</tt><tt class="py-op">)</tt>  </tt>
<a name="L448"></a><tt class="py-lineno">448</tt>  <tt class="py-line">                    <tt class="py-keyword">if</tt> <tt class="py-name">sfitness</tt><tt class="py-op">[</tt><tt class="py-name">x</tt><tt class="py-op">]</tt> <tt class="py-op">&gt;</tt> <tt class="py-name">threshold</tt><tt class="py-op">]</tt> </tt>
<a name="L449"></a><tt class="py-lineno">449</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">temp</tt><tt class="py-op">)</tt> <tt class="py-op">&gt;</tt> <tt class="py-number">2001</tt><tt class="py-op">:</tt> </tt>
<a name="L450"></a><tt class="py-lineno">450</tt>  <tt class="py-line">            <tt class="py-name">size</tt> <tt class="py-op">=</tt> <tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">temp</tt><tt class="py-op">)</tt> </tt>
<a name="L451"></a><tt class="py-lineno">451</tt>  <tt class="py-line">            <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt> <tt class="py-op">=</tt> <tt class="py-op">[</tt><tt class="py-name">temp</tt><tt class="py-op">[</tt><tt id="link-24" class="py-name"><a title="dose.copads.statisticsdistribution.BetaDistribution.random
dose.copads.statisticsdistribution.BradfordDistribution.random
dose.copads.statisticsdistribution.BurrDistribution.random
dose.copads.statisticsdistribution.CauchyDistribution.random
dose.copads.statisticsdistribution.ExponentialDistribution.random
dose.copads.statisticsdistribution.GumbelDistribution.random
dose.copads.statisticsdistribution.LogNormalDistribution.random
dose.copads.statisticsdistribution.NormalDistribution.random
dose.copads.statisticsdistribution.ParetoDistribution.random
dose.copads.statisticsdistribution.UniformDistribution.random
dose.copads.statisticsdistribution.WeiBullDistribution.random" class="py-name" href="#" onclick="return doclink('link-24', 'random', 'link-0');">random</a></tt><tt class="py-op">.</tt><tt class="py-name">randint</tt><tt class="py-op">(</tt><tt class="py-number">0</tt><tt class="py-op">,</tt> <tt class="py-name">size</tt> <tt class="py-op">-</tt> <tt class="py-number">1</tt><tt class="py-op">)</tt><tt class="py-op">]</tt>  </tt>
<a name="L452"></a><tt class="py-lineno">452</tt>  <tt class="py-line">                           <tt class="py-keyword">for</tt> <tt class="py-name">x</tt> <tt class="py-keyword">in</tt> <tt id="link-25" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.range" class="py-name" href="#" onclick="return doclink('link-25', 'range', 'link-4');">range</a></tt><tt class="py-op">(</tt><tt class="py-number">2000</tt><tt class="py-op">)</tt><tt class="py-op">]</tt> </tt>
<a name="L453"></a><tt class="py-lineno">453</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">temp</tt><tt class="py-op">)</tt> <tt class="py-op">&gt;</tt> <tt class="py-number">21</tt><tt class="py-op">:</tt>  </tt>
<a name="L454"></a><tt class="py-lineno">454</tt>  <tt class="py-line">            <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt> <tt class="py-op">=</tt> <tt class="py-name">temp</tt> </tt>
</div><a name="L455"></a><tt class="py-lineno">455</tt>  <tt class="py-line">     </tt>
<a name="Population.mating"></a><div id="Population.mating-def"><a name="L456"></a><tt class="py-lineno">456</tt> <a class="py-toggle" href="#" id="Population.mating-toggle" onclick="return toggle('Population.mating');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Population-class.html#mating">mating</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Population.mating-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Population.mating-expanded"><a name="L457"></a><tt class="py-lineno">457</tt>  <tt class="py-line">        <tt class="py-docstring">"""</tt> </tt>
<a name="L458"></a><tt class="py-lineno">458</tt>  <tt class="py-line"><tt class="py-docstring">        Function to trigger mating events in each generation. B{This function</tt> </tt>
<a name="L459"></a><tt class="py-lineno">459</tt>  <tt class="py-line"><tt class="py-docstring">        may be over-ridden by the inherited class or substituted to cater for </tt> </tt>
<a name="L460"></a><tt class="py-lineno">460</tt>  <tt class="py-line"><tt class="py-docstring">        specific mating schemes but not an absolute requirement to do so.} </tt> </tt>
<a name="L461"></a><tt class="py-lineno">461</tt>  <tt class="py-line"><tt class="py-docstring">        Mating schemes should include </tt> </tt>
<a name="L462"></a><tt class="py-lineno">462</tt>  <tt class="py-line"><tt class="py-docstring">            - selection of mating partners using Organism.fitness() function</tt> </tt>
<a name="L463"></a><tt class="py-lineno">463</tt>  <tt class="py-line"><tt class="py-docstring">                and/or other status</tt> </tt>
<a name="L464"></a><tt class="py-lineno">464</tt>  <tt class="py-line"><tt class="py-docstring">            - processes and actions of mating</tt> </tt>
<a name="L465"></a><tt class="py-lineno">465</tt>  <tt class="py-line"><tt class="py-docstring">            </tt> </tt>
<a name="L466"></a><tt class="py-lineno">466</tt>  <tt class="py-line"><tt class="py-docstring">        Here, the sample implementation randomly selects any 2 organisms from</tt> </tt>
<a name="L467"></a><tt class="py-lineno">467</tt>  <tt class="py-line"><tt class="py-docstring">        the culled list and generate a new genome for the progeny organism by</tt> </tt>
<a name="L468"></a><tt class="py-lineno">468</tt>  <tt class="py-line"><tt class="py-docstring">        single crossover.</tt> </tt>
<a name="L469"></a><tt class="py-lineno">469</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L470"></a><tt class="py-lineno">470</tt>  <tt class="py-line"><tt class="py-docstring">        @since: version 0.4</tt> </tt>
<a name="L471"></a><tt class="py-lineno">471</tt>  <tt class="py-line"><tt class="py-docstring">        """</tt> </tt>
<a name="L472"></a><tt class="py-lineno">472</tt>  <tt class="py-line">        <tt class="py-name">size</tt> <tt class="py-op">=</tt> <tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt><tt class="py-op">)</tt> </tt>
<a name="L473"></a><tt class="py-lineno">473</tt>  <tt class="py-line">        <tt class="py-name">temp</tt> <tt class="py-op">=</tt> <tt class="py-op">[</tt><tt class="py-op">]</tt> </tt>
<a name="L474"></a><tt class="py-lineno">474</tt>  <tt class="py-line">        <tt class="py-keyword">for</tt> <tt class="py-name">x</tt> <tt class="py-keyword">in</tt> <tt id="link-26" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.range" class="py-name" href="#" onclick="return doclink('link-26', 'range', 'link-4');">range</a></tt><tt class="py-op">(</tt><tt class="py-name">size</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
<a name="L475"></a><tt class="py-lineno">475</tt>  <tt class="py-line">            <tt class="py-name">organism1</tt> <tt class="py-op">=</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt><tt class="py-op">[</tt><tt id="link-27" class="py-name"><a title="dose.copads.statisticsdistribution.BetaDistribution.random
dose.copads.statisticsdistribution.BradfordDistribution.random
dose.copads.statisticsdistribution.BurrDistribution.random
dose.copads.statisticsdistribution.CauchyDistribution.random
dose.copads.statisticsdistribution.ExponentialDistribution.random
dose.copads.statisticsdistribution.GumbelDistribution.random
dose.copads.statisticsdistribution.LogNormalDistribution.random
dose.copads.statisticsdistribution.NormalDistribution.random
dose.copads.statisticsdistribution.ParetoDistribution.random
dose.copads.statisticsdistribution.UniformDistribution.random
dose.copads.statisticsdistribution.WeiBullDistribution.random" class="py-name" href="#" onclick="return doclink('link-27', 'random', 'link-0');">random</a></tt><tt class="py-op">.</tt><tt class="py-name">randint</tt><tt class="py-op">(</tt><tt class="py-number">0</tt><tt class="py-op">,</tt> <tt class="py-name">size</tt> <tt class="py-op">-</tt> <tt class="py-number">1</tt><tt class="py-op">)</tt><tt class="py-op">]</tt> </tt>
<a name="L476"></a><tt class="py-lineno">476</tt>  <tt class="py-line">            <tt class="py-name">organism2</tt> <tt class="py-op">=</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt><tt class="py-op">[</tt><tt id="link-28" class="py-name"><a title="dose.copads.statisticsdistribution.BetaDistribution.random
dose.copads.statisticsdistribution.BradfordDistribution.random
dose.copads.statisticsdistribution.BurrDistribution.random
dose.copads.statisticsdistribution.CauchyDistribution.random
dose.copads.statisticsdistribution.ExponentialDistribution.random
dose.copads.statisticsdistribution.GumbelDistribution.random
dose.copads.statisticsdistribution.LogNormalDistribution.random
dose.copads.statisticsdistribution.NormalDistribution.random
dose.copads.statisticsdistribution.ParetoDistribution.random
dose.copads.statisticsdistribution.UniformDistribution.random
dose.copads.statisticsdistribution.WeiBullDistribution.random" class="py-name" href="#" onclick="return doclink('link-28', 'random', 'link-0');">random</a></tt><tt class="py-op">.</tt><tt class="py-name">randint</tt><tt class="py-op">(</tt><tt class="py-number">0</tt><tt class="py-op">,</tt> <tt class="py-name">size</tt> <tt class="py-op">-</tt> <tt class="py-number">1</tt><tt class="py-op">)</tt><tt class="py-op">]</tt> </tt>
<a name="L477"></a><tt class="py-lineno">477</tt>  <tt class="py-line">            <tt class="py-name">crossover_pt</tt> <tt class="py-op">=</tt> <tt id="link-29" class="py-name"><a title="dose.copads.statisticsdistribution.BetaDistribution.random
dose.copads.statisticsdistribution.BradfordDistribution.random
dose.copads.statisticsdistribution.BurrDistribution.random
dose.copads.statisticsdistribution.CauchyDistribution.random
dose.copads.statisticsdistribution.ExponentialDistribution.random
dose.copads.statisticsdistribution.GumbelDistribution.random
dose.copads.statisticsdistribution.LogNormalDistribution.random
dose.copads.statisticsdistribution.NormalDistribution.random
dose.copads.statisticsdistribution.ParetoDistribution.random
dose.copads.statisticsdistribution.UniformDistribution.random
dose.copads.statisticsdistribution.WeiBullDistribution.random" class="py-name" href="#" onclick="return doclink('link-29', 'random', 'link-0');">random</a></tt><tt class="py-op">.</tt><tt class="py-name">randint</tt><tt class="py-op">(</tt><tt class="py-number">0</tt><tt class="py-op">,</tt> <tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">organism1</tt><tt class="py-op">.</tt><tt class="py-name">genome</tt><tt class="py-op">[</tt><tt class="py-number">0</tt><tt class="py-op">]</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt><tt class="py-op">)</tt><tt class="py-op">)</tt> </tt>
<a name="L478"></a><tt class="py-lineno">478</tt>  <tt class="py-line">            <tt class="py-op">(</tt><tt class="py-name">g1</tt><tt class="py-op">,</tt> <tt class="py-name">g2</tt><tt class="py-op">)</tt> <tt class="py-op">=</tt> <tt id="link-30" class="py-name" targets="Function dose.genetic.crossover()=dose.genetic-module.html#crossover"><a title="dose.genetic.crossover" class="py-name" href="#" onclick="return doclink('link-30', 'crossover', 'link-30');">crossover</a></tt><tt class="py-op">(</tt><tt class="py-name">organism1</tt><tt class="py-op">.</tt><tt class="py-name">genome</tt><tt class="py-op">[</tt><tt class="py-number">0</tt><tt class="py-op">]</tt><tt class="py-op">,</tt> <tt class="py-name">organism2</tt><tt class="py-op">.</tt><tt class="py-name">genome</tt><tt class="py-op">[</tt><tt class="py-number">0</tt><tt class="py-op">]</tt><tt class="py-op">,</tt> </tt>
<a name="L479"></a><tt class="py-lineno">479</tt>  <tt class="py-line">                                 <tt class="py-name">crossover_pt</tt><tt class="py-op">)</tt> </tt>
<a name="L480"></a><tt class="py-lineno">480</tt>  <tt class="py-line">            <tt class="py-name">temp</tt> <tt class="py-op">=</tt> <tt class="py-name">temp</tt> <tt class="py-op">+</tt> <tt class="py-op">[</tt><tt id="link-31" class="py-name" targets="Class dose.genetic.Organism=dose.genetic.Organism-class.html"><a title="dose.genetic.Organism" class="py-name" href="#" onclick="return doclink('link-31', 'Organism', 'link-31');">Organism</a></tt><tt class="py-op">(</tt><tt class="py-op">[</tt><tt class="py-name">g1</tt><tt class="py-op">]</tt><tt class="py-op">)</tt><tt class="py-op">]</tt> </tt>
<a name="L481"></a><tt class="py-lineno">481</tt>  <tt class="py-line">        <tt class="py-name">self</tt><tt class="py-op">.</tt><tt id="link-32" class="py-name" targets="Method dose.genetic.Population.add_organism()=dose.genetic.Population-class.html#add_organism"><a title="dose.genetic.Population.add_organism" class="py-name" href="#" onclick="return doclink('link-32', 'add_organism', 'link-32');">add_organism</a></tt><tt class="py-op">(</tt><tt class="py-name">temp</tt><tt class="py-op">)</tt> </tt>
</div><a name="L482"></a><tt class="py-lineno">482</tt>  <tt class="py-line">             </tt>
<a name="Population.postpopulation_control"></a><div id="Population.postpopulation_control-def"><a name="L483"></a><tt class="py-lineno">483</tt> <a class="py-toggle" href="#" id="Population.postpopulation_control-toggle" onclick="return toggle('Population.postpopulation_control');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Population-class.html#postpopulation_control">postpopulation_control</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Population.postpopulation_control-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Population.postpopulation_control-expanded"><a name="L484"></a><tt class="py-lineno">484</tt>  <tt class="py-line">        <tt class="py-docstring">"""</tt> </tt>
<a name="L485"></a><tt class="py-lineno">485</tt>  <tt class="py-line"><tt class="py-docstring">        Function to trigger population control events after mating event in </tt> </tt>
<a name="L486"></a><tt class="py-lineno">486</tt>  <tt class="py-line"><tt class="py-docstring">        each generation(For example, to simulate old-age death). B{This </tt> </tt>
<a name="L487"></a><tt class="py-lineno">487</tt>  <tt class="py-line"><tt class="py-docstring">        function may be over-ridden by the inherited class or substituted to </tt> </tt>
<a name="L488"></a><tt class="py-lineno">488</tt>  <tt class="py-line"><tt class="py-docstring">        cater for specific events.} Although this is not an absolute </tt> </tt>
<a name="L489"></a><tt class="py-lineno">489</tt>  <tt class="py-line"><tt class="py-docstring">        requirement, it is extremely encouraged to prevent exhaustion of </tt> </tt>
<a name="L490"></a><tt class="py-lineno">490</tt>  <tt class="py-line"><tt class="py-docstring">        memory space. Without population control, it will seems like a </tt> </tt>
<a name="L491"></a><tt class="py-lineno">491</tt>  <tt class="py-line"><tt class="py-docstring">        reproducing immortal population.</tt> </tt>
<a name="L492"></a><tt class="py-lineno">492</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L493"></a><tt class="py-lineno">493</tt>  <tt class="py-line"><tt class="py-docstring">        @since: version 0.4</tt> </tt>
<a name="L494"></a><tt class="py-lineno">494</tt>  <tt class="py-line"><tt class="py-docstring">        """</tt> </tt>
<a name="L495"></a><tt class="py-lineno">495</tt>  <tt class="py-line">        <tt class="py-keyword">pass</tt> </tt>
</div><a name="L496"></a><tt class="py-lineno">496</tt>  <tt class="py-line">         </tt>
<a name="Population.generation_events"></a><div id="Population.generation_events-def"><a name="L497"></a><tt class="py-lineno">497</tt> <a class="py-toggle" href="#" id="Population.generation_events-toggle" onclick="return toggle('Population.generation_events');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Population-class.html#generation_events">generation_events</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Population.generation_events-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Population.generation_events-expanded"><a name="L498"></a><tt class="py-lineno">498</tt>  <tt class="py-line">        <tt class="py-docstring">"""</tt> </tt>
<a name="L499"></a><tt class="py-lineno">499</tt>  <tt class="py-line"><tt class="py-docstring">        Function to trigger other defined events in each generation. B{This </tt> </tt>
<a name="L500"></a><tt class="py-lineno">500</tt>  <tt class="py-line"><tt class="py-docstring">        function may be over-ridden by the inherited class or substituted to </tt> </tt>
<a name="L501"></a><tt class="py-lineno">501</tt>  <tt class="py-line"><tt class="py-docstring">        cater for specific events but not an absolute requirement to do so.} </tt> </tt>
<a name="L502"></a><tt class="py-lineno">502</tt>  <tt class="py-line"><tt class="py-docstring">        Events and controls may include</tt> </tt>
<a name="L503"></a><tt class="py-lineno">503</tt>  <tt class="py-line"><tt class="py-docstring">            - processes simulating disaster or other catastrophic events</tt> </tt>
<a name="L504"></a><tt class="py-lineno">504</tt>  <tt class="py-line"><tt class="py-docstring">            - changes in mutations </tt> </tt>
<a name="L505"></a><tt class="py-lineno">505</tt>  <tt class="py-line"><tt class="py-docstring">            </tt> </tt>
<a name="L506"></a><tt class="py-lineno">506</tt>  <tt class="py-line"><tt class="py-docstring">        @since: version 0.4           </tt> </tt>
<a name="L507"></a><tt class="py-lineno">507</tt>  <tt class="py-line"><tt class="py-docstring">        """</tt> </tt>
<a name="L508"></a><tt class="py-lineno">508</tt>  <tt class="py-line">        <tt class="py-keyword">pass</tt> </tt>
</div><a name="L509"></a><tt class="py-lineno">509</tt>  <tt class="py-line">     </tt>
<a name="Population.report"></a><div id="Population.report-def"><a name="L510"></a><tt class="py-lineno">510</tt> <a class="py-toggle" href="#" id="Population.report-toggle" onclick="return toggle('Population.report');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Population-class.html#report">report</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Population.report-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Population.report-expanded"><a name="L511"></a><tt class="py-lineno">511</tt>  <tt class="py-line">        <tt class="py-docstring">"""</tt> </tt>
<a name="L512"></a><tt class="py-lineno">512</tt>  <tt class="py-line"><tt class="py-docstring">        Function to report the status of each generation. B{This function may </tt> </tt>
<a name="L513"></a><tt class="py-lineno">513</tt>  <tt class="py-line"><tt class="py-docstring">        be over-ridden by the inherited class or substituted to cater for </tt> </tt>
<a name="L514"></a><tt class="py-lineno">514</tt>  <tt class="py-line"><tt class="py-docstring">        specific reporting schemes but not an absolute requirement to do so.} </tt> </tt>
<a name="L515"></a><tt class="py-lineno">515</tt>  <tt class="py-line"><tt class="py-docstring">        At the very least, this function should report whether the goal is </tt> </tt>
<a name="L516"></a><tt class="py-lineno">516</tt>  <tt class="py-line"><tt class="py-docstring">        reached.</tt> </tt>
<a name="L517"></a><tt class="py-lineno">517</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L518"></a><tt class="py-lineno">518</tt>  <tt class="py-line"><tt class="py-docstring">        @return: dictionary of status describing the current generation</tt> </tt>
<a name="L519"></a><tt class="py-lineno">519</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L520"></a><tt class="py-lineno">520</tt>  <tt class="py-line"><tt class="py-docstring">        @since: version 0.4</tt> </tt>
<a name="L521"></a><tt class="py-lineno">521</tt>  <tt class="py-line"><tt class="py-docstring">        """</tt> </tt>
<a name="L522"></a><tt class="py-lineno">522</tt>  <tt class="py-line">        <tt class="py-name">sfitness</tt> <tt class="py-op">=</tt> <tt class="py-op">[</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt><tt class="py-op">[</tt><tt class="py-name">x</tt><tt class="py-op">]</tt><tt class="py-op">.</tt><tt id="link-33" class="py-name"><a title="dose.dose.dose_functions.fitness
dose.genetic.Organism.fitness" class="py-name" href="#" onclick="return doclink('link-33', 'fitness', 'link-21');">fitness</a></tt><tt class="py-op">(</tt><tt class="py-op">)</tt>  </tt>
<a name="L523"></a><tt class="py-lineno">523</tt>  <tt class="py-line">                    <tt class="py-keyword">for</tt> <tt class="py-name">x</tt> <tt class="py-keyword">in</tt> <tt id="link-34" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.range" class="py-name" href="#" onclick="return doclink('link-34', 'range', 'link-4');">range</a></tt><tt class="py-op">(</tt><tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt><tt class="py-op">)</tt><tt class="py-op">)</tt><tt class="py-op">]</tt> </tt>
<a name="L524"></a><tt class="py-lineno">524</tt>  <tt class="py-line">        <tt class="py-name">afitness</tt> <tt class="py-op">=</tt> <tt class="py-name">sum</tt><tt class="py-op">(</tt><tt class="py-name">sfitness</tt><tt class="py-op">)</tt> <tt class="py-op">/</tt> <tt class="py-name">float</tt><tt class="py-op">(</tt><tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt><tt class="py-op">)</tt><tt class="py-op">)</tt> </tt>
<a name="L525"></a><tt class="py-lineno">525</tt>  <tt class="py-line">        <tt class="py-keyword">return</tt> <tt class="py-op">{</tt><tt class="py-string">'generation'</tt><tt class="py-op">:</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">generation</tt><tt class="py-op">,</tt> </tt>
<a name="L526"></a><tt class="py-lineno">526</tt>  <tt class="py-line">                <tt class="py-string">'average fitness'</tt><tt class="py-op">:</tt> <tt class="py-name">afitness</tt><tt class="py-op">,</tt> </tt>
<a name="L527"></a><tt class="py-lineno">527</tt>  <tt class="py-line">                <tt class="py-string">'% to goal'</tt><tt class="py-op">:</tt> <tt class="py-name">float</tt><tt class="py-op">(</tt><tt class="py-name">afitness</tt> <tt class="py-op">-</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">goal</tt><tt class="py-op">)</tt> <tt class="py-op">/</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">goal</tt> <tt class="py-op">*</tt> <tt class="py-number">100</tt><tt class="py-op">}</tt> </tt>
</div><a name="L528"></a><tt class="py-lineno">528</tt>  <tt class="py-line">         </tt>
<a name="Population.generation_step"></a><div id="Population.generation_step-def"><a name="L529"></a><tt class="py-lineno">529</tt> <a class="py-toggle" href="#" id="Population.generation_step-toggle" onclick="return toggle('Population.generation_step');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Population-class.html#generation_step">generation_step</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Population.generation_step-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Population.generation_step-expanded"><a name="L530"></a><tt class="py-lineno">530</tt>  <tt class="py-line">        <tt class="py-docstring">"""</tt> </tt>
<a name="L531"></a><tt class="py-lineno">531</tt>  <tt class="py-line"><tt class="py-docstring">        Function to simulate events for one generation. These includes</tt> </tt>
<a name="L532"></a><tt class="py-lineno">532</tt>  <tt class="py-line"><tt class="py-docstring">            - mating (according to the mating scheme or function)</tt> </tt>
<a name="L533"></a><tt class="py-lineno">533</tt>  <tt class="py-line"><tt class="py-docstring">            - mutating each organism in the population</tt> </tt>
<a name="L534"></a><tt class="py-lineno">534</tt>  <tt class="py-line"><tt class="py-docstring">            - population size control</tt> </tt>
<a name="L535"></a><tt class="py-lineno">535</tt>  <tt class="py-line"><tt class="py-docstring">            - other events defined under generation_events function</tt> </tt>
<a name="L536"></a><tt class="py-lineno">536</tt>  <tt class="py-line"><tt class="py-docstring">            - increment of generation count</tt> </tt>
<a name="L537"></a><tt class="py-lineno">537</tt>  <tt class="py-line"><tt class="py-docstring">            - reporting the population status</tt> </tt>
<a name="L538"></a><tt class="py-lineno">538</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L539"></a><tt class="py-lineno">539</tt>  <tt class="py-line"><tt class="py-docstring">        @return: information returned from report function.</tt> </tt>
<a name="L540"></a><tt class="py-lineno">540</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L541"></a><tt class="py-lineno">541</tt>  <tt class="py-line"><tt class="py-docstring">        @since: version 0.4</tt> </tt>
<a name="L542"></a><tt class="py-lineno">542</tt>  <tt class="py-line"><tt class="py-docstring">        """</tt> </tt>
<a name="L543"></a><tt class="py-lineno">543</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">generation</tt> <tt class="py-op">&gt;</tt> <tt class="py-number">0</tt><tt class="py-op">:</tt> </tt>
<a name="L544"></a><tt class="py-lineno">544</tt>  <tt class="py-line">            <tt class="py-name">self</tt><tt class="py-op">.</tt><tt id="link-35" class="py-name" targets="Method dose.dose.dose_functions.prepopulation_control()=dose.dose.dose_functions-class.html#prepopulation_control,Method dose.genetic.Population.prepopulation_control()=dose.genetic.Population-class.html#prepopulation_control"><a title="dose.dose.dose_functions.prepopulation_control
dose.genetic.Population.prepopulation_control" class="py-name" href="#" onclick="return doclink('link-35', 'prepopulation_control', 'link-35');">prepopulation_control</a></tt><tt class="py-op">(</tt><tt class="py-op">)</tt> </tt>
<a name="L545"></a><tt class="py-lineno">545</tt>  <tt class="py-line">        <tt class="py-name">self</tt><tt class="py-op">.</tt><tt id="link-36" class="py-name" targets="Method dose.dose.dose_functions.mating()=dose.dose.dose_functions-class.html#mating,Method dose.genetic.Population.mating()=dose.genetic.Population-class.html#mating"><a title="dose.dose.dose_functions.mating
dose.genetic.Population.mating" class="py-name" href="#" onclick="return doclink('link-36', 'mating', 'link-36');">mating</a></tt><tt class="py-op">(</tt><tt class="py-op">)</tt> </tt>
<a name="L546"></a><tt class="py-lineno">546</tt>  <tt class="py-line">        <tt class="py-name">self</tt><tt class="py-op">.</tt><tt id="link-37" class="py-name" targets="Method dose.dose.dose_functions.postpopulation_control()=dose.dose.dose_functions-class.html#postpopulation_control,Method dose.genetic.Population.postpopulation_control()=dose.genetic.Population-class.html#postpopulation_control"><a title="dose.dose.dose_functions.postpopulation_control
dose.genetic.Population.postpopulation_control" class="py-name" href="#" onclick="return doclink('link-37', 'postpopulation_control', 'link-37');">postpopulation_control</a></tt><tt class="py-op">(</tt><tt class="py-op">)</tt> </tt>
<a name="L547"></a><tt class="py-lineno">547</tt>  <tt class="py-line">        <tt class="py-keyword">for</tt> <tt class="py-name">organism</tt> <tt class="py-keyword">in</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt><tt class="py-op">:</tt> </tt>
<a name="L548"></a><tt class="py-lineno">548</tt>  <tt class="py-line">            <tt class="py-name">organism</tt><tt class="py-op">.</tt><tt id="link-38" class="py-name" targets="Method dose.dose.dose_functions.mutation_scheme()=dose.dose.dose_functions-class.html#mutation_scheme,Method dose.genetic.Organism.mutation_scheme()=dose.genetic.Organism-class.html#mutation_scheme"><a title="dose.dose.dose_functions.mutation_scheme
dose.genetic.Organism.mutation_scheme" class="py-name" href="#" onclick="return doclink('link-38', 'mutation_scheme', 'link-38');">mutation_scheme</a></tt><tt class="py-op">(</tt><tt class="py-op">)</tt>  </tt>
<a name="L549"></a><tt class="py-lineno">549</tt>  <tt class="py-line">        <tt class="py-name">self</tt><tt class="py-op">.</tt><tt id="link-39" class="py-name" targets="Method dose.dose.dose_functions.generation_events()=dose.dose.dose_functions-class.html#generation_events,Method dose.genetic.Population.generation_events()=dose.genetic.Population-class.html#generation_events"><a title="dose.dose.dose_functions.generation_events
dose.genetic.Population.generation_events" class="py-name" href="#" onclick="return doclink('link-39', 'generation_events', 'link-39');">generation_events</a></tt><tt class="py-op">(</tt><tt class="py-op">)</tt> </tt>
<a name="L550"></a><tt class="py-lineno">550</tt>  <tt class="py-line">        <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">generation</tt> <tt class="py-op">=</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">generation</tt> <tt class="py-op">+</tt> <tt class="py-number">1</tt> </tt>
<a name="L551"></a><tt class="py-lineno">551</tt>  <tt class="py-line">        <tt class="py-keyword">return</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt id="link-40" class="py-name" targets="Method dose.dose.dose_functions.report()=dose.dose.dose_functions-class.html#report,Method dose.genetic.Population.report()=dose.genetic.Population-class.html#report"><a title="dose.dose.dose_functions.report
dose.genetic.Population.report" class="py-name" href="#" onclick="return doclink('link-40', 'report', 'link-40');">report</a></tt><tt class="py-op">(</tt><tt class="py-op">)</tt> </tt>
</div><a name="L552"></a><tt class="py-lineno">552</tt>  <tt class="py-line">     </tt>
<a name="Population.add_organism"></a><div id="Population.add_organism-def"><a name="L553"></a><tt class="py-lineno">553</tt> <a class="py-toggle" href="#" id="Population.add_organism-toggle" onclick="return toggle('Population.add_organism');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Population-class.html#add_organism">add_organism</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">,</tt> <tt class="py-param">organism</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Population.add_organism-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Population.add_organism-expanded"><a name="L554"></a><tt class="py-lineno">554</tt>  <tt class="py-line">        <tt class="py-docstring">"""Add a new organism(s) to the population.</tt> </tt>
<a name="L555"></a><tt class="py-lineno">555</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L556"></a><tt class="py-lineno">556</tt>  <tt class="py-line"><tt class="py-docstring">        @param organism: list of new Organism object(s)</tt> </tt>
<a name="L557"></a><tt class="py-lineno">557</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L558"></a><tt class="py-lineno">558</tt>  <tt class="py-line"><tt class="py-docstring">        @since: version 0.4"""</tt> </tt>
<a name="L559"></a><tt class="py-lineno">559</tt>  <tt class="py-line">        <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt> <tt class="py-op">=</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt> <tt class="py-op">+</tt> <tt class="py-name">organism</tt> </tt>
</div><a name="L560"></a><tt class="py-lineno">560</tt>  <tt class="py-line">         </tt>
<a name="Population.freeze"></a><div id="Population.freeze-def"><a name="L561"></a><tt class="py-lineno">561</tt> <a class="py-toggle" href="#" id="Population.freeze-toggle" onclick="return toggle('Population.freeze');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Population-class.html#freeze">freeze</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">,</tt> <tt class="py-param">prefix</tt><tt class="py-op">=</tt><tt class="py-string">'pop'</tt><tt class="py-op">,</tt> <tt class="py-param">proportion</tt><tt class="py-op">=</tt><tt class="py-number">0.01</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Population.freeze-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Population.freeze-expanded"><a name="L562"></a><tt class="py-lineno">562</tt>  <tt class="py-line">        <tt class="py-docstring">"""</tt> </tt>
<a name="L563"></a><tt class="py-lineno">563</tt>  <tt class="py-line"><tt class="py-docstring">        Preserves part or the entire population. If the population size or </tt> </tt>
<a name="L564"></a><tt class="py-lineno">564</tt>  <tt class="py-line"><tt class="py-docstring">        the preserved proportion is below 100, the entire population will be </tt> </tt>
<a name="L565"></a><tt class="py-lineno">565</tt>  <tt class="py-line"><tt class="py-docstring">        preserved. The preserved sample will be written into a file with name</tt> </tt>
<a name="L566"></a><tt class="py-lineno">566</tt>  <tt class="py-line"><tt class="py-docstring">        in the following format - &lt;prefix&gt;&lt;generation count&gt;_&lt;sample size&gt;.gap</tt> </tt>
<a name="L567"></a><tt class="py-lineno">567</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L568"></a><tt class="py-lineno">568</tt>  <tt class="py-line"><tt class="py-docstring">        @param prefix: prefix of file name. Default = 'pop'.</tt> </tt>
<a name="L569"></a><tt class="py-lineno">569</tt>  <tt class="py-line"><tt class="py-docstring">        @param proportion: proportion of population to be preserved.</tt> </tt>
<a name="L570"></a><tt class="py-lineno">570</tt>  <tt class="py-line"><tt class="py-docstring">            Default = 0.01, preserves 1% of the population.</tt> </tt>
<a name="L571"></a><tt class="py-lineno">571</tt>  <tt class="py-line"><tt class="py-docstring">            </tt> </tt>
<a name="L572"></a><tt class="py-lineno">572</tt>  <tt class="py-line"><tt class="py-docstring">        @since: version 0.4</tt> </tt>
<a name="L573"></a><tt class="py-lineno">573</tt>  <tt class="py-line"><tt class="py-docstring">        """</tt> </tt>
<a name="L574"></a><tt class="py-lineno">574</tt>  <tt class="py-line">        <tt class="py-comment"># In Python 3, cPickle is no longer needed: Py3 looks for</tt> </tt>
<a name="L575"></a><tt class="py-lineno">575</tt>  <tt class="py-line">        <tt class="py-comment"># an optimized version, and if it founds none, will load the</tt> </tt>
<a name="L576"></a><tt class="py-lineno">576</tt>  <tt class="py-line">        <tt class="py-comment"># pure python implementation of pickle. </tt> </tt>
<a name="L577"></a><tt class="py-lineno">577</tt>  <tt class="py-line">        <tt class="py-keyword">try</tt><tt class="py-op">:</tt> </tt>
<a name="L578"></a><tt class="py-lineno">578</tt>  <tt class="py-line">            <tt class="py-keyword">import</tt> <tt class="py-name">cPickle</tt> <tt class="py-keyword">as</tt> <tt class="py-name">pickle</tt> </tt>
<a name="L579"></a><tt class="py-lineno">579</tt>  <tt class="py-line">        <tt class="py-keyword">except</tt> <tt class="py-name">ImportError</tt><tt class="py-op">:</tt> </tt>
<a name="L580"></a><tt class="py-lineno">580</tt>  <tt class="py-line">            <tt class="py-keyword">import</tt> <tt class="py-name">pickle</tt> </tt>
<a name="L581"></a><tt class="py-lineno">581</tt>  <tt class="py-line"> </tt>
<a name="L582"></a><tt class="py-lineno">582</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">proportion</tt> <tt class="py-op">&gt;</tt> <tt class="py-number">1.0</tt><tt class="py-op">:</tt> <tt class="py-name">proportion</tt> <tt class="py-op">=</tt> <tt class="py-number">1.0</tt> </tt>
<a name="L583"></a><tt class="py-lineno">583</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt><tt class="py-op">)</tt> <tt class="py-op">&lt;</tt> <tt class="py-number">101</tt> <tt class="py-keyword">or</tt> <tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt><tt class="py-op">)</tt> <tt class="py-op">*</tt> <tt class="py-name">proportion</tt> <tt class="py-op">&lt;</tt> <tt class="py-number">101</tt><tt class="py-op">:</tt> </tt>
<a name="L584"></a><tt class="py-lineno">584</tt>  <tt class="py-line">            <tt id="link-41" class="py-name" targets="Function dose.copads.operations.sample()=dose.copads.operations-module.html#sample,Variable dose.copads.samplestatistics.TwoSample.sample=dose.copads.samplestatistics.TwoSample-class.html#sample"><a title="dose.copads.operations.sample
dose.copads.samplestatistics.TwoSample.sample" class="py-name" href="#" onclick="return doclink('link-41', 'sample', 'link-41');">sample</a></tt> <tt class="py-op">=</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt> </tt>
<a name="L585"></a><tt class="py-lineno">585</tt>  <tt class="py-line">        <tt class="py-keyword">else</tt><tt class="py-op">:</tt> </tt>
<a name="L586"></a><tt class="py-lineno">586</tt>  <tt class="py-line">            <tt class="py-name">size</tt> <tt class="py-op">=</tt> <tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt><tt class="py-op">)</tt> </tt>
<a name="L587"></a><tt class="py-lineno">587</tt>  <tt class="py-line">            <tt id="link-42" class="py-name"><a title="dose.copads.operations.sample
dose.copads.samplestatistics.TwoSample.sample" class="py-name" href="#" onclick="return doclink('link-42', 'sample', 'link-41');">sample</a></tt> <tt class="py-op">=</tt> <tt class="py-op">[</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt><tt class="py-op">[</tt><tt id="link-43" class="py-name"><a title="dose.copads.statisticsdistribution.BetaDistribution.random
dose.copads.statisticsdistribution.BradfordDistribution.random
dose.copads.statisticsdistribution.BurrDistribution.random
dose.copads.statisticsdistribution.CauchyDistribution.random
dose.copads.statisticsdistribution.ExponentialDistribution.random
dose.copads.statisticsdistribution.GumbelDistribution.random
dose.copads.statisticsdistribution.LogNormalDistribution.random
dose.copads.statisticsdistribution.NormalDistribution.random
dose.copads.statisticsdistribution.ParetoDistribution.random
dose.copads.statisticsdistribution.UniformDistribution.random
dose.copads.statisticsdistribution.WeiBullDistribution.random" class="py-name" href="#" onclick="return doclink('link-43', 'random', 'link-0');">random</a></tt><tt class="py-op">.</tt><tt class="py-name">randint</tt><tt class="py-op">(</tt><tt class="py-number">0</tt><tt class="py-op">,</tt> <tt class="py-name">size</tt> <tt class="py-op">-</tt> <tt class="py-number">1</tt><tt class="py-op">)</tt><tt class="py-op">]</tt> </tt>
<a name="L588"></a><tt class="py-lineno">588</tt>  <tt class="py-line">                      <tt class="py-keyword">for</tt> <tt class="py-name">x</tt> <tt class="py-keyword">in</tt> <tt id="link-44" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.range" class="py-name" href="#" onclick="return doclink('link-44', 'range', 'link-4');">range</a></tt><tt class="py-op">(</tt><tt class="py-name">int</tt><tt class="py-op">(</tt><tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt><tt class="py-op">)</tt> <tt class="py-op">*</tt> <tt class="py-name">proportion</tt><tt class="py-op">)</tt><tt class="py-op">)</tt><tt class="py-op">]</tt> </tt>
<a name="L589"></a><tt class="py-lineno">589</tt>  <tt class="py-line">        <tt id="link-45" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.name" class="py-name" href="#" onclick="return doclink('link-45', 'name', 'link-15');">name</a></tt> <tt class="py-op">=</tt> <tt class="py-string">''</tt><tt class="py-op">.</tt><tt class="py-name">join</tt><tt class="py-op">(</tt><tt class="py-op">[</tt><tt class="py-name">prefix</tt><tt class="py-op">,</tt> <tt class="py-name">str</tt><tt class="py-op">(</tt><tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">generation</tt><tt class="py-op">)</tt><tt class="py-op">,</tt> <tt class="py-string">'_'</tt><tt class="py-op">,</tt>  </tt>
<a name="L590"></a><tt class="py-lineno">590</tt>  <tt class="py-line">                        <tt class="py-name">str</tt><tt class="py-op">(</tt><tt class="py-name">len</tt><tt class="py-op">(</tt><tt id="link-46" class="py-name"><a title="dose.copads.operations.sample
dose.copads.samplestatistics.TwoSample.sample" class="py-name" href="#" onclick="return doclink('link-46', 'sample', 'link-41');">sample</a></tt><tt class="py-op">)</tt><tt class="py-op">)</tt><tt class="py-op">,</tt> <tt class="py-string">'.gap'</tt><tt class="py-op">]</tt><tt class="py-op">)</tt> </tt>
<a name="L591"></a><tt class="py-lineno">591</tt>  <tt class="py-line">        <tt class="py-name">f</tt> <tt class="py-op">=</tt> <tt class="py-name">open</tt><tt class="py-op">(</tt><tt id="link-47" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.name" class="py-name" href="#" onclick="return doclink('link-47', 'name', 'link-15');">name</a></tt><tt class="py-op">,</tt> <tt class="py-string">'wb'</tt><tt class="py-op">)</tt> </tt>
<a name="L592"></a><tt class="py-lineno">592</tt>  <tt class="py-line">        <tt class="py-name">pickle</tt><tt class="py-op">.</tt><tt class="py-name">dump</tt><tt class="py-op">(</tt><tt id="link-48" class="py-name"><a title="dose.copads.operations.sample
dose.copads.samplestatistics.TwoSample.sample" class="py-name" href="#" onclick="return doclink('link-48', 'sample', 'link-41');">sample</a></tt><tt class="py-op">,</tt> <tt class="py-name">f</tt><tt class="py-op">)</tt> </tt>
<a name="L593"></a><tt class="py-lineno">593</tt>  <tt class="py-line">        <tt class="py-name">f</tt><tt class="py-op">.</tt><tt class="py-name">close</tt><tt class="py-op">(</tt><tt class="py-op">)</tt> </tt>
</div><a name="L594"></a><tt class="py-lineno">594</tt>  <tt class="py-line">         </tt>
<a name="Population.revive"></a><div id="Population.revive-def"><a name="L595"></a><tt class="py-lineno">595</tt> <a class="py-toggle" href="#" id="Population.revive-toggle" onclick="return toggle('Population.revive');">-</a><tt class="py-line">    <tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic.Population-class.html#revive">revive</a><tt class="py-op">(</tt><tt class="py-param">self</tt><tt class="py-op">,</tt> <tt class="py-param">filename</tt><tt class="py-op">,</tt> <tt class="py-param">type</tt><tt class="py-op">=</tt><tt class="py-string">'replace'</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="Population.revive-collapsed" style="display:none;" pad="+++" indent="++++++++"></div><div id="Population.revive-expanded"><a name="L596"></a><tt class="py-lineno">596</tt>  <tt class="py-line">        <tt class="py-docstring">"""</tt> </tt>
<a name="L597"></a><tt class="py-lineno">597</tt>  <tt class="py-line"><tt class="py-docstring">        Revives a frozen population.</tt> </tt>
<a name="L598"></a><tt class="py-lineno">598</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L599"></a><tt class="py-lineno">599</tt>  <tt class="py-line"><tt class="py-docstring">        @param filename: file name of frozen population (generated by </tt> </tt>
<a name="L600"></a><tt class="py-lineno">600</tt>  <tt class="py-line"><tt class="py-docstring">            Population.freeze() function)</tt> </tt>
<a name="L601"></a><tt class="py-lineno">601</tt>  <tt class="py-line"><tt class="py-docstring">        @param type: type of revival. Allows 'replace' (replace the current </tt> </tt>
<a name="L602"></a><tt class="py-lineno">602</tt>  <tt class="py-line"><tt class="py-docstring">            population with the revived population) or 'add' (add the revived</tt> </tt>
<a name="L603"></a><tt class="py-lineno">603</tt>  <tt class="py-line"><tt class="py-docstring">            population to the current population). Default = 'replace'.</tt> </tt>
<a name="L604"></a><tt class="py-lineno">604</tt>  <tt class="py-line"><tt class="py-docstring">            </tt> </tt>
<a name="L605"></a><tt class="py-lineno">605</tt>  <tt class="py-line"><tt class="py-docstring">        @since: version 0.4</tt> </tt>
<a name="L606"></a><tt class="py-lineno">606</tt>  <tt class="py-line"><tt class="py-docstring">        """</tt> </tt>
<a name="L607"></a><tt class="py-lineno">607</tt>  <tt class="py-line">        <tt class="py-comment"># In Python 3, cPickle is no longer needed: Py3 looks for</tt> </tt>
<a name="L608"></a><tt class="py-lineno">608</tt>  <tt class="py-line">        <tt class="py-comment"># an optimized version, and if it founds none, will load the</tt> </tt>
<a name="L609"></a><tt class="py-lineno">609</tt>  <tt class="py-line">        <tt class="py-comment"># pure python implementation of pickle. </tt> </tt>
<a name="L610"></a><tt class="py-lineno">610</tt>  <tt class="py-line">        <tt class="py-keyword">try</tt><tt class="py-op">:</tt> </tt>
<a name="L611"></a><tt class="py-lineno">611</tt>  <tt class="py-line">            <tt class="py-keyword">import</tt> <tt class="py-name">cPickle</tt> <tt class="py-keyword">as</tt> <tt class="py-name">pickle</tt> </tt>
<a name="L612"></a><tt class="py-lineno">612</tt>  <tt class="py-line">        <tt class="py-keyword">except</tt> <tt class="py-name">ImportError</tt><tt class="py-op">:</tt> </tt>
<a name="L613"></a><tt class="py-lineno">613</tt>  <tt class="py-line">            <tt class="py-keyword">import</tt> <tt class="py-name">pickle</tt> </tt>
<a name="L614"></a><tt class="py-lineno">614</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">type</tt> <tt class="py-op">==</tt> <tt class="py-string">'replace'</tt><tt class="py-op">:</tt> </tt>
<a name="L615"></a><tt class="py-lineno">615</tt>  <tt class="py-line">            <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt> <tt class="py-op">=</tt> <tt class="py-name">pickle</tt><tt class="py-op">.</tt><tt class="py-name">load</tt><tt class="py-op">(</tt><tt class="py-name">open</tt><tt class="py-op">(</tt><tt class="py-name">filename</tt><tt class="py-op">,</tt> <tt class="py-string">'rb'</tt><tt class="py-op">)</tt><tt class="py-op">)</tt> </tt>
<a name="L616"></a><tt class="py-lineno">616</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">type</tt> <tt class="py-op">==</tt> <tt class="py-string">'add'</tt><tt class="py-op">:</tt> </tt>
<a name="L617"></a><tt class="py-lineno">617</tt>  <tt class="py-line">            <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt> <tt class="py-op">=</tt> <tt class="py-name">self</tt><tt class="py-op">.</tt><tt class="py-name">agents</tt> <tt class="py-op">+</tt> <tt class="py-name">pickle</tt><tt class="py-op">.</tt><tt class="py-name">load</tt><tt class="py-op">(</tt><tt class="py-name">open</tt><tt class="py-op">(</tt><tt class="py-name">filename</tt><tt class="py-op">,</tt> <tt class="py-string">'rb'</tt><tt class="py-op">)</tt><tt class="py-op">)</tt> </tt>
</div></div><a name="L618"></a><tt class="py-lineno">618</tt>  <tt class="py-line"> </tt>
<a name="crossover"></a><div id="crossover-def"><a name="L619"></a><tt class="py-lineno">619</tt> <a class="py-toggle" href="#" id="crossover-toggle" onclick="return toggle('crossover');">-</a><tt class="py-line"><tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic-module.html#crossover">crossover</a><tt class="py-op">(</tt><tt class="py-param">chromosome1</tt><tt class="py-op">,</tt> <tt class="py-param">chromosome2</tt><tt class="py-op">,</tt> <tt class="py-param">position</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="crossover-collapsed" style="display:none;" pad="+++" indent="++++"></div><div id="crossover-expanded"><a name="L620"></a><tt class="py-lineno">620</tt>  <tt class="py-line">    <tt class="py-docstring">"""</tt> </tt>
<a name="L621"></a><tt class="py-lineno">621</tt>  <tt class="py-line"><tt class="py-docstring">    Cross-over operator - swaps the data on the 2 given chromosomes after </tt> </tt>
<a name="L622"></a><tt class="py-lineno">622</tt>  <tt class="py-line"><tt class="py-docstring">    a given position.</tt> </tt>
<a name="L623"></a><tt class="py-lineno">623</tt>  <tt class="py-line"><tt class="py-docstring">    </tt> </tt>
<a name="L624"></a><tt class="py-lineno">624</tt>  <tt class="py-line"><tt class="py-docstring">    @param chromosome1: Chromosome object</tt> </tt>
<a name="L625"></a><tt class="py-lineno">625</tt>  <tt class="py-line"><tt class="py-docstring">    @param chromosome2: Chromosome object</tt> </tt>
<a name="L626"></a><tt class="py-lineno">626</tt>  <tt class="py-line"><tt class="py-docstring">    @param position: base position of the swap over</tt> </tt>
<a name="L627"></a><tt class="py-lineno">627</tt>  <tt class="py-line"><tt class="py-docstring">    @type position: integer</tt> </tt>
<a name="L628"></a><tt class="py-lineno">628</tt>  <tt class="py-line"><tt class="py-docstring">    @return: (resulting chromosome1, resulting chromosome2)</tt> </tt>
<a name="L629"></a><tt class="py-lineno">629</tt>  <tt class="py-line"><tt class="py-docstring"></tt> </tt>
<a name="L630"></a><tt class="py-lineno">630</tt>  <tt class="py-line"><tt class="py-docstring">    New chromosomes inherit their parent's crossed sequences, bases and</tt> </tt>
<a name="L631"></a><tt class="py-lineno">631</tt>  <tt class="py-line"><tt class="py-docstring">    background mutation rate.</tt> </tt>
<a name="L632"></a><tt class="py-lineno">632</tt>  <tt class="py-line"><tt class="py-docstring">    </tt> </tt>
<a name="L633"></a><tt class="py-lineno">633</tt>  <tt class="py-line"><tt class="py-docstring">    @since: version 0.4</tt> </tt>
<a name="L634"></a><tt class="py-lineno">634</tt>  <tt class="py-line"><tt class="py-docstring">    """</tt> </tt>
<a name="L635"></a><tt class="py-lineno">635</tt>  <tt class="py-line">    <tt class="py-name">seq1</tt> <tt class="py-op">=</tt> <tt class="py-name">chromosome1</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt> </tt>
<a name="L636"></a><tt class="py-lineno">636</tt>  <tt class="py-line">    <tt class="py-name">seq2</tt> <tt class="py-op">=</tt> <tt class="py-name">chromosome2</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt>  </tt>
<a name="L637"></a><tt class="py-lineno">637</tt>  <tt class="py-line">    <tt class="py-name">position</tt> <tt class="py-op">=</tt> <tt class="py-name">int</tt><tt class="py-op">(</tt><tt class="py-name">position</tt><tt class="py-op">)</tt> </tt>
<a name="L638"></a><tt class="py-lineno">638</tt>  <tt class="py-line">    <tt class="py-keyword">if</tt> <tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">seq1</tt><tt class="py-op">)</tt> <tt class="py-op">&gt;</tt> <tt class="py-name">position</tt> <tt class="py-keyword">and</tt> <tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">seq2</tt><tt class="py-op">)</tt> <tt class="py-op">&gt;</tt> <tt class="py-name">position</tt><tt class="py-op">:</tt> </tt>
<a name="L639"></a><tt class="py-lineno">639</tt>  <tt class="py-line">        <tt class="py-name">new1</tt> <tt class="py-op">=</tt> <tt id="link-49" class="py-name"><a title="dose.genetic.Chromosome" class="py-name" href="#" onclick="return doclink('link-49', 'Chromosome', 'link-14');">Chromosome</a></tt><tt class="py-op">(</tt><tt class="py-name">seq1</tt><tt class="py-op">[</tt><tt class="py-op">:</tt><tt class="py-name">position</tt><tt class="py-op">]</tt> <tt class="py-op">+</tt> <tt class="py-name">seq2</tt><tt class="py-op">[</tt><tt class="py-name">position</tt><tt class="py-op">:</tt><tt class="py-op">]</tt><tt class="py-op">,</tt>  </tt>
<a name="L640"></a><tt class="py-lineno">640</tt>  <tt class="py-line">                          <tt class="py-name">chromosome1</tt><tt class="py-op">.</tt><tt class="py-name">base</tt><tt class="py-op">,</tt> <tt class="py-name">chromosome1</tt><tt class="py-op">.</tt><tt class="py-name">background_mutation</tt><tt class="py-op">)</tt> </tt>
<a name="L641"></a><tt class="py-lineno">641</tt>  <tt class="py-line">        <tt class="py-name">new2</tt> <tt class="py-op">=</tt> <tt id="link-50" class="py-name"><a title="dose.genetic.Chromosome" class="py-name" href="#" onclick="return doclink('link-50', 'Chromosome', 'link-14');">Chromosome</a></tt><tt class="py-op">(</tt><tt class="py-name">seq2</tt><tt class="py-op">[</tt><tt class="py-op">:</tt><tt class="py-name">position</tt><tt class="py-op">]</tt> <tt class="py-op">+</tt> <tt class="py-name">seq1</tt><tt class="py-op">[</tt><tt class="py-name">position</tt><tt class="py-op">:</tt><tt class="py-op">]</tt><tt class="py-op">,</tt> </tt>
<a name="L642"></a><tt class="py-lineno">642</tt>  <tt class="py-line">                          <tt class="py-name">chromosome2</tt><tt class="py-op">.</tt><tt class="py-name">base</tt><tt class="py-op">,</tt> <tt class="py-name">chromosome2</tt><tt class="py-op">.</tt><tt class="py-name">background_mutation</tt><tt class="py-op">)</tt> </tt>
<a name="L643"></a><tt class="py-lineno">643</tt>  <tt class="py-line">        <tt class="py-keyword">return</tt> <tt class="py-op">(</tt><tt class="py-name">new1</tt><tt class="py-op">,</tt> <tt class="py-name">new2</tt><tt class="py-op">)</tt> </tt>
<a name="L644"></a><tt class="py-lineno">644</tt>  <tt class="py-line">    <tt class="py-keyword">elif</tt> <tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">seq1</tt><tt class="py-op">)</tt> <tt class="py-op">&gt;</tt> <tt class="py-name">position</tt><tt class="py-op">:</tt> </tt>
<a name="L645"></a><tt class="py-lineno">645</tt>  <tt class="py-line">        <tt class="py-name">new1</tt> <tt class="py-op">=</tt> <tt id="link-51" class="py-name"><a title="dose.genetic.Chromosome" class="py-name" href="#" onclick="return doclink('link-51', 'Chromosome', 'link-14');">Chromosome</a></tt><tt class="py-op">(</tt><tt class="py-name">seq1</tt><tt class="py-op">[</tt><tt class="py-op">:</tt><tt class="py-name">position</tt><tt class="py-op">]</tt><tt class="py-op">,</tt> <tt class="py-name">chromosome1</tt><tt class="py-op">.</tt><tt class="py-name">base</tt><tt class="py-op">,</tt>  </tt>
<a name="L646"></a><tt class="py-lineno">646</tt>  <tt class="py-line">                          <tt class="py-name">chromosome1</tt><tt class="py-op">.</tt><tt class="py-name">background_mutation</tt><tt class="py-op">)</tt> </tt>
<a name="L647"></a><tt class="py-lineno">647</tt>  <tt class="py-line">        <tt class="py-name">new2</tt> <tt class="py-op">=</tt> <tt id="link-52" class="py-name"><a title="dose.genetic.Chromosome" class="py-name" href="#" onclick="return doclink('link-52', 'Chromosome', 'link-14');">Chromosome</a></tt><tt class="py-op">(</tt><tt class="py-name">chromosome2</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt> <tt class="py-op">+</tt> <tt class="py-name">seq1</tt><tt class="py-op">[</tt><tt class="py-name">position</tt><tt class="py-op">:</tt><tt class="py-op">]</tt><tt class="py-op">,</tt> </tt>
<a name="L648"></a><tt class="py-lineno">648</tt>  <tt class="py-line">                          <tt class="py-name">chromosome2</tt><tt class="py-op">.</tt><tt class="py-name">base</tt><tt class="py-op">,</tt> <tt class="py-name">chromosome2</tt><tt class="py-op">.</tt><tt class="py-name">background_mutation</tt><tt class="py-op">)</tt> </tt>
<a name="L649"></a><tt class="py-lineno">649</tt>  <tt class="py-line">        <tt class="py-keyword">return</tt> <tt class="py-op">(</tt><tt class="py-name">new1</tt><tt class="py-op">,</tt> <tt class="py-name">new2</tt><tt class="py-op">)</tt> </tt>
<a name="L650"></a><tt class="py-lineno">650</tt>  <tt class="py-line">    <tt class="py-keyword">elif</tt> <tt class="py-name">len</tt><tt class="py-op">(</tt><tt class="py-name">seq2</tt><tt class="py-op">)</tt> <tt class="py-op">&gt;</tt> <tt class="py-name">position</tt><tt class="py-op">:</tt> </tt>
<a name="L651"></a><tt class="py-lineno">651</tt>  <tt class="py-line">        <tt class="py-name">new1</tt> <tt class="py-op">=</tt> <tt id="link-53" class="py-name"><a title="dose.genetic.Chromosome" class="py-name" href="#" onclick="return doclink('link-53', 'Chromosome', 'link-14');">Chromosome</a></tt><tt class="py-op">(</tt><tt class="py-name">chromosome1</tt><tt class="py-op">.</tt><tt class="py-name">sequence</tt> <tt class="py-op">+</tt> <tt class="py-name">seq2</tt><tt class="py-op">[</tt><tt class="py-name">position</tt><tt class="py-op">:</tt><tt class="py-op">]</tt><tt class="py-op">,</tt> </tt>
<a name="L652"></a><tt class="py-lineno">652</tt>  <tt class="py-line">                          <tt class="py-name">chromosome1</tt><tt class="py-op">.</tt><tt class="py-name">base</tt><tt class="py-op">,</tt> <tt class="py-name">chromosome1</tt><tt class="py-op">.</tt><tt class="py-name">background_mutation</tt><tt class="py-op">)</tt> </tt>
<a name="L653"></a><tt class="py-lineno">653</tt>  <tt class="py-line">        <tt class="py-name">new2</tt><tt class="py-op">=</tt> <tt id="link-54" class="py-name"><a title="dose.genetic.Chromosome" class="py-name" href="#" onclick="return doclink('link-54', 'Chromosome', 'link-14');">Chromosome</a></tt><tt class="py-op">(</tt><tt class="py-name">seq2</tt><tt class="py-op">[</tt><tt class="py-op">:</tt><tt class="py-name">position</tt><tt class="py-op">]</tt><tt class="py-op">,</tt> <tt class="py-name">chromosome2</tt><tt class="py-op">.</tt><tt class="py-name">base</tt><tt class="py-op">,</tt>  </tt>
<a name="L654"></a><tt class="py-lineno">654</tt>  <tt class="py-line">                         <tt class="py-name">chromosome2</tt><tt class="py-op">.</tt><tt class="py-name">background_mutation</tt><tt class="py-op">)</tt> </tt>
<a name="L655"></a><tt class="py-lineno">655</tt>  <tt class="py-line">        <tt class="py-keyword">return</tt> <tt class="py-op">(</tt><tt class="py-name">new1</tt><tt class="py-op">,</tt> <tt class="py-name">new2</tt><tt class="py-op">)</tt> </tt>
<a name="L656"></a><tt class="py-lineno">656</tt>  <tt class="py-line">    <tt class="py-keyword">else</tt><tt class="py-op">:</tt> </tt>
<a name="L657"></a><tt class="py-lineno">657</tt>  <tt class="py-line">        <tt class="py-keyword">return</tt> <tt class="py-op">(</tt><tt class="py-name">chromosome1</tt><tt class="py-op">,</tt> <tt class="py-name">chromosome2</tt><tt class="py-op">)</tt>         </tt>
</div><a name="L658"></a><tt class="py-lineno">658</tt>  <tt class="py-line">             </tt>
<a name="L659"></a><tt class="py-lineno">659</tt>  <tt class="py-line"><tt id="link-55" class="py-name" targets="Variable dose.genetic.population_data=dose.genetic-module.html#population_data"><a title="dose.genetic.population_data" class="py-name" href="#" onclick="return doclink('link-55', 'population_data', 'link-55');">population_data</a></tt> <tt class="py-op">=</tt> \ </tt>
<a name="L660"></a><tt class="py-lineno">660</tt>  <tt class="py-line"><tt class="py-op">{</tt> </tt>
<a name="L661"></a><tt class="py-lineno">661</tt>  <tt class="py-line">    <tt class="py-string">'chromosome_bases'</tt> <tt class="py-op">:</tt> <tt class="py-op">[</tt><tt class="py-number">1</tt><tt class="py-op">,</tt> <tt class="py-number">2</tt><tt class="py-op">,</tt> <tt class="py-number">3</tt><tt class="py-op">,</tt> <tt class="py-number">4</tt><tt class="py-op">]</tt><tt class="py-op">,</tt> </tt>
<a name="L662"></a><tt class="py-lineno">662</tt>  <tt class="py-line">    <tt class="py-string">'chromosome_length'</tt> <tt class="py-op">:</tt> <tt class="py-number">200</tt><tt class="py-op">,</tt> </tt>
<a name="L663"></a><tt class="py-lineno">663</tt>  <tt class="py-line">    <tt class="py-string">'chromosome_type'</tt> <tt class="py-op">:</tt> <tt class="py-string">'defined'</tt><tt class="py-op">,</tt> </tt>
<a name="L664"></a><tt class="py-lineno">664</tt>  <tt class="py-line">    <tt class="py-string">'initial_chromosome'</tt> <tt class="py-op">:</tt> <tt class="py-op">[</tt><tt class="py-number">1</tt><tt class="py-op">]</tt> <tt class="py-op">*</tt> <tt class="py-number">200</tt><tt class="py-op">,</tt> </tt>
<a name="L665"></a><tt class="py-lineno">665</tt>  <tt class="py-line">    <tt class="py-string">'background_mutation'</tt> <tt class="py-op">:</tt> <tt class="py-number">0.0001</tt><tt class="py-op">,</tt> </tt>
<a name="L666"></a><tt class="py-lineno">666</tt>  <tt class="py-line">    <tt class="py-string">'genome_size'</tt> <tt class="py-op">:</tt> <tt class="py-number">1</tt><tt class="py-op">,</tt> </tt>
<a name="L667"></a><tt class="py-lineno">667</tt>  <tt class="py-line">    <tt class="py-string">'population_size'</tt> <tt class="py-op">:</tt> <tt class="py-number">200</tt><tt class="py-op">,</tt> </tt>
<a name="L668"></a><tt class="py-lineno">668</tt>  <tt class="py-line">    <tt class="py-string">'fitness_function'</tt> <tt class="py-op">:</tt> <tt class="py-string">'default'</tt><tt class="py-op">,</tt> </tt>
<a name="L669"></a><tt class="py-lineno">669</tt>  <tt class="py-line">    <tt class="py-string">'mutation_scheme'</tt> <tt class="py-op">:</tt> <tt class="py-string">'default'</tt><tt class="py-op">,</tt> </tt>
<a name="L670"></a><tt class="py-lineno">670</tt>  <tt class="py-line">    <tt class="py-string">'additional_mutation'</tt> <tt class="py-op">:</tt> <tt class="py-number">0.01</tt><tt class="py-op">,</tt> </tt>
<a name="L671"></a><tt class="py-lineno">671</tt>  <tt class="py-line">    <tt class="py-string">'mutation_type'</tt> <tt class="py-op">:</tt> <tt class="py-string">'point'</tt><tt class="py-op">,</tt> </tt>
<a name="L672"></a><tt class="py-lineno">672</tt>  <tt class="py-line">    <tt class="py-string">'goal'</tt> <tt class="py-op">:</tt> <tt class="py-number">4</tt><tt class="py-op">,</tt> </tt>
<a name="L673"></a><tt class="py-lineno">673</tt>  <tt class="py-line">    <tt class="py-string">'maximum_generations'</tt> <tt class="py-op">:</tt> <tt class="py-string">'infinite'</tt><tt class="py-op">,</tt> </tt>
<a name="L674"></a><tt class="py-lineno">674</tt>  <tt class="py-line">    <tt class="py-string">'prepopulation_control'</tt> <tt class="py-op">:</tt> <tt class="py-string">'default'</tt><tt class="py-op">,</tt> </tt>
<a name="L675"></a><tt class="py-lineno">675</tt>  <tt class="py-line">    <tt class="py-string">'mating'</tt> <tt class="py-op">:</tt> <tt class="py-string">'default'</tt><tt class="py-op">,</tt> </tt>
<a name="L676"></a><tt class="py-lineno">676</tt>  <tt class="py-line">    <tt class="py-string">'postpopulation_control'</tt> <tt class="py-op">:</tt> <tt class="py-string">'default'</tt><tt class="py-op">,</tt> </tt>
<a name="L677"></a><tt class="py-lineno">677</tt>  <tt class="py-line">    <tt class="py-string">'generation_events'</tt> <tt class="py-op">:</tt> <tt class="py-string">'default'</tt><tt class="py-op">,</tt> </tt>
<a name="L678"></a><tt class="py-lineno">678</tt>  <tt class="py-line">    <tt class="py-string">'population_report'</tt> <tt class="py-op">:</tt> <tt class="py-string">'default'</tt> </tt>
<a name="L679"></a><tt class="py-lineno">679</tt>  <tt class="py-line"><tt class="py-op">}</tt> </tt>
<a name="L680"></a><tt class="py-lineno">680</tt>  <tt class="py-line"> </tt>
<a name="population_constructor"></a><div id="population_constructor-def"><a name="L681"></a><tt class="py-lineno">681</tt> <a class="py-toggle" href="#" id="population_constructor-toggle" onclick="return toggle('population_constructor');">-</a><tt class="py-line"><tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic-module.html#population_constructor">population_constructor</a><tt class="py-op">(</tt><tt class="py-param">data</tt><tt class="py-op">=</tt><tt id="link-56" class="py-name"><a title="dose.genetic.population_data" class="py-name" href="#" onclick="return doclink('link-56', 'population_data', 'link-55');">population_data</a></tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="population_constructor-collapsed" style="display:none;" pad="+++" indent="++++"></div><div id="population_constructor-expanded"><a name="L682"></a><tt class="py-lineno">682</tt>  <tt class="py-line">    <tt class="py-docstring">"""</tt> </tt>
<a name="L683"></a><tt class="py-lineno">683</tt>  <tt class="py-line"><tt class="py-docstring">    Function to construct a population based on a dictionary of population </tt> </tt>
<a name="L684"></a><tt class="py-lineno">684</tt>  <tt class="py-line"><tt class="py-docstring">    data.</tt> </tt>
<a name="L685"></a><tt class="py-lineno">685</tt>  <tt class="py-line"><tt class="py-docstring">    </tt> </tt>
<a name="L686"></a><tt class="py-lineno">686</tt>  <tt class="py-line"><tt class="py-docstring">    Population data contains the following keys:</tt> </tt>
<a name="L687"></a><tt class="py-lineno">687</tt>  <tt class="py-line"><tt class="py-docstring">        - 'chromosome_bases' = List of allowable bases. </tt> </tt>
<a name="L688"></a><tt class="py-lineno">688</tt>  <tt class="py-line"><tt class="py-docstring">            Default = [1, 2, 3, 4].</tt> </tt>
<a name="L689"></a><tt class="py-lineno">689</tt>  <tt class="py-line"><tt class="py-docstring">        - 'chromosome_length' = Length of a chromosome. Default = 200.</tt> </tt>
<a name="L690"></a><tt class="py-lineno">690</tt>  <tt class="py-line"><tt class="py-docstring">        - 'chromosome_type' = Type of chromosome. Default = 'defined'.</tt> </tt>
<a name="L691"></a><tt class="py-lineno">691</tt>  <tt class="py-line"><tt class="py-docstring">        - 'initial_chromosome' = Initial chromosome. Default = [1] * 200.</tt> </tt>
<a name="L692"></a><tt class="py-lineno">692</tt>  <tt class="py-line"><tt class="py-docstring">        - 'background_mutation' = Background mutation rate. </tt> </tt>
<a name="L693"></a><tt class="py-lineno">693</tt>  <tt class="py-line"><tt class="py-docstring">            Default = 0.0001 (0.01%).</tt> </tt>
<a name="L694"></a><tt class="py-lineno">694</tt>  <tt class="py-line"><tt class="py-docstring">        - 'genome_size' = Number of chromosomes per organism. Default = 1.</tt> </tt>
<a name="L695"></a><tt class="py-lineno">695</tt>  <tt class="py-line"><tt class="py-docstring">        - 'population_size' = Size of initial population (number of organisms).</tt> </tt>
<a name="L696"></a><tt class="py-lineno">696</tt>  <tt class="py-line"><tt class="py-docstring">            Default = 200.</tt> </tt>
<a name="L697"></a><tt class="py-lineno">697</tt>  <tt class="py-line"><tt class="py-docstring">        - 'fitness_function' = Fitness evaluation function. Accepts 'default' </tt> </tt>
<a name="L698"></a><tt class="py-lineno">698</tt>  <tt class="py-line"><tt class="py-docstring">            or a function. Please refer to Organism. Default = 'default'.</tt> </tt>
<a name="L699"></a><tt class="py-lineno">699</tt>  <tt class="py-line"><tt class="py-docstring">        - 'mutation_scheme' = Function simulating the mutation scheme of an </tt> </tt>
<a name="L700"></a><tt class="py-lineno">700</tt>  <tt class="py-line"><tt class="py-docstring">            organism. Accepts 'default' or a function. Please refer to </tt> </tt>
<a name="L701"></a><tt class="py-lineno">701</tt>  <tt class="py-line"><tt class="py-docstring">            Organism. Default = 'default'.</tt> </tt>
<a name="L702"></a><tt class="py-lineno">702</tt>  <tt class="py-line"><tt class="py-docstring">        - 'additional_mutation' = Mutation rate on top of background </tt> </tt>
<a name="L703"></a><tt class="py-lineno">703</tt>  <tt class="py-line"><tt class="py-docstring">            mutation rate. Default = 0.01 (1%).</tt> </tt>
<a name="L704"></a><tt class="py-lineno">704</tt>  <tt class="py-line"><tt class="py-docstring">        - 'mutation_type' = Type of default mutation. Default = 'point'.</tt> </tt>
<a name="L705"></a><tt class="py-lineno">705</tt>  <tt class="py-line"><tt class="py-docstring">        - 'goal' = Goal of the population, as evaluated by fitness function.</tt> </tt>
<a name="L706"></a><tt class="py-lineno">706</tt>  <tt class="py-line"><tt class="py-docstring">            Default = 4.</tt> </tt>
<a name="L707"></a><tt class="py-lineno">707</tt>  <tt class="py-line"><tt class="py-docstring">        - 'maximum_generations' = Number of generations to simulate. Accepts </tt> </tt>
<a name="L708"></a><tt class="py-lineno">708</tt>  <tt class="py-line"><tt class="py-docstring">            an integer or 'infinite'. Default = 'infinite'.</tt> </tt>
<a name="L709"></a><tt class="py-lineno">709</tt>  <tt class="py-line"><tt class="py-docstring">        - 'prepopulation_control' = Function simulating pre-mating population </tt> </tt>
<a name="L710"></a><tt class="py-lineno">710</tt>  <tt class="py-line"><tt class="py-docstring">            control. Please refer to Population. Default = 'default'.</tt> </tt>
<a name="L711"></a><tt class="py-lineno">711</tt>  <tt class="py-line"><tt class="py-docstring">        - 'mating' = Function simulating mating procedure (mate selection </tt> </tt>
<a name="L712"></a><tt class="py-lineno">712</tt>  <tt class="py-line"><tt class="py-docstring">            and act of mating control. Please refer to Population. </tt> </tt>
<a name="L713"></a><tt class="py-lineno">713</tt>  <tt class="py-line"><tt class="py-docstring">            Default = 'default'.</tt> </tt>
<a name="L714"></a><tt class="py-lineno">714</tt>  <tt class="py-line"><tt class="py-docstring">        - 'postpopulation_control' = Function simulating post-mating population </tt> </tt>
<a name="L715"></a><tt class="py-lineno">715</tt>  <tt class="py-line"><tt class="py-docstring">            control. Please refer to Population. Default = 'default'.</tt> </tt>
<a name="L716"></a><tt class="py-lineno">716</tt>  <tt class="py-line"><tt class="py-docstring">        - 'generation_events' = Function simulating other possible (usually </tt> </tt>
<a name="L717"></a><tt class="py-lineno">717</tt>  <tt class="py-line"><tt class="py-docstring">            rare or random) events in the generation. Please refer to </tt> </tt>
<a name="L718"></a><tt class="py-lineno">718</tt>  <tt class="py-line"><tt class="py-docstring">            Population. Default = 'default'.</tt> </tt>
<a name="L719"></a><tt class="py-lineno">719</tt>  <tt class="py-line"><tt class="py-docstring">        - 'population_report' = Function to generate the status report of </tt> </tt>
<a name="L720"></a><tt class="py-lineno">720</tt>  <tt class="py-line"><tt class="py-docstring">            the generation. Please refer to Population. Default = 'default'.</tt> </tt>
<a name="L721"></a><tt class="py-lineno">721</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L722"></a><tt class="py-lineno">722</tt>  <tt class="py-line"><tt class="py-docstring">    @param data: population data</tt> </tt>
<a name="L723"></a><tt class="py-lineno">723</tt>  <tt class="py-line"><tt class="py-docstring">    @type data: dictionary</tt> </tt>
<a name="L724"></a><tt class="py-lineno">724</tt>  <tt class="py-line"><tt class="py-docstring">        </tt> </tt>
<a name="L725"></a><tt class="py-lineno">725</tt>  <tt class="py-line"><tt class="py-docstring">    @return: Population object</tt> </tt>
<a name="L726"></a><tt class="py-lineno">726</tt>  <tt class="py-line"><tt class="py-docstring"></tt> </tt>
<a name="L727"></a><tt class="py-lineno">727</tt>  <tt class="py-line"><tt class="py-docstring">    @see: Lim, JZR, Aw, ZQ, Goh, DJW, How, JA, Low, SXZ, Loo, BZL,</tt> </tt>
<a name="L728"></a><tt class="py-lineno">728</tt>  <tt class="py-line"><tt class="py-docstring">    Ling, MHT. 2010. A genetic algorithm framework grounded in biology.</tt> </tt>
<a name="L729"></a><tt class="py-lineno">729</tt>  <tt class="py-line"><tt class="py-docstring">    The Python Papers Source Codes 2: 6.</tt> </tt>
<a name="L730"></a><tt class="py-lineno">730</tt>  <tt class="py-line"><tt class="py-docstring">    </tt> </tt>
<a name="L731"></a><tt class="py-lineno">731</tt>  <tt class="py-line"><tt class="py-docstring">    @since: version 0.4</tt> </tt>
<a name="L732"></a><tt class="py-lineno">732</tt>  <tt class="py-line"><tt class="py-docstring">    """</tt> </tt>
<a name="L733"></a><tt class="py-lineno">733</tt>  <tt class="py-line">    <tt class="py-name">chr</tt> <tt class="py-op">=</tt> <tt id="link-57" class="py-name"><a title="dose.genetic.Chromosome" class="py-name" href="#" onclick="return doclink('link-57', 'Chromosome', 'link-14');">Chromosome</a></tt><tt class="py-op">(</tt><tt id="link-58" class="py-name" targets="Variable dose.copads.samplestatistics.SingleSample.data=dose.copads.samplestatistics.SingleSample-class.html#data"><a title="dose.copads.samplestatistics.SingleSample.data" class="py-name" href="#" onclick="return doclink('link-58', 'data', 'link-58');">data</a></tt><tt class="py-op">[</tt><tt class="py-string">'initial_chromosome'</tt><tt class="py-op">]</tt><tt class="py-op">,</tt>  </tt>
<a name="L734"></a><tt class="py-lineno">734</tt>  <tt class="py-line">                     <tt id="link-59" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.data" class="py-name" href="#" onclick="return doclink('link-59', 'data', 'link-58');">data</a></tt><tt class="py-op">[</tt><tt class="py-string">'chromosome_bases'</tt><tt class="py-op">]</tt><tt class="py-op">,</tt> </tt>
<a name="L735"></a><tt class="py-lineno">735</tt>  <tt class="py-line">                     <tt id="link-60" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.data" class="py-name" href="#" onclick="return doclink('link-60', 'data', 'link-58');">data</a></tt><tt class="py-op">[</tt><tt class="py-string">'background_mutation'</tt><tt class="py-op">]</tt><tt class="py-op">)</tt> </tt>
<a name="L736"></a><tt class="py-lineno">736</tt>  <tt class="py-line">    <tt class="py-name">org</tt> <tt class="py-op">=</tt> <tt id="link-61" class="py-name"><a title="dose.genetic.Organism" class="py-name" href="#" onclick="return doclink('link-61', 'Organism', 'link-31');">Organism</a></tt><tt class="py-op">(</tt><tt class="py-op">[</tt><tt class="py-name">chr</tt><tt class="py-op">]</tt><tt class="py-op">*</tt><tt id="link-62" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.data" class="py-name" href="#" onclick="return doclink('link-62', 'data', 'link-58');">data</a></tt><tt class="py-op">[</tt><tt class="py-string">'genome_size'</tt><tt class="py-op">]</tt><tt class="py-op">,</tt> </tt>
<a name="L737"></a><tt class="py-lineno">737</tt>  <tt class="py-line">                   <tt id="link-63" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.data" class="py-name" href="#" onclick="return doclink('link-63', 'data', 'link-58');">data</a></tt><tt class="py-op">[</tt><tt class="py-string">'mutation_type'</tt><tt class="py-op">]</tt><tt class="py-op">,</tt> </tt>
<a name="L738"></a><tt class="py-lineno">738</tt>  <tt class="py-line">                   <tt id="link-64" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.data" class="py-name" href="#" onclick="return doclink('link-64', 'data', 'link-58');">data</a></tt><tt class="py-op">[</tt><tt class="py-string">'additional_mutation'</tt><tt class="py-op">]</tt><tt class="py-op">)</tt> </tt>
<a name="L739"></a><tt class="py-lineno">739</tt>  <tt class="py-line">    <tt class="py-name">org_set</tt> <tt class="py-op">=</tt> <tt class="py-op">[</tt><tt class="py-name">org</tt><tt class="py-op">.</tt><tt id="link-65" class="py-name" targets="Method dose.genetic.Organism.clone()=dose.genetic.Organism-class.html#clone"><a title="dose.genetic.Organism.clone" class="py-name" href="#" onclick="return doclink('link-65', 'clone', 'link-65');">clone</a></tt><tt class="py-op">(</tt><tt class="py-op">)</tt> <tt class="py-keyword">for</tt> <tt class="py-name">x</tt> <tt class="py-keyword">in</tt> <tt id="link-66" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.range" class="py-name" href="#" onclick="return doclink('link-66', 'range', 'link-4');">range</a></tt><tt class="py-op">(</tt><tt id="link-67" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.data" class="py-name" href="#" onclick="return doclink('link-67', 'data', 'link-58');">data</a></tt><tt class="py-op">[</tt><tt class="py-string">'population_size'</tt><tt class="py-op">]</tt><tt class="py-op">)</tt><tt class="py-op">]</tt> </tt>
<a name="L740"></a><tt class="py-lineno">740</tt>  <tt class="py-line">    <tt class="py-name">pop</tt> <tt class="py-op">=</tt> <tt id="link-68" class="py-name" targets="Class dose.genetic.Population=dose.genetic.Population-class.html"><a title="dose.genetic.Population" class="py-name" href="#" onclick="return doclink('link-68', 'Population', 'link-68');">Population</a></tt><tt class="py-op">(</tt><tt id="link-69" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.data" class="py-name" href="#" onclick="return doclink('link-69', 'data', 'link-58');">data</a></tt><tt class="py-op">[</tt><tt class="py-string">'goal'</tt><tt class="py-op">]</tt><tt class="py-op">,</tt>  </tt>
<a name="L741"></a><tt class="py-lineno">741</tt>  <tt class="py-line">                     <tt class="py-name">int</tt><tt class="py-op">(</tt><tt id="link-70" class="py-name"><a title="dose.copads.samplestatistics.SingleSample.data" class="py-name" href="#" onclick="return doclink('link-70', 'data', 'link-58');">data</a></tt><tt class="py-op">[</tt><tt class="py-string">'maximum_generations'</tt><tt class="py-op">]</tt><tt class="py-op">)</tt><tt class="py-op">,</tt>  </tt>
<a name="L742"></a><tt class="py-lineno">742</tt>  <tt class="py-line">                     <tt class="py-name">org_set</tt><tt class="py-op">)</tt> </tt>
<a name="L743"></a><tt class="py-lineno">743</tt>  <tt class="py-line">    <tt class="py-keyword">return</tt> <tt class="py-name">pop</tt> </tt>
</div><a name="L744"></a><tt class="py-lineno">744</tt>  <tt class="py-line">     </tt>
<a name="population_simulate"></a><div id="population_simulate-def"><a name="L745"></a><tt class="py-lineno">745</tt> <a class="py-toggle" href="#" id="population_simulate-toggle" onclick="return toggle('population_simulate');">-</a><tt class="py-line"><tt class="py-keyword">def</tt> <a class="py-def-name" href="dose.genetic-module.html#population_simulate">population_simulate</a><tt class="py-op">(</tt><tt class="py-param">population</tt><tt class="py-op">,</tt>  </tt>
<a name="L746"></a><tt class="py-lineno">746</tt>  <tt class="py-line">                        <tt class="py-param">printfreq</tt><tt class="py-op">=</tt><tt class="py-number">100</tt><tt class="py-op">,</tt>  </tt>
<a name="L747"></a><tt class="py-lineno">747</tt>  <tt class="py-line">                        <tt class="py-param">freezefreq</tt><tt class="py-op">=</tt><tt class="py-string">'never'</tt><tt class="py-op">,</tt>  </tt>
<a name="L748"></a><tt class="py-lineno">748</tt>  <tt class="py-line">                        <tt class="py-param">freezefile</tt><tt class="py-op">=</tt><tt class="py-string">'pop'</tt><tt class="py-op">,</tt> </tt>
<a name="L749"></a><tt class="py-lineno">749</tt>  <tt class="py-line">                        <tt class="py-param">freezeproportion</tt><tt class="py-op">=</tt><tt class="py-number">0.01</tt><tt class="py-op">,</tt>  </tt>
<a name="L750"></a><tt class="py-lineno">750</tt>  <tt class="py-line">                        <tt class="py-param">resultfile</tt><tt class="py-op">=</tt><tt class="py-string">'result.txt'</tt><tt class="py-op">)</tt><tt class="py-op">:</tt> </tt>
</div><div id="population_simulate-collapsed" style="display:none;" pad="+++" indent="++++"></div><div id="population_simulate-expanded"><a name="L751"></a><tt class="py-lineno">751</tt>  <tt class="py-line">    <tt class="py-docstring">"""</tt> </tt>
<a name="L752"></a><tt class="py-lineno">752</tt>  <tt class="py-line"><tt class="py-docstring">    Function to simulate the population - start the GA.</tt> </tt>
<a name="L753"></a><tt class="py-lineno">753</tt>  <tt class="py-line"><tt class="py-docstring">    </tt> </tt>
<a name="L754"></a><tt class="py-lineno">754</tt>  <tt class="py-line"><tt class="py-docstring">    @param population: Population to run.</tt> </tt>
<a name="L755"></a><tt class="py-lineno">755</tt>  <tt class="py-line"><tt class="py-docstring">    @type population: Population object</tt> </tt>
<a name="L756"></a><tt class="py-lineno">756</tt>  <tt class="py-line"><tt class="py-docstring">    @param printfreq: Reporting intervals on screen. Default = 100.</tt> </tt>
<a name="L757"></a><tt class="py-lineno">757</tt>  <tt class="py-line"><tt class="py-docstring">    @param freezefreq: Generation intervals to freeze population into a file.</tt> </tt>
<a name="L758"></a><tt class="py-lineno">758</tt>  <tt class="py-line"><tt class="py-docstring">        See Population.freeze method. Accepts an integer or 'never'. </tt> </tt>
<a name="L759"></a><tt class="py-lineno">759</tt>  <tt class="py-line"><tt class="py-docstring">        Default = 'never'.</tt> </tt>
<a name="L760"></a><tt class="py-lineno">760</tt>  <tt class="py-line"><tt class="py-docstring">    @param freezefile: Prefix of file name of frozen population. </tt> </tt>
<a name="L761"></a><tt class="py-lineno">761</tt>  <tt class="py-line"><tt class="py-docstring">        See Population.freeze method. Default = 'pop'.</tt> </tt>
<a name="L762"></a><tt class="py-lineno">762</tt>  <tt class="py-line"><tt class="py-docstring">    @param freezeproportion: Proportion of population to be preserved.</tt> </tt>
<a name="L763"></a><tt class="py-lineno">763</tt>  <tt class="py-line"><tt class="py-docstring">        See Population.freeze method. Default = 0.01, preserves 1% of the </tt> </tt>
<a name="L764"></a><tt class="py-lineno">764</tt>  <tt class="py-line"><tt class="py-docstring">        population.</tt> </tt>
<a name="L765"></a><tt class="py-lineno">765</tt>  <tt class="py-line"><tt class="py-docstring">    @param resultfile: Name of file to print out results of each generation. </tt> </tt>
<a name="L766"></a><tt class="py-lineno">766</tt>  <tt class="py-line"><tt class="py-docstring">        Format of output is dependent on reporting method of the population</tt> </tt>
<a name="L767"></a><tt class="py-lineno">767</tt>  <tt class="py-line"><tt class="py-docstring">        (Population.report). If 'None', it will print out the results into</tt> </tt>
<a name="L768"></a><tt class="py-lineno">768</tt>  <tt class="py-line"><tt class="py-docstring">        a file. Default = 'result.txt'.</tt> </tt>
<a name="L769"></a><tt class="py-lineno">769</tt>  <tt class="py-line"><tt class="py-docstring"></tt> </tt>
<a name="L770"></a><tt class="py-lineno">770</tt>  <tt class="py-line"><tt class="py-docstring">    @see: Lim, JZR, Aw, ZQ, Goh, DJW, How, JA, Low, SXZ, Loo, BZL,</tt> </tt>
<a name="L771"></a><tt class="py-lineno">771</tt>  <tt class="py-line"><tt class="py-docstring">    Ling, MHT. 2010. A genetic algorithm framework grounded in biology.</tt> </tt>
<a name="L772"></a><tt class="py-lineno">772</tt>  <tt class="py-line"><tt class="py-docstring">    The Python Papers Source Codes 2: 6.</tt> </tt>
<a name="L773"></a><tt class="py-lineno">773</tt>  <tt class="py-line"><tt class="py-docstring">    </tt> </tt>
<a name="L774"></a><tt class="py-lineno">774</tt>  <tt class="py-line"><tt class="py-docstring">    @since: version 0.4</tt> </tt>
<a name="L775"></a><tt class="py-lineno">775</tt>  <tt class="py-line"><tt class="py-docstring">    """</tt> </tt>
<a name="L776"></a><tt class="py-lineno">776</tt>  <tt class="py-line">    <tt class="py-keyword">if</tt> <tt class="py-name">freezefreq</tt> <tt class="py-op">==</tt> <tt class="py-string">'never'</tt><tt class="py-op">:</tt> <tt class="py-name">freezefreq</tt> <tt class="py-op">=</tt> <tt class="py-name">int</tt><tt class="py-op">(</tt><tt class="py-number">12e14</tt><tt class="py-op">)</tt> </tt>
<a name="L777"></a><tt class="py-lineno">777</tt>  <tt class="py-line">    <tt class="py-keyword">if</tt> <tt class="py-name">resultfile</tt> <tt class="py-op">!=</tt> <tt class="py-string">'None'</tt><tt class="py-op">:</tt> <tt class="py-name">result</tt> <tt class="py-op">=</tt> <tt class="py-name">open</tt><tt class="py-op">(</tt><tt class="py-name">resultfile</tt><tt class="py-op">,</tt> <tt class="py-string">'w'</tt><tt class="py-op">)</tt> </tt>
<a name="L778"></a><tt class="py-lineno">778</tt>  <tt class="py-line">    <tt id="link-71" class="py-name"><a title="dose.dose.dose_functions.report
dose.genetic.Population.report" class="py-name" href="#" onclick="return doclink('link-71', 'report', 'link-40');">report</a></tt> <tt class="py-op">=</tt> <tt class="py-name">population</tt><tt class="py-op">.</tt><tt id="link-72" class="py-name" targets="Method dose.genetic.Population.generation_step()=dose.genetic.Population-class.html#generation_step"><a title="dose.genetic.Population.generation_step" class="py-name" href="#" onclick="return doclink('link-72', 'generation_step', 'link-72');">generation_step</a></tt><tt class="py-op">(</tt><tt class="py-op">)</tt> </tt>
<a name="L779"></a><tt class="py-lineno">779</tt>  <tt class="py-line">    <tt class="py-name">reportitems</tt> <tt class="py-op">=</tt> <tt class="py-op">[</tt><tt class="py-string">'|'</tt><tt class="py-op">.</tt><tt class="py-name">join</tt><tt class="py-op">(</tt><tt class="py-op">[</tt><tt class="py-name">str</tt><tt class="py-op">(</tt><tt class="py-name">key</tt><tt class="py-op">)</tt><tt class="py-op">,</tt> <tt class="py-name">str</tt><tt class="py-op">(</tt><tt id="link-73" class="py-name"><a title="dose.dose.dose_functions.report
dose.genetic.Population.report" class="py-name" href="#" onclick="return doclink('link-73', 'report', 'link-40');">report</a></tt><tt class="py-op">[</tt><tt class="py-name">key</tt><tt class="py-op">]</tt><tt class="py-op">)</tt><tt class="py-op">]</tt><tt class="py-op">)</tt> </tt>
<a name="L780"></a><tt class="py-lineno">780</tt>  <tt class="py-line">                    <tt class="py-keyword">for</tt> <tt class="py-name">key</tt> <tt class="py-keyword">in</tt> <tt class="py-name">list</tt><tt class="py-op">(</tt><tt id="link-74" class="py-name"><a title="dose.dose.dose_functions.report
dose.genetic.Population.report" class="py-name" href="#" onclick="return doclink('link-74', 'report', 'link-40');">report</a></tt><tt class="py-op">.</tt><tt class="py-name">keys</tt><tt class="py-op">(</tt><tt class="py-op">)</tt><tt class="py-op">)</tt><tt class="py-op">]</tt> </tt>
<a name="L781"></a><tt class="py-lineno">781</tt>  <tt class="py-line">    <tt class="py-name">result</tt><tt class="py-op">.</tt><tt class="py-name">writelines</tt><tt class="py-op">(</tt><tt class="py-string">'|'</tt><tt class="py-op">.</tt><tt class="py-name">join</tt><tt class="py-op">(</tt><tt class="py-name">reportitems</tt><tt class="py-op">)</tt><tt class="py-op">)</tt> </tt>
<a name="L782"></a><tt class="py-lineno">782</tt>  <tt class="py-line">    <tt class="py-name">result</tt><tt class="py-op">.</tt><tt class="py-name">writelines</tt><tt class="py-op">(</tt><tt class="py-name">os</tt><tt class="py-op">.</tt><tt class="py-name">linesep</tt><tt class="py-op">)</tt> </tt>
<a name="L783"></a><tt class="py-lineno">783</tt>  <tt class="py-line">    <tt class="py-keyword">while</tt> <tt class="py-name">population</tt><tt class="py-op">.</tt><tt class="py-name">generation</tt> <tt class="py-op">&lt;</tt> <tt class="py-name">population</tt><tt class="py-op">.</tt><tt class="py-name">maxgenerations</tt><tt class="py-op">:</tt> </tt>
<a name="L784"></a><tt class="py-lineno">784</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">resultfile</tt> <tt class="py-op">!=</tt> <tt class="py-string">'None'</tt><tt class="py-op">:</tt>  </tt>
<a name="L785"></a><tt class="py-lineno">785</tt>  <tt class="py-line">            <tt id="link-75" class="py-name"><a title="dose.dose.dose_functions.report
dose.genetic.Population.report" class="py-name" href="#" onclick="return doclink('link-75', 'report', 'link-40');">report</a></tt> <tt class="py-op">=</tt> <tt class="py-name">population</tt><tt class="py-op">.</tt><tt id="link-76" class="py-name"><a title="dose.genetic.Population.generation_step" class="py-name" href="#" onclick="return doclink('link-76', 'generation_step', 'link-72');">generation_step</a></tt><tt class="py-op">(</tt><tt class="py-op">)</tt> </tt>
<a name="L786"></a><tt class="py-lineno">786</tt>  <tt class="py-line">            <tt class="py-name">reportitems</tt> <tt class="py-op">=</tt> <tt class="py-op">[</tt><tt class="py-string">'|'</tt><tt class="py-op">.</tt><tt class="py-name">join</tt><tt class="py-op">(</tt><tt class="py-op">[</tt><tt class="py-name">str</tt><tt class="py-op">(</tt><tt class="py-name">key</tt><tt class="py-op">)</tt><tt class="py-op">,</tt> <tt class="py-name">str</tt><tt class="py-op">(</tt><tt id="link-77" class="py-name"><a title="dose.dose.dose_functions.report
dose.genetic.Population.report" class="py-name" href="#" onclick="return doclink('link-77', 'report', 'link-40');">report</a></tt><tt class="py-op">[</tt><tt class="py-name">key</tt><tt class="py-op">]</tt><tt class="py-op">)</tt><tt class="py-op">]</tt><tt class="py-op">)</tt> </tt>
<a name="L787"></a><tt class="py-lineno">787</tt>  <tt class="py-line">                           <tt class="py-keyword">for</tt> <tt class="py-name">key</tt> <tt class="py-keyword">in</tt> <tt class="py-name">list</tt><tt class="py-op">(</tt><tt id="link-78" class="py-name"><a title="dose.dose.dose_functions.report
dose.genetic.Population.report" class="py-name" href="#" onclick="return doclink('link-78', 'report', 'link-40');">report</a></tt><tt class="py-op">.</tt><tt class="py-name">keys</tt><tt class="py-op">(</tt><tt class="py-op">)</tt><tt class="py-op">)</tt><tt class="py-op">]</tt> </tt>
<a name="L788"></a><tt class="py-lineno">788</tt>  <tt class="py-line">            <tt class="py-name">result</tt><tt class="py-op">.</tt><tt class="py-name">writelines</tt><tt class="py-op">(</tt><tt class="py-string">'|'</tt><tt class="py-op">.</tt><tt class="py-name">join</tt><tt class="py-op">(</tt><tt class="py-name">reportitems</tt><tt class="py-op">)</tt><tt class="py-op">)</tt> </tt>
<a name="L789"></a><tt class="py-lineno">789</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">population</tt><tt class="py-op">.</tt><tt class="py-name">generation</tt> <tt class="py-op">%</tt> <tt class="py-name">int</tt><tt class="py-op">(</tt><tt class="py-name">printfreq</tt><tt class="py-op">)</tt> <tt class="py-op">==</tt> <tt class="py-number">0</tt><tt class="py-op">:</tt> </tt>
<a name="L790"></a><tt class="py-lineno">790</tt>  <tt class="py-line">            <tt class="py-keyword">print</tt><tt class="py-op">(</tt><tt class="py-string">'|'</tt><tt class="py-op">.</tt><tt class="py-name">join</tt><tt class="py-op">(</tt><tt class="py-name">reportitems</tt><tt class="py-op">)</tt><tt class="py-op">)</tt> </tt>
<a name="L791"></a><tt class="py-lineno">791</tt>  <tt class="py-line">        <tt class="py-name">result</tt><tt class="py-op">.</tt><tt class="py-name">writelines</tt><tt class="py-op">(</tt><tt class="py-name">os</tt><tt class="py-op">.</tt><tt class="py-name">linesep</tt><tt class="py-op">)</tt> </tt>
<a name="L792"></a><tt class="py-lineno">792</tt>  <tt class="py-line">        <tt class="py-keyword">if</tt> <tt class="py-name">population</tt><tt class="py-op">.</tt><tt class="py-name">generation</tt> <tt class="py-op">%</tt> <tt class="py-name">int</tt><tt class="py-op">(</tt><tt class="py-name">freezefreq</tt><tt class="py-op">)</tt> <tt class="py-op">==</tt> <tt class="py-number">0</tt><tt class="py-op">:</tt> </tt>
<a name="L793"></a><tt class="py-lineno">793</tt>  <tt class="py-line">            <tt class="py-name">population</tt><tt class="py-op">.</tt><tt id="link-79" class="py-name" targets="Method dose.genetic.Population.freeze()=dose.genetic.Population-class.html#freeze"><a title="dose.genetic.Population.freeze" class="py-name" href="#" onclick="return doclink('link-79', 'freeze', 'link-79');">freeze</a></tt><tt class="py-op">(</tt><tt class="py-name">freezefile</tt><tt class="py-op">,</tt> <tt class="py-name">freezeproportion</tt><tt class="py-op">)</tt> </tt>
<a name="L794"></a><tt class="py-lineno">794</tt>  <tt class="py-line">    <tt class="py-name">population</tt><tt class="py-op">.</tt><tt id="link-80" class="py-name"><a title="dose.genetic.Population.freeze" class="py-name" href="#" onclick="return doclink('link-80', 'freeze', 'link-79');">freeze</a></tt><tt class="py-op">(</tt><tt class="py-name">freezefile</tt><tt class="py-op">,</tt> <tt class="py-number">1.0</tt><tt class="py-op">)</tt> </tt>
<a name="L795"></a><tt class="py-lineno">795</tt>  <tt class="py-line">    <tt class="py-name">result</tt><tt class="py-op">.</tt><tt class="py-name">close</tt><tt class="py-op">(</tt><tt class="py-op">)</tt> </tt>
</div><a name="L796"></a><tt class="py-lineno">796</tt>  <tt class="py-line"> </tt><script type="text/javascript">
<!--
expandto(location.href);
// -->
</script>
</pre>
<br />
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="dose-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            ><a class="navbar" target="_top" href="https://github.com/mauriceling/dose">https://github.com/mauriceling/dose</a></th>
          </tr></table></th>
  </tr>
</table>
<table border="0" cellpadding="0" cellspacing="0" width="100%%">
  <tr>
    <td align="left" class="footer">
    Generated by Epydoc 3.0.1 on Thu Feb 15 15:37:57 2018
    </td>
    <td align="right" class="footer">
      <a target="mainFrame" href="http://epydoc.sourceforge.net"
        >http://epydoc.sourceforge.net</a>
    </td>
  </tr>
</table>

<script type="text/javascript">
  <!--
  // Private objects are initially displayed (because if
  // javascript is turned off then we want them to be
  // visible); but by default, we want to hide them.  So hide
  // them unless we have a cookie that says to show them.
  checkCookie();
  // -->
</script>
</body>
</html>
