<?xml version="1.0" encoding="ascii"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
          "DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
  <title>dose.metadata</title>
  <link rel="stylesheet" href="epydoc.css" type="text/css" />
  <script type="text/javascript" src="epydoc.js"></script>
</head>

<body bgcolor="white" text="black" link="blue" vlink="#204080"
      alink="#204080">
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="dose-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            ><a class="navbar" target="_top" href="https://github.com/mauriceling/dose">https://github.com/mauriceling/dose</a></th>
          </tr></table></th>
  </tr>
</table>
<table width="100%" cellpadding="0" cellspacing="0">
  <tr valign="top">
    <td width="100%">
      <span class="breadcrumbs">
        <a href="dose-module.html">Package&nbsp;dose</a> ::
        Module&nbsp;metadata
      </span>
    </td>
    <td>
      <table cellpadding="0" cellspacing="0">
        <!-- hide/show private -->
        <tr><td align="right"><span class="options">[<a href="javascript:void(0);" class="privatelink"
    onclick="toggle_private();">hide&nbsp;private</a>]</span></td></tr>
        <tr><td align="right"><span class="options"
            >[<a href="frames.html" target="_top">frames</a
            >]&nbsp;|&nbsp;<a href="dose.metadata-pysrc.html"
            target="_top">no&nbsp;frames</a>]</span></td></tr>
      </table>
    </td>
  </tr>
</table>
<h1 class="epydoc">Source Code for <a href="dose.metadata-module.html">Module dose.metadata</a></h1>
<pre class="py-src">
<a name="L1"></a><tt class="py-lineno">  1</tt>  <tt class="py-line"><tt class="py-docstring">'''</tt> </tt>
<a name="L2"></a><tt class="py-lineno">  2</tt>  <tt class="py-line"><tt class="py-docstring">Digital Organism Simulation Environment (DOSE) Metadata</tt> </tt>
<a name="L3"></a><tt class="py-lineno">  3</tt>  <tt class="py-line"><tt class="py-docstring"></tt> </tt>
<a name="L4"></a><tt class="py-lineno">  4</tt>  <tt class="py-line"><tt class="py-docstring">Date created: 14th February 2014</tt> </tt>
<a name="L5"></a><tt class="py-lineno">  5</tt>  <tt class="py-line"><tt class="py-docstring">'''</tt> </tt>
<a name="L6"></a><tt class="py-lineno">  6</tt>  <tt class="py-line"> </tt>
<a name="L7"></a><tt class="py-lineno">  7</tt>  <tt class="py-line"><tt id="link-0" class="py-name" targets="Variable dose.metadata.pypi_name=dose.metadata-module.html#pypi_name"><a title="dose.metadata.pypi_name" class="py-name" href="#" onclick="return doclink('link-0', 'pypi_name', 'link-0');">pypi_name</a></tt> <tt class="py-op">=</tt> <tt class="py-string">'digital organism simulation environment'</tt> </tt>
<a name="L8"></a><tt class="py-lineno">  8</tt>  <tt class="py-line"><tt id="link-1" class="py-name" targets="Variable dose.metadata.description=dose.metadata-module.html#description"><a title="dose.metadata.description" class="py-name" href="#" onclick="return doclink('link-1', 'description', 'link-1');">description</a></tt> <tt class="py-op">=</tt> <tt class="py-string">'Digital Organism Simulation Environment (DOSE)'</tt> </tt>
<a name="L9"></a><tt class="py-lineno">  9</tt>  <tt class="py-line"><tt id="link-2" class="py-name" targets="Variable dose.metadata.version=dose.metadata-module.html#version"><a title="dose.metadata.version" class="py-name" href="#" onclick="return doclink('link-2', 'version', 'link-2');">version</a></tt> <tt class="py-op">=</tt> <tt class="py-string">'1.0.5'</tt> </tt>
<a name="L10"></a><tt class="py-lineno"> 10</tt>  <tt class="py-line"><tt id="link-3" class="py-name" targets="Variable dose.metadata.maintainer=dose.metadata-module.html#maintainer"><a title="dose.metadata.maintainer" class="py-name" href="#" onclick="return doclink('link-3', 'maintainer', 'link-3');">maintainer</a></tt> <tt class="py-op">=</tt> <tt class="py-string">'Maurice HT Ling'</tt> </tt>
<a name="L11"></a><tt class="py-lineno"> 11</tt>  <tt class="py-line"><tt id="link-4" class="py-name" targets="Variable dose.metadata.email=dose.metadata-module.html#email"><a title="dose.metadata.email" class="py-name" href="#" onclick="return doclink('link-4', 'email', 'link-4');">email</a></tt> <tt class="py-op">=</tt> <tt class="py-string">'mauriceling@acm.org'</tt> </tt>
<a name="L12"></a><tt class="py-lineno"> 12</tt>  <tt class="py-line"><tt id="link-5" class="py-name" targets="Variable dose.metadata.download_url=dose.metadata-module.html#download_url"><a title="dose.metadata.download_url" class="py-name" href="#" onclick="return doclink('link-5', 'download_url', 'link-5');">download_url</a></tt> <tt class="py-op">=</tt> <tt class="py-string">'https://github.com/mauriceling/dose/releases'</tt> </tt>
<a name="L13"></a><tt class="py-lineno"> 13</tt>  <tt class="py-line"><tt id="link-6" class="py-name" targets="Variable dose.metadata.project_start_year=dose.metadata-module.html#project_start_year"><a title="dose.metadata.project_start_year" class="py-name" href="#" onclick="return doclink('link-6', 'project_start_year', 'link-6');">project_start_year</a></tt> <tt class="py-op">=</tt> <tt class="py-string">'2010'</tt> </tt>
<a name="L14"></a><tt class="py-lineno"> 14</tt>  <tt class="py-line"><tt id="link-7" class="py-name" targets="Variable dose.metadata.project_website=dose.metadata-module.html#project_website"><a title="dose.metadata.project_website" class="py-name" href="#" onclick="return doclink('link-7', 'project_website', 'link-7');">project_website</a></tt> <tt class="py-op">=</tt> <tt class="py-string">'https://github.com/mauriceling/dose'</tt> </tt>
<a name="L15"></a><tt class="py-lineno"> 15</tt>  <tt class="py-line"><tt id="link-8" class="py-name" targets="Variable dose.metadata.documentation_website=dose.metadata-module.html#documentation_website"><a title="dose.metadata.documentation_website" class="py-name" href="#" onclick="return doclink('link-8', 'documentation_website', 'link-8');">documentation_website</a></tt> <tt class="py-op">=</tt> <tt class="py-string">'http://maurice.vodien.com/project-dose'</tt> </tt>
<a name="L16"></a><tt class="py-lineno"> 16</tt>  <tt class="py-line"><tt id="link-9" class="py-name" targets="Variable dose.metadata.short_license=dose.metadata-module.html#short_license"><a title="dose.metadata.short_license" class="py-name" href="#" onclick="return doclink('link-9', 'short_license', 'link-9');">short_license</a></tt> <tt class="py-op">=</tt> <tt class="py-string">'GNU General Public License version 3'</tt> </tt>
<a name="L17"></a><tt class="py-lineno"> 17</tt>  <tt class="py-line"><tt id="link-10" class="py-name" targets="Variable dose.metadata.long_license=dose.metadata-module.html#long_license"><a title="dose.metadata.long_license" class="py-name" href="#" onclick="return doclink('link-10', 'long_license', 'link-10');">long_license</a></tt> <tt class="py-op">=</tt> <tt class="py-string">'''</tt> </tt>
<a name="L18"></a><tt class="py-lineno"> 18</tt>  <tt class="py-line"><tt class="py-string">Unless otherwise specified, dose.copads package will be licensed </tt> </tt>
<a name="L19"></a><tt class="py-lineno"> 19</tt>  <tt class="py-line"><tt class="py-string">under Python Software Foundation License version 2; all other files will </tt> </tt>
<a name="L20"></a><tt class="py-lineno"> 20</tt>  <tt class="py-line"><tt class="py-string">be licensed GNU General Public License version 3.'''</tt> </tt>
<a name="L21"></a><tt class="py-lineno"> 21</tt>  <tt class="py-line"><tt id="link-11" class="py-name" targets="Variable dose.metadata.platforms=dose.metadata-module.html#platforms"><a title="dose.metadata.platforms" class="py-name" href="#" onclick="return doclink('link-11', 'platforms', 'link-11');">platforms</a></tt> <tt class="py-op">=</tt> <tt class="py-op">[</tt><tt class="py-string">'Operating System :: OS Independent'</tt><tt class="py-op">]</tt> </tt>
<a name="L22"></a><tt class="py-lineno"> 22</tt>  <tt class="py-line"><tt id="link-12" class="py-name" targets="Variable dose.metadata.trove_classifiers=dose.metadata-module.html#trove_classifiers"><a title="dose.metadata.trove_classifiers" class="py-name" href="#" onclick="return doclink('link-12', 'trove_classifiers', 'link-12');">trove_classifiers</a></tt> <tt class="py-op">=</tt> \ </tt>
<a name="L23"></a><tt class="py-lineno"> 23</tt>  <tt class="py-line">    <tt class="py-op">[</tt><tt class="py-string">'Development Status :: 3 - Alpha'</tt><tt class="py-op">,</tt> </tt>
<a name="L24"></a><tt class="py-lineno"> 24</tt>  <tt class="py-line">    <tt class="py-string">'Intended Audience :: Developers'</tt><tt class="py-op">,</tt> </tt>
<a name="L25"></a><tt class="py-lineno"> 25</tt>  <tt class="py-line">    <tt class="py-string">'Intended Audience :: Education'</tt><tt class="py-op">,</tt> </tt>
<a name="L26"></a><tt class="py-lineno"> 26</tt>  <tt class="py-line">    <tt class="py-string">'Intended Audience :: Science/Research'</tt><tt class="py-op">,</tt> </tt>
<a name="L27"></a><tt class="py-lineno"> 27</tt>  <tt class="py-line">    <tt class="py-string">'License :: OSI Approved :: GNU General Public License v3 (GPLv3)'</tt><tt class="py-op">,</tt> </tt>
<a name="L28"></a><tt class="py-lineno"> 28</tt>  <tt class="py-line">    <tt class="py-string">'Natural Language :: English'</tt><tt class="py-op">,</tt> </tt>
<a name="L29"></a><tt class="py-lineno"> 29</tt>  <tt class="py-line">    <tt class="py-string">'Operating System :: OS Independent'</tt><tt class="py-op">,</tt> </tt>
<a name="L30"></a><tt class="py-lineno"> 30</tt>  <tt class="py-line">    <tt class="py-string">'Programming Language :: Python :: 2 :: Only'</tt><tt class="py-op">,</tt> </tt>
<a name="L31"></a><tt class="py-lineno"> 31</tt>  <tt class="py-line">    <tt class="py-string">'Topic :: Education'</tt><tt class="py-op">,</tt> </tt>
<a name="L32"></a><tt class="py-lineno"> 32</tt>  <tt class="py-line">    <tt class="py-string">'Topic :: Scientific/Engineering :: Artificial Life'</tt><tt class="py-op">,</tt> </tt>
<a name="L33"></a><tt class="py-lineno"> 33</tt>  <tt class="py-line">    <tt class="py-string">'Topic :: Scientific/Engineering :: Bio-Informatics'</tt><tt class="py-op">,</tt> </tt>
<a name="L34"></a><tt class="py-lineno"> 34</tt>  <tt class="py-line">    <tt class="py-string">'Topic :: Software Development :: Libraries :: Python Modules'</tt> </tt>
<a name="L35"></a><tt class="py-lineno"> 35</tt>  <tt class="py-line">    <tt class="py-op">]</tt> </tt>
<a name="L36"></a><tt class="py-lineno"> 36</tt>  <tt class="py-line"> </tt>
<a name="L37"></a><tt class="py-lineno"> 37</tt>  <tt class="py-line"><tt class="py-keyword">from</tt> <tt class="py-name">datetime</tt> <tt class="py-keyword">import</tt> <tt class="py-name">datetime</tt> </tt>
<a name="L38"></a><tt class="py-lineno"> 38</tt>  <tt class="py-line"><tt id="link-13" class="py-name" targets="Variable dose.metadata.copyright_year=dose.metadata-module.html#copyright_year"><a title="dose.metadata.copyright_year" class="py-name" href="#" onclick="return doclink('link-13', 'copyright_year', 'link-13');">copyright_year</a></tt> <tt class="py-op">=</tt> <tt class="py-string">'-'</tt><tt class="py-op">.</tt><tt class="py-name">join</tt><tt class="py-op">(</tt><tt class="py-op">[</tt><tt id="link-14" class="py-name"><a title="dose.metadata.project_start_year" class="py-name" href="#" onclick="return doclink('link-14', 'project_start_year', 'link-6');">project_start_year</a></tt><tt class="py-op">,</tt>  </tt>
<a name="L39"></a><tt class="py-lineno"> 39</tt>  <tt class="py-line">                           <tt class="py-name">str</tt><tt class="py-op">(</tt><tt class="py-name">datetime</tt><tt class="py-op">.</tt><tt class="py-name">now</tt><tt class="py-op">(</tt><tt class="py-op">)</tt><tt class="py-op">.</tt><tt class="py-name">year</tt><tt class="py-op">)</tt><tt class="py-op">]</tt><tt class="py-op">)</tt> </tt>
<a name="L40"></a><tt class="py-lineno"> 40</tt>  <tt class="py-line">                            </tt>
<a name="L41"></a><tt class="py-lineno"> 41</tt>  <tt class="py-line"><tt id="link-15" class="py-name" targets="Variable dose.metadata.long_description=dose.metadata-module.html#long_description"><a title="dose.metadata.long_description" class="py-name" href="#" onclick="return doclink('link-15', 'long_description', 'link-15');">long_description</a></tt> <tt class="py-op">=</tt> <tt class="py-string">'''%s</tt> </tt>
<a name="L42"></a><tt class="py-lineno"> 42</tt>  <tt class="py-line"><tt class="py-string"></tt> </tt>
<a name="L43"></a><tt class="py-lineno"> 43</tt>  <tt class="py-line"><tt class="py-string">Life is fascinating and deeply intriguing. Despite so, life forms on Earth </tt> </tt>
<a name="L44"></a><tt class="py-lineno"> 44</tt>  <tt class="py-line"><tt class="py-string">or carbon-based life forms as a group is just one form, one possible sample </tt> </tt>
<a name="L45"></a><tt class="py-lineno"> 45</tt>  <tt class="py-line"><tt class="py-string">of possibly a whole magnitude of life. Even then, there are many aspects of </tt> </tt>
<a name="L46"></a><tt class="py-lineno"> 46</tt>  <tt class="py-line"><tt class="py-string">life that cannot be deciphered even by examining current life forms; for </tt> </tt>
<a name="L47"></a><tt class="py-lineno"> 47</tt>  <tt class="py-line"><tt class="py-string">example, how did chemical reactions organize themselves into biochemical </tt> </tt>
<a name="L48"></a><tt class="py-lineno"> 48</tt>  <tt class="py-line"><tt class="py-string">pathways? How did life start? How is intelligence formed?</tt> </tt>
<a name="L49"></a><tt class="py-lineno"> 49</tt>  <tt class="py-line"><tt class="py-string"></tt> </tt>
<a name="L50"></a><tt class="py-lineno"> 50</tt>  <tt class="py-line"><tt class="py-string">To answer such questions, we will have to restart our evolutionary time to </tt> </tt>
<a name="L51"></a><tt class="py-lineno"> 51</tt>  <tt class="py-line"><tt class="py-string">the very beginning - clearly an impossibly gargantuan task. At the same </tt> </tt>
<a name="L52"></a><tt class="py-lineno"> 52</tt>  <tt class="py-line"><tt class="py-string">time, studying biological/carbon-based life forms is expensive, time </tt> </tt>
<a name="L53"></a><tt class="py-lineno"> 53</tt>  <tt class="py-line"><tt class="py-string">consuming and destructive. As a molecular biologist, there is no way I can </tt> </tt>
<a name="L54"></a><tt class="py-lineno"> 54</tt>  <tt class="py-line"><tt class="py-string">examine the entire genome of even a bacteria in an inanimate state, then </tt> </tt>
<a name="L55"></a><tt class="py-lineno"> 55</tt>  <tt class="py-line"><tt class="py-string">somehow allow it to continue living as if time had just stopped while I </tt> </tt>
<a name="L56"></a><tt class="py-lineno"> 56</tt>  <tt class="py-line"><tt class="py-string">am examining it.</tt> </tt>
<a name="L57"></a><tt class="py-lineno"> 57</tt>  <tt class="py-line"><tt class="py-string"></tt> </tt>
<a name="L58"></a><tt class="py-lineno"> 58</tt>  <tt class="py-line"><tt class="py-string">However, if I can simulate a bacteria or any life form in a computer, </tt> </tt>
<a name="L59"></a><tt class="py-lineno"> 59</tt>  <tt class="py-line"><tt class="py-string">then I can make a digital copy of the bacterium, pull it apart to study it </tt> </tt>
<a name="L60"></a><tt class="py-lineno"> 60</tt>  <tt class="py-line"><tt class="py-string">while the original bacterium continues "living" in my virtual world without </tt> </tt>
<a name="L61"></a><tt class="py-lineno"> 61</tt>  <tt class="py-line"><tt class="py-string">even knowing that it had been duplicated. Many biologists thought of virtual </tt> </tt>
<a name="L62"></a><tt class="py-lineno"> 62</tt>  <tt class="py-line"><tt class="py-string">life forms as a new way to learn about life itself. Studying of virtual life </tt> </tt>
<a name="L63"></a><tt class="py-lineno"> 63</tt>  <tt class="py-line"><tt class="py-string">forms is known as Artificial Life and I term "virtual life forms" as </tt> </tt>
<a name="L64"></a><tt class="py-lineno"> 64</tt>  <tt class="py-line"><tt class="py-string">"digital organisms". There are several advantages in experimenting using </tt> </tt>
<a name="L65"></a><tt class="py-lineno"> 65</tt>  <tt class="py-line"><tt class="py-string">digital organisms. Firstly, generation time can be much faster compared to </tt> </tt>
<a name="L66"></a><tt class="py-lineno"> 66</tt>  <tt class="py-line"><tt class="py-string">most biological life. Secondly, it is usually cheaper to examine computer </tt> </tt>
<a name="L67"></a><tt class="py-lineno"> 67</tt>  <tt class="py-line"><tt class="py-string">simulations than working on actual biological life. Perhaps the most </tt> </tt>
<a name="L68"></a><tt class="py-lineno"> 68</tt>  <tt class="py-line"><tt class="py-string">important advantage of looking at life from this perspective is that by </tt> </tt>
<a name="L69"></a><tt class="py-lineno"> 69</tt>  <tt class="py-line"><tt class="py-string">recreating life in a different medium, we are not limited to our own system </tt> </tt>
<a name="L70"></a><tt class="py-lineno"> 70</tt>  <tt class="py-line"><tt class="py-string">of carbon-based life; hence, studying life as what-it-could-be.</tt> </tt>
<a name="L71"></a><tt class="py-lineno"> 71</tt>  <tt class="py-line"><tt class="py-string"></tt> </tt>
<a name="L72"></a><tt class="py-lineno"> 72</tt>  <tt class="py-line"><tt class="py-string">Digital Organisms Simulation Environment (DOSE) is essentially a virtual </tt> </tt>
<a name="L73"></a><tt class="py-lineno"> 73</tt>  <tt class="py-line"><tt class="py-string">world simulator for studying digital organisms. I will argue that digital </tt> </tt>
<a name="L74"></a><tt class="py-lineno"> 74</tt>  <tt class="py-line"><tt class="py-string">organisms are considered living organisms (Koh and Ling, 2013). Despite so, </tt> </tt>
<a name="L75"></a><tt class="py-lineno"> 75</tt>  <tt class="py-line"><tt class="py-string">being a molecular biologist by training, I have a hard time mapping </tt> </tt>
<a name="L76"></a><tt class="py-lineno"> 76</tt>  <tt class="py-line"><tt class="py-string">components of digital organisms into biological life whenever such </tt> </tt>
<a name="L77"></a><tt class="py-lineno"> 77</tt>  <tt class="py-line"><tt class="py-string">components are too abstract.</tt> </tt>
<a name="L78"></a><tt class="py-lineno"> 78</tt>  <tt class="py-line"><tt class="py-string"></tt> </tt>
<a name="L79"></a><tt class="py-lineno"> 79</tt>  <tt class="py-line"><tt class="py-string">Hence, I decided to design an artificial life / digital organism simulator </tt> </tt>
<a name="L80"></a><tt class="py-lineno"> 80</tt>  <tt class="py-line"><tt class="py-string">that bears resemblance to biological life and ecology. These are the </tt> </tt>
<a name="L81"></a><tt class="py-lineno"> 81</tt>  <tt class="py-line"><tt class="py-string">foundation papers:</tt> </tt>
<a name="L82"></a><tt class="py-lineno"> 82</tt>  <tt class="py-line"><tt class="py-string"></tt> </tt>
<a name="L83"></a><tt class="py-lineno"> 83</tt>  <tt class="py-line"><tt class="py-string">1. Lim, JZR, Aw, ZQ, Goh, DJW, How, JA, Low, SXZ, Loo, BZL, Ling, MHT. 2010. </tt> </tt>
<a name="L84"></a><tt class="py-lineno"> 84</tt>  <tt class="py-line"><tt class="py-string">A genetic algorithm framework grounded in biology. The Python Papers Source </tt> </tt>
<a name="L85"></a><tt class="py-lineno"> 85</tt>  <tt class="py-line"><tt class="py-string">Codes 2: 6. </tt> </tt>
<a name="L86"></a><tt class="py-lineno"> 86</tt>  <tt class="py-line"><tt class="py-string"></tt> </tt>
<a name="L87"></a><tt class="py-lineno"> 87</tt>  <tt class="py-line"><tt class="py-string">This manuscript describes the implementation of a GA framework that uses </tt> </tt>
<a name="L88"></a><tt class="py-lineno"> 88</tt>  <tt class="py-line"><tt class="py-string">biological hierarchy - from chromosomes to organisms to population.</tt> </tt>
<a name="L89"></a><tt class="py-lineno"> 89</tt>  <tt class="py-line"><tt class="py-string"></tt> </tt>
<a name="L90"></a><tt class="py-lineno"> 90</tt>  <tt class="py-line"><tt class="py-string">2. Ling, MHT. 2012. An Artificial Life Simulation Library Based on Genetic </tt> </tt>
<a name="L91"></a><tt class="py-lineno"> 91</tt>  <tt class="py-line"><tt class="py-string">Algorithm, 3-Character Genetic Code and Biological Hierarchy. The Python </tt> </tt>
<a name="L92"></a><tt class="py-lineno"> 92</tt>  <tt class="py-line"><tt class="py-string">Papers 7: 5. </tt> </tt>
<a name="L93"></a><tt class="py-lineno"> 93</tt>  <tt class="py-line"><tt class="py-string"></tt> </tt>
<a name="L94"></a><tt class="py-lineno"> 94</tt>  <tt class="py-line"><tt class="py-string">Genetic algorithm (GA) is inspired by biological evolution of genetic </tt> </tt>
<a name="L95"></a><tt class="py-lineno"> 95</tt>  <tt class="py-line"><tt class="py-string">organisms by optimizing the genotypic combinations encoded within each </tt> </tt>
<a name="L96"></a><tt class="py-lineno"> 96</tt>  <tt class="py-line"><tt class="py-string">individual with the help of evolutionary operators, suggesting that GA </tt> </tt>
<a name="L97"></a><tt class="py-lineno"> 97</tt>  <tt class="py-line"><tt class="py-string">may be a suitable model for studying real-life evolutionary processes. </tt> </tt>
<a name="L98"></a><tt class="py-lineno"> 98</tt>  <tt class="py-line"><tt class="py-string">This paper describes the design of a Python library for artificial life </tt> </tt>
<a name="L99"></a><tt class="py-lineno"> 99</tt>  <tt class="py-line"><tt class="py-string">simulation, Digital Organism Simulation Environment (DOSE), based on GA </tt> </tt>
<a name="L100"></a><tt class="py-lineno">100</tt>  <tt class="py-line"><tt class="py-string">and biological hierarchy starting from genetic sequence to population. </tt> </tt>
<a name="L101"></a><tt class="py-lineno">101</tt>  <tt class="py-line"><tt class="py-string">A 3-character instruction set that does not take any operand is </tt> </tt>
<a name="L102"></a><tt class="py-lineno">102</tt>  <tt class="py-line"><tt class="py-string">introduced as genetic code for digital organism. This mimics the 3-</tt> </tt>
<a name="L103"></a><tt class="py-lineno">103</tt>  <tt class="py-line"><tt class="py-string">nucleotide codon structure in naturally occurring DNA. In addition, </tt> </tt>
<a name="L104"></a><tt class="py-lineno">104</tt>  <tt class="py-line"><tt class="py-string">the context of a 3-dimensional world composing of ecological cells is </tt> </tt>
<a name="L105"></a><tt class="py-lineno">105</tt>  <tt class="py-line"><tt class="py-string">introduced to simulate a physical ecosystem.</tt> </tt>
<a name="L106"></a><tt class="py-lineno">106</tt>  <tt class="py-line"><tt class="py-string"></tt> </tt>
<a name="L107"></a><tt class="py-lineno">107</tt>  <tt class="py-line"><tt class="py-string">- Ling, MHT. 2012. Ragaraja 1.0: The Genome Interpreter of Digital </tt> </tt>
<a name="L108"></a><tt class="py-lineno">108</tt>  <tt class="py-line"><tt class="py-string">Organism Simulation Environment (DOSE). The Python Papers Source Codes </tt> </tt>
<a name="L109"></a><tt class="py-lineno">109</tt>  <tt class="py-line"><tt class="py-string">4: 2. </tt> </tt>
<a name="L110"></a><tt class="py-lineno">110</tt>  <tt class="py-line"><tt class="py-string"></tt> </tt>
<a name="L111"></a><tt class="py-lineno">111</tt>  <tt class="py-line"><tt class="py-string">This manuscript describes the implementation and test of Ragaraja </tt> </tt>
<a name="L112"></a><tt class="py-lineno">112</tt>  <tt class="py-line"><tt class="py-string">instruction set version 1.0, which is the core genomic interpreter of </tt> </tt>
<a name="L113"></a><tt class="py-lineno">113</tt>  <tt class="py-line"><tt class="py-string">DOSE.</tt> </tt>
<a name="L114"></a><tt class="py-lineno">114</tt>  <tt class="py-line"><tt class="py-string"></tt> </tt>
<a name="L115"></a><tt class="py-lineno">115</tt>  <tt class="py-line"><tt class="py-string">From this foundation, the complete suite of Digital Organisms Simulation </tt> </tt>
<a name="L116"></a><tt class="py-lineno">116</tt>  <tt class="py-line"><tt class="py-string">Environment (DOSE) can be build.</tt> </tt>
<a name="L117"></a><tt class="py-lineno">117</tt>  <tt class="py-line"><tt class="py-string"></tt> </tt>
<a name="L118"></a><tt class="py-lineno">118</tt>  <tt class="py-line"><tt class="py-string">Project website: U{%s}</tt> </tt>
<a name="L119"></a><tt class="py-lineno">119</tt>  <tt class="py-line"><tt class="py-string"></tt> </tt>
<a name="L120"></a><tt class="py-lineno">120</tt>  <tt class="py-line"><tt class="py-string">Documentation can be found at %s</tt> </tt>
<a name="L121"></a><tt class="py-lineno">121</tt>  <tt class="py-line"><tt class="py-string"></tt> </tt>
<a name="L122"></a><tt class="py-lineno">122</tt>  <tt class="py-line"><tt class="py-string">License: %s</tt> </tt>
<a name="L123"></a><tt class="py-lineno">123</tt>  <tt class="py-line"><tt class="py-string"></tt> </tt>
<a name="L124"></a><tt class="py-lineno">124</tt>  <tt class="py-line"><tt class="py-string">Copyright %s, Maurice HT Ling (on behalf of all authors).</tt> </tt>
<a name="L125"></a><tt class="py-lineno">125</tt>  <tt class="py-line"><tt class="py-string">'''</tt> <tt class="py-op">%</tt> <tt class="py-op">(</tt><tt id="link-16" class="py-name"><a title="dose.metadata.description" class="py-name" href="#" onclick="return doclink('link-16', 'description', 'link-1');">description</a></tt><tt class="py-op">,</tt> </tt>
<a name="L126"></a><tt class="py-lineno">126</tt>  <tt class="py-line">       <tt id="link-17" class="py-name"><a title="dose.metadata.project_website" class="py-name" href="#" onclick="return doclink('link-17', 'project_website', 'link-7');">project_website</a></tt><tt class="py-op">,</tt>  </tt>
<a name="L127"></a><tt class="py-lineno">127</tt>  <tt class="py-line">       <tt id="link-18" class="py-name"><a title="dose.metadata.documentation_website" class="py-name" href="#" onclick="return doclink('link-18', 'documentation_website', 'link-8');">documentation_website</a></tt><tt class="py-op">,</tt>  </tt>
<a name="L128"></a><tt class="py-lineno">128</tt>  <tt class="py-line">       <tt id="link-19" class="py-name"><a title="dose.metadata.long_license" class="py-name" href="#" onclick="return doclink('link-19', 'long_license', 'link-10');">long_license</a></tt><tt class="py-op">,</tt>  </tt>
<a name="L129"></a><tt class="py-lineno">129</tt>  <tt class="py-line">       <tt id="link-20" class="py-name"><a title="dose.metadata.copyright_year" class="py-name" href="#" onclick="return doclink('link-20', 'copyright_year', 'link-13');">copyright_year</a></tt><tt class="py-op">)</tt> </tt>
<a name="L130"></a><tt class="py-lineno">130</tt>  <tt class="py-line"> </tt><script type="text/javascript">
<!--
expandto(location.href);
// -->
</script>
</pre>
<br />
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="dose-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            ><a class="navbar" target="_top" href="https://github.com/mauriceling/dose">https://github.com/mauriceling/dose</a></th>
          </tr></table></th>
  </tr>
</table>
<table border="0" cellpadding="0" cellspacing="0" width="100%%">
  <tr>
    <td align="left" class="footer">
    Generated by Epydoc 3.0.1 on Thu Feb 15 15:37:58 2018
    </td>
    <td align="right" class="footer">
      <a target="mainFrame" href="http://epydoc.sourceforge.net"
        >http://epydoc.sourceforge.net</a>
    </td>
  </tr>
</table>

<script type="text/javascript">
  <!--
  // Private objects are initially displayed (because if
  // javascript is turned off then we want them to be
  // visible); but by default, we want to hide them.  So hide
  // them unless we have a cookie that says to show them.
  checkCookie();
  // -->
</script>
</body>
</html>
